// @discoveryjs/json-ext/dist/json-ext.js
"use strict";

var _excluded = ["replacer", "getKeys", "space"],
  _excluded2 = ["replacer", "getKeys"];
function _createForOfIteratorHelper(r, e) { var t = "undefined" != typeof Symbol && r[Symbol.iterator] || r["@@iterator"]; if (!t) { if (Array.isArray(r) || (t = _unsupportedIterableToArray(r)) || e && r && "number" == typeof r.length) { t && (r = t); var _n = 0, F = function F() {}; return { s: F, n: function n() { return _n >= r.length ? { done: !0 } : { done: !1, value: r[_n++] }; }, e: function e(r) { throw r; }, f: F }; } throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); } var o, a = !0, u = !1; return { s: function s() { t = t.call(r); }, n: function n() { var r = t.next(); return a = r.done, r; }, e: function e(r) { u = !0, o = r; }, f: function f() { try { a || null == t["return"] || t["return"](); } finally { if (u) throw o; } } }; }
function _regeneratorValues(e) { if (null != e) { var t = e["function" == typeof Symbol && Symbol.iterator || "@@iterator"], r = 0; if (t) return t.call(e); if ("function" == typeof e.next) return e; if (!isNaN(e.length)) return { next: function next() { return e && r >= e.length && (e = void 0), { value: e && e[r++], done: !e }; } }; } throw new TypeError(_typeof(e) + " is not iterable"); }
function _regenerator() { /*! regenerator-runtime -- Copyright (c) 2014-present, Facebook, Inc. -- license (MIT): https://github.com/babel/babel/blob/main/packages/babel-helpers/LICENSE */ var e, t, r = "function" == typeof Symbol ? Symbol : {}, n = r.iterator || "@@iterator", o = r.toStringTag || "@@toStringTag"; function i(r, n, o, i) { var c = n && n.prototype instanceof Generator ? n : Generator, u = Object.create(c.prototype); return _regeneratorDefine2(u, "_invoke", function (r, n, o) { var i, c, u, f = 0, p = o || [], y = !1, G = { p: 0, n: 0, v: e, a: d, f: d.bind(e, 4), d: function d(t, r) { return i = t, c = 0, u = e, G.n = r, a; } }; function d(r, n) { for (c = r, u = n, t = 0; !y && f && !o && t < p.length; t++) { var o, i = p[t], d = G.p, l = i[2]; r > 3 ? (o = l === n) && (u = i[(c = i[4]) ? 5 : (c = 3, 3)], i[4] = i[5] = e) : i[0] <= d && ((o = r < 2 && d < i[1]) ? (c = 0, G.v = n, G.n = i[1]) : d < l && (o = r < 3 || i[0] > n || n > l) && (i[4] = r, i[5] = n, G.n = l, c = 0)); } if (o || r > 1) return a; throw y = !0, n; } return function (o, p, l) { if (f > 1) throw TypeError("Generator is already running"); for (y && 1 === p && d(p, l), c = p, u = l; (t = c < 2 ? e : u) || !y;) { i || (c ? c < 3 ? (c > 1 && (G.n = -1), d(c, u)) : G.n = u : G.v = u); try { if (f = 2, i) { if (c || (o = "next"), t = i[o]) { if (!(t = t.call(i, u))) throw TypeError("iterator result is not an object"); if (!t.done) return t; u = t.value, c < 2 && (c = 0); } else 1 === c && (t = i["return"]) && t.call(i), c < 2 && (u = TypeError("The iterator does not provide a '" + o + "' method"), c = 1); i = e; } else if ((t = (y = G.n < 0) ? u : r.call(n, G)) !== a) break; } catch (t) { i = e, c = 1, u = t; } finally { f = 1; } } return { value: t, done: y }; }; }(r, o, i), !0), u; } var a = {}; function Generator() {} function GeneratorFunction() {} function GeneratorFunctionPrototype() {} t = Object.getPrototypeOf; var c = [][n] ? t(t([][n]())) : (_regeneratorDefine2(t = {}, n, function () { return this; }), t), u = GeneratorFunctionPrototype.prototype = Generator.prototype = Object.create(c); function f(e) { return Object.setPrototypeOf ? Object.setPrototypeOf(e, GeneratorFunctionPrototype) : (e.__proto__ = GeneratorFunctionPrototype, _regeneratorDefine2(e, o, "GeneratorFunction")), e.prototype = Object.create(u), e; } return GeneratorFunction.prototype = GeneratorFunctionPrototype, _regeneratorDefine2(u, "constructor", GeneratorFunctionPrototype), _regeneratorDefine2(GeneratorFunctionPrototype, "constructor", GeneratorFunction), GeneratorFunction.displayName = "GeneratorFunction", _regeneratorDefine2(GeneratorFunctionPrototype, o, "GeneratorFunction"), _regeneratorDefine2(u), _regeneratorDefine2(u, o, "Generator"), _regeneratorDefine2(u, n, function () { return this; }), _regeneratorDefine2(u, "toString", function () { return "[object Generator]"; }), (_regenerator = function _regenerator() { return { w: i, m: f }; })(); }
function _regeneratorDefine2(e, r, n, t) { var i = Object.defineProperty; try { i({}, "", {}); } catch (e) { i = 0; } _regeneratorDefine2 = function _regeneratorDefine(e, r, n, t) { function o(r, n) { _regeneratorDefine2(e, r, function (e) { return this._invoke(r, n, e); }); } r ? i ? i(e, r, { value: n, enumerable: !t, configurable: !t, writable: !t }) : e[r] = n : (o("next", 0), o("throw", 1), o("return", 2)); }, _regeneratorDefine2(e, r, n, t); }
function _objectWithoutProperties(e, t) { if (null == e) return {}; var o, r, i = _objectWithoutPropertiesLoose(e, t); if (Object.getOwnPropertySymbols) { var n = Object.getOwnPropertySymbols(e); for (r = 0; r < n.length; r++) o = n[r], -1 === t.indexOf(o) && {}.propertyIsEnumerable.call(e, o) && (i[o] = e[o]); } return i; }
function _objectWithoutPropertiesLoose(r, e) { if (null == r) return {}; var t = {}; for (var n in r) if ({}.hasOwnProperty.call(r, n)) { if (-1 !== e.indexOf(n)) continue; t[n] = r[n]; } return t; }
function asyncGeneratorStep(n, t, e, r, o, a, c) { try { var i = n[a](c), u = i.value; } catch (n) { return void e(n); } i.done ? t(u) : Promise.resolve(u).then(r, o); }
function _asyncToGenerator(n) { return function () { var t = this, e = arguments; return new Promise(function (r, o) { var a = n.apply(t, e); function _next(n) { asyncGeneratorStep(a, r, o, _next, _throw, "next", n); } function _throw(n) { asyncGeneratorStep(a, r, o, _next, _throw, "throw", n); } _next(void 0); }); }; }
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function _defineProperty(e, r, t) { return (r = _toPropertyKey(r)) in e ? Object.defineProperty(e, r, { value: t, enumerable: !0, configurable: !0, writable: !0 }) : e[r] = t, e; }
function _toPropertyKey(t) { var i = _toPrimitive(t, "string"); return "symbol" == _typeof(i) ? i : i + ""; }
function _toPrimitive(t, r) { if ("object" != _typeof(t) || !t) return t; var e = t[Symbol.toPrimitive]; if (void 0 !== e) { var i = e.call(t, r || "default"); if ("object" != _typeof(i)) return i; throw new TypeError("@@toPrimitive must return a primitive value."); } return ("string" === r ? String : Number)(t); }
function _toConsumableArray(r) { return _arrayWithoutHoles(r) || _iterableToArray(r) || _unsupportedIterableToArray(r) || _nonIterableSpread(); }
function _nonIterableSpread() { throw new TypeError("Invalid attempt to spread non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }
function _unsupportedIterableToArray(r, a) { if (r) { if ("string" == typeof r) return _arrayLikeToArray(r, a); var t = {}.toString.call(r).slice(8, -1); return "Object" === t && r.constructor && (t = r.constructor.name), "Map" === t || "Set" === t ? Array.from(r) : "Arguments" === t || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(t) ? _arrayLikeToArray(r, a) : void 0; } }
function _iterableToArray(r) { if ("undefined" != typeof Symbol && null != r[Symbol.iterator] || null != r["@@iterator"]) return Array.from(r); }
function _arrayWithoutHoles(r) { if (Array.isArray(r)) return _arrayLikeToArray(r); }
function _arrayLikeToArray(r, a) { (null == a || a > r.length) && (a = r.length); for (var e = 0, n = Array(a); e < a; e++) n[e] = r[e]; return n; }
function _typeof(o) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (o) { return typeof o; } : function (o) { return o && "function" == typeof Symbol && o.constructor === Symbol && o !== Symbol.prototype ? "symbol" : typeof o; }, _typeof(o); }
function _awaitAsyncGenerator(e) { return new _OverloadYield(e, 0); }
function _wrapAsyncGenerator(e) { return function () { return new AsyncGenerator(e.apply(this, arguments)); }; }
function AsyncGenerator(e) { var t, n; function resume(t, n) { try { var r = e[t](n), o = r.value, u = o instanceof _OverloadYield; Promise.resolve(u ? o.v : o).then(function (n) { if (u) { var i = "return" === t && o.k ? t : "next"; if (!o.k || n.done) return resume(i, n); n = e[i](n).value; } settle(!!r.done, n); }, function (e) { resume("throw", e); }); } catch (e) { settle(2, e); } } function settle(e, r) { 2 === e ? t.reject(r) : t.resolve({ value: r, done: e }), (t = t.next) ? resume(t.key, t.arg) : n = null; } this._invoke = function (e, r) { return new Promise(function (o, u) { var i = { key: e, arg: r, resolve: o, reject: u, next: null }; n ? n = n.next = i : (t = n = i, resume(e, r)); }); }, "function" != typeof e["return"] && (this["return"] = void 0); }
AsyncGenerator.prototype["function" == typeof Symbol && Symbol.asyncIterator || "@@asyncIterator"] = function () { return this; }, AsyncGenerator.prototype.next = function (e) { return this._invoke("next", e); }, AsyncGenerator.prototype["throw"] = function (e) { return this._invoke("throw", e); }, AsyncGenerator.prototype["return"] = function (e) { return this._invoke("return", e); };
function _OverloadYield(e, d) { this.v = e, this.k = d; }
function _asyncIterator(r) { var n, t, o, e = 2; for ("undefined" != typeof Symbol && (t = Symbol.asyncIterator, o = Symbol.iterator); e--;) { if (t && null != (n = r[t])) return n.call(r); if (o && null != (n = r[o])) return new AsyncFromSyncIterator(n.call(r)); t = "@@asyncIterator", o = "@@iterator"; } throw new TypeError("Object is not async iterable"); }
function AsyncFromSyncIterator(r) { function AsyncFromSyncIteratorContinuation(r) { if (Object(r) !== r) return Promise.reject(new TypeError(r + " is not an object.")); var n = r.done; return Promise.resolve(r.value).then(function (r) { return { value: r, done: n }; }); } return AsyncFromSyncIterator = function AsyncFromSyncIterator(r) { this.s = r, this.n = r.next; }, AsyncFromSyncIterator.prototype = { s: null, n: null, next: function next() { return AsyncFromSyncIteratorContinuation(this.n.apply(this.s, arguments)); }, "return": function _return(r) { var n = this.s["return"]; return void 0 === n ? Promise.resolve({ value: r, done: !0 }) : AsyncFromSyncIteratorContinuation(n.apply(this.s, arguments)); }, "throw": function _throw(r) { var n = this.s["return"]; return void 0 === n ? Promise.reject(r) : AsyncFromSyncIteratorContinuation(n.apply(this.s, arguments)); } }, new AsyncFromSyncIterator(r); }
(function (global, factory) {
  (typeof exports === "undefined" ? "undefined" : _typeof(exports)) === 'object' && typeof module !== 'undefined' ? module.exports = factory() : typeof define === 'function' && define.amd ? define(factory) : global.jsonExt = factory();
})(typeof globalThis != 'undefined' ? globalThis : typeof window != 'undefined' ? window : typeof global != 'undefined' ? global : typeof self != 'undefined' ? self : void 0, function () {
  var _marked = /*#__PURE__*/_regenerator().m(stringifyChunked);
  // src/utils.js
  function isIterable(value) {
    return _typeof(value) === "object" && value !== null && (typeof value[Symbol.iterator] === "function" || typeof value[Symbol.asyncIterator] === "function");
  }
  function replaceValue(holder, key, value, replacer) {
    if (value && typeof value.toJSON === "function") {
      value = value.toJSON();
    }
    if (replacer !== null) {
      value = replacer.call(holder, String(key), value);
    }
    switch (_typeof(value)) {
      case "function":
      case "symbol":
        value = void 0;
        break;
      case "object":
        if (value !== null) {
          var cls = value.constructor;
          if (cls === String || cls === Number || cls === Boolean) {
            value = value.valueOf();
          }
        }
        break;
    }
    return value;
  }
  function normalizeReplacer(replacer) {
    if (typeof replacer === "function") {
      return replacer;
    }
    if (Array.isArray(replacer)) {
      var allowlist = new Set(replacer.map(function (item) {
        var cls = item && item.constructor;
        return cls === String || cls === Number ? String(item) : null;
      }).filter(function (item) {
        return typeof item === "string";
      }));
      return _toConsumableArray(allowlist);
    }
    return null;
  }
  function normalizeSpace(space) {
    if (typeof space === "number") {
      if (!Number.isFinite(space) || space < 1) {
        return false;
      }
      return " ".repeat(Math.min(space, 10));
    }
    if (typeof space === "string") {
      return space.slice(0, 10) || false;
    }
    return false;
  }
  function normalizeStringifyOptions(optionsOrReplacer, space) {
    if (optionsOrReplacer === null || Array.isArray(optionsOrReplacer) || _typeof(optionsOrReplacer) !== "object") {
      optionsOrReplacer = {
        replacer: optionsOrReplacer,
        space: space
      };
    }
    var replacer = normalizeReplacer(optionsOrReplacer.replacer);
    var getKeys = Object.keys;
    if (Array.isArray(replacer)) {
      var allowlist = replacer;
      getKeys = function getKeys() {
        return allowlist;
      };
      replacer = null;
    }
    return _objectSpread(_objectSpread({}, optionsOrReplacer), {}, {
      replacer: replacer,
      getKeys: getKeys,
      space: normalizeSpace(optionsOrReplacer.space)
    });
  }
  function resolveStringifyMode() {
    var mode = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : "json";
    if (mode === "json" || mode === "jsonl") {
      return mode;
    }
    throw new TypeError('Invalid options: `mode` should be "json" or "jsonl"');
  }

  // src/parse-chunked.js
  var NO_VALUE = /* @__PURE__ */Symbol("empty");
  var STACK_OBJECT = 1;
  var STACK_ARRAY = 2;
  var MODE_JSON = 0;
  var MODE_JSONL = 1;
  var MODE_JSONL_AUTO = 2;
  var decoder = new TextDecoder();
  function adjustPosition(error, jsonParseOffset) {
    if (error.name === "SyntaxError" && jsonParseOffset) {
      error.message = error.message.replace(/at position (\d+)/, function (_, pos) {
        return "at position " + (Number(pos) + jsonParseOffset);
      });
    }
    return error;
  }
  function append(array, elements) {
    var initialLength = array.length;
    array.length += elements.length;
    for (var i = 0; i < elements.length; i++) {
      array[initialLength + i] = elements[i];
    }
  }
  function resolveParseMode(mode) {
    switch (mode) {
      case "json":
        return MODE_JSON;
      case "jsonl":
        return MODE_JSONL;
      case "auto":
        return MODE_JSONL_AUTO;
      default:
        throw new TypeError('Invalid options: `mode` should be "json", "jsonl", or "auto"');
    }
  }
  function parseChunkedOptions(value) {
    var _options$mode, _options$reviver, _options$onRootValue, _options$onChunk;
    var options = typeof value === "function" ? {
      reviver: value
    } : value || {};
    return {
      mode: resolveParseMode((_options$mode = options.mode) !== null && _options$mode !== void 0 ? _options$mode : "json"),
      reviver: (_options$reviver = options.reviver) !== null && _options$reviver !== void 0 ? _options$reviver : null,
      onRootValue: (_options$onRootValue = options.onRootValue) !== null && _options$onRootValue !== void 0 ? _options$onRootValue : null,
      onChunk: (_options$onChunk = options.onChunk) !== null && _options$onChunk !== void 0 ? _options$onChunk : null
    };
  }
  function applyReviver(value, reviver) {
    return walk({
      "": value
    }, "", value);
    function walk(holder, key, value2) {
      if (value2 && _typeof(value2) === "object") {
        for (var _i = 0, _Object$keys = Object.keys(value2); _i < _Object$keys.length; _i++) {
          var childKey = _Object$keys[_i];
          var childValue = value2[childKey];
          var newValue = walk(value2, childKey, childValue);
          if (newValue === void 0) {
            delete value2[childKey];
          } else if (newValue !== childValue) {
            value2[childKey] = newValue;
          }
        }
      }
      return reviver.call(holder, key, value2);
    }
  }
  function parseChunked(_x, _x2) {
    return _parseChunked.apply(this, arguments);
  }
  function _parseChunked() {
    _parseChunked = _asyncToGenerator(/*#__PURE__*/_regenerator().m(function _callee2(chunkEmitter, optionsOrReviver) {
      var _parseChunkedOptions, mode, reviver, onRootValue, onChunk, iterable, parser, _iteratorAbruptCompletion, _didIteratorError, _iteratorError, _iterator, _step, chunk, _t, _t2;
      return _regenerator().w(function (_context4) {
        while (1) switch (_context4.p = _context4.n) {
          case 0:
            _parseChunkedOptions = parseChunkedOptions(optionsOrReviver), mode = _parseChunkedOptions.mode, reviver = _parseChunkedOptions.reviver, onRootValue = _parseChunkedOptions.onRootValue, onChunk = _parseChunkedOptions.onChunk;
            iterable = typeof chunkEmitter === "function" ? chunkEmitter() : chunkEmitter;
            if (!isIterable(iterable)) {
              _context4.n = 16;
              break;
            }
            parser = createChunkParser(mode, reviver, onRootValue, onChunk);
            _context4.p = 1;
            _iteratorAbruptCompletion = false;
            _didIteratorError = false;
            _context4.p = 2;
            _iterator = _asyncIterator(iterable);
          case 3:
            _context4.n = 4;
            return _iterator.next();
          case 4:
            if (!(_iteratorAbruptCompletion = !(_step = _context4.v).done)) {
              _context4.n = 7;
              break;
            }
            chunk = _step.value;
            if (!(typeof chunk !== "string" && !ArrayBuffer.isView(chunk))) {
              _context4.n = 5;
              break;
            }
            throw new TypeError("Invalid chunk: Expected string, TypedArray or Buffer");
          case 5:
            parser.push(chunk);
          case 6:
            _iteratorAbruptCompletion = false;
            _context4.n = 3;
            break;
          case 7:
            _context4.n = 9;
            break;
          case 8:
            _context4.p = 8;
            _t = _context4.v;
            _didIteratorError = true;
            _iteratorError = _t;
          case 9:
            _context4.p = 9;
            _context4.p = 10;
            if (!(_iteratorAbruptCompletion && _iterator["return"] != null)) {
              _context4.n = 11;
              break;
            }
            _context4.n = 11;
            return _iterator["return"]();
          case 11:
            _context4.p = 11;
            if (!_didIteratorError) {
              _context4.n = 12;
              break;
            }
            throw _iteratorError;
          case 12:
            return _context4.f(11);
          case 13:
            return _context4.f(9);
          case 14:
            return _context4.a(2, parser.finish());
          case 15:
            _context4.p = 15;
            _t2 = _context4.v;
            throw adjustPosition(_t2, parser.jsonParseOffset);
          case 16:
            throw new TypeError("Invalid chunk emitter: Expected an Iterable, AsyncIterable, generator, async generator, or a function returning an Iterable or AsyncIterable");
          case 17:
            return _context4.a(2);
        }
      }, _callee2, null, [[10,, 11, 13], [2, 8, 9, 14], [1, 15]]);
    }));
    return _parseChunked.apply(this, arguments);
  }
  function createChunkParser(parseMode, reviver, onRootValue, onChunk) {
    var rootValues = parseMode === MODE_JSONL ? [] : null;
    var rootValuesCount = 0;
    var currentRootValue = NO_VALUE;
    var currentRootValueCursor = null;
    var consumedChunkLength = 0;
    var parsedChunkLength = 0;
    var prevArray = null;
    var prevArraySlices = [];
    var stack = new Array(100);
    var lastFlushDepth = 0;
    var flushDepth = 0;
    var stateString = false;
    var stateStringEscape = false;
    var seenNonWhiteSpace = false;
    var allowNewRootValue = true;
    var pendingByteSeq = null;
    var pendingChunk = null;
    var jsonParseOffset = 0;
    var state = Object.freeze({
      get mode() {
        return parseMode === MODE_JSONL ? "jsonl" : "json";
      },
      get returnValue() {
        return typeof onRootValue === "function" ? rootValuesCount : rootValues !== null ? rootValues : currentRootValue !== NO_VALUE ? currentRootValue : void 0;
      },
      get currentRootValue() {
        return currentRootValue !== NO_VALUE ? currentRootValue : void 0;
      },
      get rootValuesCount() {
        return rootValuesCount;
      },
      get consumed() {
        return consumedChunkLength;
      },
      get parsed() {
        return parsedChunkLength;
      }
    });
    return {
      push: push,
      finish: finish,
      state: state,
      get jsonParseOffset() {
        return jsonParseOffset;
      }
    };
    function startRootValue(fragment) {
      if (!allowNewRootValue) {
        jsonParseOffset -= 2;
        JSON.parse("[]" + fragment);
      }
      if (currentRootValue !== NO_VALUE && parseMode === MODE_JSONL_AUTO) {
        parseMode = MODE_JSONL;
        rootValues = [currentRootValue];
      }
      allowNewRootValue = false;
      currentRootValue = JSON.parse(fragment);
    }
    function finishRootValue() {
      rootValuesCount++;
      if (typeof reviver === "function") {
        currentRootValue = applyReviver(currentRootValue, reviver);
      }
      if (typeof onRootValue === "function") {
        onRootValue(currentRootValue, state);
      } else if (parseMode === MODE_JSONL) {
        rootValues.push(currentRootValue);
      }
    }
    function mergeArraySlices() {
      if (prevArray === null) {
        return;
      }
      if (prevArraySlices.length !== 0) {
        var _prevArray;
        var newArray = prevArraySlices.length === 1 ? prevArray.concat(prevArraySlices[0]) : (_prevArray = prevArray).concat.apply(_prevArray, _toConsumableArray(prevArraySlices));
        if (currentRootValueCursor.prev !== null) {
          currentRootValueCursor.prev.value[currentRootValueCursor.key] = newArray;
        } else {
          currentRootValue = newArray;
        }
        currentRootValueCursor.value = newArray;
        prevArraySlices = [];
      }
      prevArray = null;
    }
    function parseAndAppend(fragment, wrap) {
      if (stack[lastFlushDepth - 1] === STACK_OBJECT) {
        if (wrap) {
          jsonParseOffset--;
          fragment = "{" + fragment + "}";
        }
        Object.assign(currentRootValueCursor.value, JSON.parse(fragment));
      } else {
        if (wrap) {
          jsonParseOffset--;
          fragment = "[" + fragment + "]";
        }
        if (prevArray === currentRootValueCursor.value) {
          prevArraySlices.push(JSON.parse(fragment));
        } else {
          append(currentRootValueCursor.value, JSON.parse(fragment));
          prevArray = currentRootValueCursor.value;
        }
      }
    }
    function prepareAddition(fragment) {
      var _currentRootValueCurs = currentRootValueCursor,
        value = _currentRootValueCurs.value;
      var expectComma = Array.isArray(value) ? value.length !== 0 : Object.keys(value).length !== 0;
      if (expectComma) {
        if (fragment[0] === ",") {
          jsonParseOffset++;
          return fragment.slice(1);
        }
        if (fragment[0] !== "}" && fragment[0] !== "]") {
          jsonParseOffset -= 3;
          return "[[]" + fragment;
        }
      }
      return fragment;
    }
    function flush(chunk, start, end) {
      var fragment = chunk.slice(start, end);
      jsonParseOffset = consumedChunkLength + start;
      parsedChunkLength += end - start;
      if (pendingChunk !== null) {
        fragment = pendingChunk + fragment;
        jsonParseOffset -= pendingChunk.length;
        parsedChunkLength += pendingChunk.length;
        pendingChunk = null;
      }
      if (flushDepth === lastFlushDepth) {
        if (lastFlushDepth === 0) {
          startRootValue(fragment);
        } else {
          parseAndAppend(prepareAddition(fragment), true);
        }
      } else if (flushDepth > lastFlushDepth) {
        for (var i = flushDepth - 1; i >= lastFlushDepth; i--) {
          fragment += stack[i] === STACK_OBJECT ? "}" : "]";
        }
        if (lastFlushDepth === 0) {
          startRootValue(fragment);
          currentRootValueCursor = {
            value: currentRootValue,
            key: null,
            prev: null
          };
        } else {
          parseAndAppend(prepareAddition(fragment), true);
          mergeArraySlices();
        }
        for (var _i2 = lastFlushDepth || 1; _i2 < flushDepth; _i2++) {
          var _currentRootValueCurs2 = currentRootValueCursor,
            value = _currentRootValueCurs2.value;
          var key = null;
          if (stack[_i2 - 1] === STACK_OBJECT) {
            for (key in value);
            value = value[key];
          } else {
            key = value.length - 1;
            value = value[key];
          }
          currentRootValueCursor = {
            value: value,
            key: key,
            prev: currentRootValueCursor
          };
        }
      } else {
        fragment = prepareAddition(fragment);
        for (var _i3 = lastFlushDepth - 1; _i3 >= flushDepth; _i3--) {
          jsonParseOffset--;
          fragment = (stack[_i3] === STACK_OBJECT ? "{" : "[") + fragment;
        }
        parseAndAppend(fragment, false);
        mergeArraySlices();
        for (var _i4 = lastFlushDepth - 1; _i4 >= flushDepth; _i4--) {
          currentRootValueCursor = currentRootValueCursor.prev;
        }
      }
      if (flushDepth === 0) {
        finishRootValue();
      }
      lastFlushDepth = flushDepth;
      seenNonWhiteSpace = false;
    }
    function ensureChunkString(chunk) {
      if (typeof chunk !== "string") {
        if (pendingByteSeq !== null) {
          var origRawChunk = chunk;
          chunk = new Uint8Array(pendingByteSeq.length + origRawChunk.length);
          chunk.set(pendingByteSeq);
          chunk.set(origRawChunk, pendingByteSeq.length);
          pendingByteSeq = null;
        }
        if (chunk[chunk.length - 1] > 127) {
          for (var seqLength = 0; seqLength < chunk.length; seqLength++) {
            var _byte = chunk[chunk.length - 1 - seqLength];
            if (_byte >> 6 === 3) {
              seqLength++;
              if (seqLength !== 4 && _byte >> 3 === 30 || seqLength !== 3 && _byte >> 4 === 14 || seqLength !== 2 && _byte >> 5 === 6) {
                pendingByteSeq = chunk.slice(chunk.length - seqLength);
                chunk = chunk.subarray(0, -seqLength);
              }
              break;
            }
          }
        }
        chunk = decoder.decode(chunk);
      }
      return chunk;
    }
    function push(chunk) {
      chunk = ensureChunkString(chunk);
      var chunkLength = chunk.length;
      var prevParsedChunkLength = parsedChunkLength;
      var lastFlushPoint = 0;
      var flushPoint = 0;
      scan: for (var i = 0; i < chunkLength; i++) {
        if (stateString) {
          for (; i < chunkLength; i++) {
            if (stateStringEscape) {
              stateStringEscape = false;
            } else {
              switch (chunk.charCodeAt(i)) {
                case 34:
                  stateString = false;
                  continue scan;
                case 92:
                  stateStringEscape = true;
              }
            }
          }
          break;
        }
        switch (chunk.charCodeAt(i)) {
          case 34:
            stateString = true;
            stateStringEscape = false;
            seenNonWhiteSpace = true;
            break;
          case 44:
            flushPoint = i;
            break;
          case 123:
            flushPoint = i + 1;
            stack[flushDepth++] = STACK_OBJECT;
            seenNonWhiteSpace = true;
            break;
          case 91:
            flushPoint = i + 1;
            stack[flushDepth++] = STACK_ARRAY;
            seenNonWhiteSpace = true;
            break;
          case 93:
          /* ] */
          case 125:
            flushPoint = i + 1;
            if (flushDepth === 0) {
              break scan;
            }
            flushDepth--;
            if (flushDepth < lastFlushDepth) {
              flush(chunk, lastFlushPoint, flushPoint);
              lastFlushPoint = flushPoint;
            }
            break;
          case 9:
          /* \t */
          case 10:
          /* \n */
          case 13:
          /* \r */
          case 32:
            if (flushDepth === 0) {
              if (seenNonWhiteSpace) {
                flushPoint = i;
                flush(chunk, lastFlushPoint, flushPoint);
                lastFlushPoint = flushPoint;
              }
              if (parseMode !== MODE_JSON && allowNewRootValue === false && (chunk.charCodeAt(i) === 10 || chunk.charCodeAt(i) === 13)) {
                allowNewRootValue = true;
              }
              if (flushPoint === i) {
                parsedChunkLength++;
              }
            }
            if (lastFlushPoint === i) {
              lastFlushPoint++;
            }
            if (flushPoint === i) {
              flushPoint++;
            }
            break;
          default:
            seenNonWhiteSpace = true;
        }
      }
      if (flushPoint > lastFlushPoint) {
        flush(chunk, lastFlushPoint, flushPoint);
      }
      if (flushPoint < chunkLength) {
        if (pendingChunk !== null) {
          pendingChunk += chunk;
        } else {
          pendingChunk = chunk.slice(flushPoint, chunkLength);
        }
      }
      consumedChunkLength += chunkLength;
      if (typeof onChunk === "function") {
        onChunk(parsedChunkLength - prevParsedChunkLength, chunk, pendingChunk, state);
      }
    }
    function finish() {
      if (pendingChunk !== null || currentRootValue === NO_VALUE && parseMode !== MODE_JSONL) {
        flushDepth = 0;
        flush("", 0, 0);
      }
      if (typeof onChunk === "function") {
        parsedChunkLength = consumedChunkLength;
        onChunk(0, null, null, state);
      }
      var result = state.returnValue;
      rootValues = null;
      currentRootValue = NO_VALUE;
      return result;
    }
  }

  // src/stringify-chunked.js
  function encodeString(value) {
    if (/[^\x20\x21\x23-\x5B\x5D-\uD799]/.test(value)) {
      return JSON.stringify(value);
    }
    return '"' + value + '"';
  }
  function stringifyChunked(value) {
    var _len,
      args,
      _key,
      _normalizeStringifyOp,
      replacer,
      getKeys,
      space,
      options,
      highWaterMark,
      roots,
      rootCount,
      keyStrings,
      stack,
      rootValue,
      prevState,
      state,
      stateValue,
      stateEmpty,
      stateKeys,
      stateIndex,
      buffer,
      _loop,
      i,
      printObject,
      printArray,
      printEntryPrelude,
      printEntry,
      pushPrimitive,
      pushState,
      popState,
      _args2 = arguments;
    return _regenerator().w(function (_context2) {
      while (1) switch (_context2.n) {
        case 0:
          popState = function _popState() {
            stack.pop();
            var value2 = stack.length > 0 ? stack[stack.length - 1] : rootValue;
            state = Array.isArray(value2) ? printArray : printObject;
            stateValue = value2;
            stateEmpty = false;
            stateKeys = prevState.keys;
            stateIndex = prevState.index;
            prevState = prevState.prev;
          };
          pushState = function _pushState() {
            prevState = {
              keys: stateKeys,
              index: stateIndex,
              prev: prevState
            };
          };
          pushPrimitive = function _pushPrimitive(value2) {
            var _value2$constructor;
            switch (_typeof(value2)) {
              case "string":
                buffer += encodeString(value2);
                break;
              case "number":
                buffer += Number.isFinite(value2) ? String(value2) : "null";
                break;
              case "boolean":
                buffer += value2 ? "true" : "false";
                break;
              case "undefined":
              case "object":
                buffer += "null";
                break;
              default:
                throw new TypeError("Do not know how to serialize a ".concat(((_value2$constructor = value2.constructor) === null || _value2$constructor === void 0 ? void 0 : _value2$constructor.name) || _typeof(value2)));
            }
          };
          printEntry = function _printEntry(key, value2) {
            value2 = replaceValue(stateValue, key, value2, replacer);
            if (value2 === null || _typeof(value2) !== "object") {
              if (state !== printObject || value2 !== void 0) {
                printEntryPrelude(key);
                pushPrimitive(value2);
              }
            } else {
              if (stack.includes(value2)) {
                throw new TypeError("Converting circular structure to JSON");
              }
              printEntryPrelude(key);
              stack.push(value2);
              pushState();
              state = Array.isArray(value2) ? printArray : printObject;
              stateValue = value2;
              stateEmpty = true;
              stateIndex = 0;
            }
          };
          printEntryPrelude = function _printEntryPrelude(key) {
            if (stateEmpty) {
              stateEmpty = false;
            } else {
              buffer += ",";
            }
            if (space && prevState !== null) {
              buffer += "\n".concat(space.repeat(stack.length));
            }
            if (state === printObject) {
              var keyString = keyStrings.get(key);
              if (keyString === void 0) {
                keyStrings.set(key, keyString = encodeString(key) + (space ? ": " : ":"));
              }
              buffer += keyString;
            }
          };
          printArray = function _printArray() {
            if (stateIndex === 0) {
              buffer += "[";
            }
            if (stateIndex === stateValue.length) {
              buffer += space && !stateEmpty ? "\n".concat(space.repeat(stack.length - 1), "]") : "]";
              popState();
              return;
            }
            printEntry(stateIndex, stateValue[stateIndex++]);
          };
          printObject = function _printObject() {
            if (stateIndex === 0) {
              stateKeys = getKeys(stateValue);
              buffer += "{";
            }
            if (stateIndex === stateKeys.length) {
              buffer += space && !stateEmpty ? "\n".concat(space.repeat(stack.length - 1), "}") : "}";
              popState();
              return;
            }
            var key = stateKeys[stateIndex++];
            printEntry(key, stateValue[key]);
          };
          for (_len = _args2.length, args = new Array(_len > 1 ? _len - 1 : 0), _key = 1; _key < _len; _key++) {
            args[_key - 1] = _args2[_key];
          }
          _normalizeStringifyOp = normalizeStringifyOptions.apply(void 0, args), replacer = _normalizeStringifyOp.replacer, getKeys = _normalizeStringifyOp.getKeys, space = _normalizeStringifyOp.space, options = _objectWithoutProperties(_normalizeStringifyOp, _excluded);
          highWaterMark = Number(options.highWaterMark) || 16384;
          roots = resolveStringifyMode(options.mode) === "jsonl" && Array.isArray(value) ? value : [value];
          rootCount = roots.length;
          keyStrings = /* @__PURE__ */new Map();
          stack = [];
          rootValue = null;
          prevState = null;
          state = null;
          stateValue = null;
          stateEmpty = true;
          stateKeys = [];
          stateIndex = 0;
          buffer = "";
          _loop = /*#__PURE__*/_regenerator().m(function _loop(i) {
            return _regenerator().w(function (_context) {
              while (1) switch (_context.n) {
                case 0:
                  if (rootValue !== null) {
                    buffer += "\n";
                  }
                  rootValue = {
                    "": roots[i]
                  };
                  prevState = null;
                  state = function state() {
                    return printEntry("", roots[i]);
                  };
                  stateValue = rootValue;
                  stateEmpty = true;
                  stateKeys = [""];
                  stateIndex = 0;
                case 1:
                  state();
                  if (!(buffer.length >= highWaterMark || prevState === null && i === rootCount - 1)) {
                    _context.n = 3;
                    break;
                  }
                  _context.n = 2;
                  return buffer;
                case 2:
                  buffer = "";
                case 3:
                  if (prevState !== null) {
                    _context.n = 1;
                    break;
                  }
                case 4:
                  return _context.a(2);
              }
            }, _loop);
          });
          i = 0;
        case 1:
          if (!(i < rootCount)) {
            _context2.n = 3;
            break;
          }
          return _context2.d(_regeneratorValues(_loop(i)), 2);
        case 2:
          i++;
          _context2.n = 1;
          break;
        case 3:
          return _context2.a(2);
      }
    }, _marked);
  }

  // src/stringify-info.js
  var hasOwn = typeof Object.hasOwn === "function" ? Object.hasOwn : function (object, key) {
    return Object.hasOwnProperty.call(object, key);
  };
  var escapableCharCodeSubstitution = {
    // JSON Single Character Escape Sequences
    8: "\\b",
    9: "\\t",
    10: "\\n",
    12: "\\f",
    13: "\\r",
    34: '\\"',
    92: "\\\\"
  };
  var charLength2048 = Uint8Array.from({
    length: 2048
  }, function (_, code) {
    if (hasOwn(escapableCharCodeSubstitution, code)) {
      return 2;
    }
    if (code < 32) {
      return 6;
    }
    return code < 128 ? 1 : 2;
  });
  function isLeadingSurrogate(code) {
    return code >= 55296 && code <= 56319;
  }
  function isTrailingSurrogate(code) {
    return code >= 56320 && code <= 57343;
  }
  function stringLength(str) {
    if (!/[^\x20\x21\x23-\x5B\x5D-\x7F]/.test(str)) {
      return str.length + 2;
    }
    var len = 0;
    var prevLeadingSurrogate = false;
    for (var i = 0; i < str.length; i++) {
      var code = str.charCodeAt(i);
      if (code < 2048) {
        len += charLength2048[code];
      } else if (isLeadingSurrogate(code)) {
        len += 6;
        prevLeadingSurrogate = true;
        continue;
      } else if (isTrailingSurrogate(code)) {
        len = prevLeadingSurrogate ? len - 2 : len + 6;
      } else {
        len += 3;
      }
      prevLeadingSurrogate = false;
    }
    return len + 2;
  }
  function intLength(num) {
    var len = 0;
    if (num < 0) {
      len = 1;
      num = -num;
    }
    if (num >= 1e9) {
      len += 9;
      num = (num - num % 1e9) / 1e9;
    }
    if (num >= 1e4) {
      if (num >= 1e6) {
        return len + (num >= 1e8 ? 9 : num >= 1e7 ? 8 : 7);
      }
      return len + (num >= 1e5 ? 6 : 5);
    }
    return len + (num >= 100 ? num >= 1e3 ? 4 : 3 : num >= 10 ? 2 : 1);
  }
  function primitiveLength(value) {
    switch (_typeof(value)) {
      case "string":
        return stringLength(value);
      case "number":
        return Number.isFinite(value) ? Number.isInteger(value) ? intLength(value) : String(value).length : 4;
      case "boolean":
        return value ? 4 : 5;
      case "undefined":
      case "object":
        return 4;
      /* null */
      default:
        return 0;
    }
  }
  function stringifyInfo(value) {
    var _options$space;
    for (var _len2 = arguments.length, args = new Array(_len2 > 1 ? _len2 - 1 : 0), _key2 = 1; _key2 < _len2; _key2++) {
      args[_key2 - 1] = arguments[_key2];
    }
    var _normalizeStringifyOp2 = normalizeStringifyOptions.apply(void 0, args),
      replacer = _normalizeStringifyOp2.replacer,
      getKeys = _normalizeStringifyOp2.getKeys,
      options = _objectWithoutProperties(_normalizeStringifyOp2, _excluded2);
    var continueOnCircular = Boolean(options.continueOnCircular);
    var space = ((_options$space = options.space) === null || _options$space === void 0 ? void 0 : _options$space.length) || 0;
    var roots = resolveStringifyMode(options.mode) === "jsonl" && Array.isArray(value) ? value : [value];
    var keysLength = /* @__PURE__ */new Map();
    var visited = /* @__PURE__ */new Map();
    var circular = /* @__PURE__ */new Set();
    var stack = [];
    var stop = false;
    var bytes = 0;
    var spaceBytes = 0;
    var objects = 0;
    for (var i = 0; i < roots.length; i++) {
      if (i > 0) {
        bytes += 1;
      }
      walk({
        "": roots[i]
      }, "", roots[i]);
    }
    if (bytes === 0 && roots.length === 1) {
      bytes += 9;
    }
    return {
      bytes: isNaN(bytes) ? Infinity : bytes + spaceBytes,
      spaceBytes: space > 0 && isNaN(bytes) ? Infinity : spaceBytes,
      circular: _toConsumableArray(circular)
    };
    function walk(holder, key, value2) {
      if (stop) {
        return;
      }
      value2 = replaceValue(holder, key, value2, replacer);
      if (value2 === null || _typeof(value2) !== "object") {
        if (value2 !== void 0 || Array.isArray(holder)) {
          bytes += primitiveLength(value2);
        }
      } else {
        if (stack.includes(value2)) {
          circular.add(value2);
          bytes += 4;
          if (!continueOnCircular) {
            stop = true;
          }
          return;
        }
        if (visited.has(value2)) {
          bytes += visited.get(value2);
          return;
        }
        objects++;
        var prevObjects = objects;
        var valueBytes = bytes;
        var valueLength = 0;
        stack.push(value2);
        if (Array.isArray(value2)) {
          valueLength = value2.length;
          for (var _i5 = 0; _i5 < valueLength; _i5++) {
            walk(value2, _i5, value2[_i5]);
          }
        } else {
          var prevLength = bytes;
          var _iterator2 = _createForOfIteratorHelper(getKeys(value2)),
            _step2;
          try {
            for (_iterator2.s(); !(_step2 = _iterator2.n()).done;) {
              var key2 = _step2.value;
              walk(value2, key2, value2[key2]);
              if (prevLength !== bytes) {
                var keyLen = keysLength.get(key2);
                if (keyLen === void 0) {
                  keysLength.set(key2, keyLen = stringLength(key2) + 1);
                }
                bytes += keyLen;
                valueLength++;
                prevLength = bytes;
              }
            }
          } catch (err) {
            _iterator2.e(err);
          } finally {
            _iterator2.f();
          }
        }
        bytes += valueLength === 0 ? 2 : 1 + valueLength;
        if (space > 0 && valueLength > 0) {
          spaceBytes +=
          // a space between ":" and a value for each object entry
          (Array.isArray(value2) ? 0 : valueLength) +
          // the formula results from folding the following components:
          // - for each key-value or element: ident + newline
          //   (1 + stack.length * space) * valueLength
          // - ident (one space less) before "}" or "]" + newline
          //   (stack.length - 1) * space + 1
          (1 + stack.length * space) * (valueLength + 1) - space;
        }
        stack.pop();
        if (prevObjects !== objects) {
          visited.set(value2, bytes - valueBytes);
        }
      }
    }
  }

  // src/web-streams.js
  function parseFromWebStream(stream) {
    return parseChunked(isIterable(stream) ? stream : /*#__PURE__*/_wrapAsyncGenerator(/*#__PURE__*/_regenerator().m(function _callee() {
      var reader, _yield$_awaitAsyncGen, value, done;
      return _regenerator().w(function (_context3) {
        while (1) switch (_context3.p = _context3.n) {
          case 0:
            reader = stream.getReader();
            _context3.p = 1;
          case 2:
            if (!true) {
              _context3.n = 6;
              break;
            }
            _context3.n = 3;
            return _awaitAsyncGenerator(reader.read());
          case 3:
            _yield$_awaitAsyncGen = _context3.v;
            value = _yield$_awaitAsyncGen.value;
            done = _yield$_awaitAsyncGen.done;
            if (!done) {
              _context3.n = 4;
              break;
            }
            return _context3.a(3, 6);
          case 4:
            _context3.n = 5;
            return value;
          case 5:
            _context3.n = 2;
            break;
          case 6:
            _context3.p = 6;
            reader.releaseLock();
            return _context3.f(6);
          case 7:
            return _context3.a(2);
        }
      }, _callee, null, [[1,, 6, 7]]);
    })));
  }
  function createStringifyWebStream(value, replacer, space) {
    if (typeof ReadableStream.from === "function") {
      return ReadableStream.from(stringifyChunked(value, replacer, space));
    }
    return new ReadableStream({
      start: function start() {
        this.generator = stringifyChunked(value, replacer, space);
      },
      pull: function pull(controller) {
        var _this$generator$next = this.generator.next(),
          value2 = _this$generator$next.value,
          done = _this$generator$next.done;
        if (done) {
          controller.close();
        } else {
          controller.enqueue(value2);
        }
      },
      cancel: function cancel() {
        this.generator = null;
      }
    });
  }
  return {
    createStringifyWebStream: createStringifyWebStream,
    parseChunked: parseChunked,
    parseFromWebStream: parseFromWebStream,
    stringifyChunked: stringifyChunked,
    stringifyInfo: stringifyInfo
  };
});
