// @vitejs/plugin-react/dist/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = viteReact;
exports["module.exports"] = viteReactForCjs;
exports.reactCompilerPreset = void 0;
var _nodeFs = require("node:fs");
var _nodePath = require("node:path");
var _nodeUrl = require("node:url");
var _pluginutils = require("@rolldown/pluginutils");
var _internal = require("vite/internal");
function _regenerator() { /*! regenerator-runtime -- Copyright (c) 2014-present, Facebook, Inc. -- license (MIT): https://github.com/babel/babel/blob/main/packages/babel-helpers/LICENSE */ var e, t, r = "function" == typeof Symbol ? Symbol : {}, n = r.iterator || "@@iterator", o = r.toStringTag || "@@toStringTag"; function i(r, n, o, i) { var c = n && n.prototype instanceof Generator ? n : Generator, u = Object.create(c.prototype); return _regeneratorDefine2(u, "_invoke", function (r, n, o) { var i, c, u, f = 0, p = o || [], y = !1, G = { p: 0, n: 0, v: e, a: d, f: d.bind(e, 4), d: function d(t, r) { return i = t, c = 0, u = e, G.n = r, a; } }; function d(r, n) { for (c = r, u = n, t = 0; !y && f && !o && t < p.length; t++) { var o, i = p[t], d = G.p, l = i[2]; r > 3 ? (o = l === n) && (u = i[(c = i[4]) ? 5 : (c = 3, 3)], i[4] = i[5] = e) : i[0] <= d && ((o = r < 2 && d < i[1]) ? (c = 0, G.v = n, G.n = i[1]) : d < l && (o = r < 3 || i[0] > n || n > l) && (i[4] = r, i[5] = n, G.n = l, c = 0)); } if (o || r > 1) return a; throw y = !0, n; } return function (o, p, l) { if (f > 1) throw TypeError("Generator is already running"); for (y && 1 === p && d(p, l), c = p, u = l; (t = c < 2 ? e : u) || !y;) { i || (c ? c < 3 ? (c > 1 && (G.n = -1), d(c, u)) : G.n = u : G.v = u); try { if (f = 2, i) { if (c || (o = "next"), t = i[o]) { if (!(t = t.call(i, u))) throw TypeError("iterator result is not an object"); if (!t.done) return t; u = t.value, c < 2 && (c = 0); } else 1 === c && (t = i["return"]) && t.call(i), c < 2 && (u = TypeError("The iterator does not provide a '" + o + "' method"), c = 1); i = e; } else if ((t = (y = G.n < 0) ? u : r.call(n, G)) !== a) break; } catch (t) { i = e, c = 1, u = t; } finally { f = 1; } } return { value: t, done: y }; }; }(r, o, i), !0), u; } var a = {}; function Generator() {} function GeneratorFunction() {} function GeneratorFunctionPrototype() {} t = Object.getPrototypeOf; var c = [][n] ? t(t([][n]())) : (_regeneratorDefine2(t = {}, n, function () { return this; }), t), u = GeneratorFunctionPrototype.prototype = Generator.prototype = Object.create(c); function f(e) { return Object.setPrototypeOf ? Object.setPrototypeOf(e, GeneratorFunctionPrototype) : (e.__proto__ = GeneratorFunctionPrototype, _regeneratorDefine2(e, o, "GeneratorFunction")), e.prototype = Object.create(u), e; } return GeneratorFunction.prototype = GeneratorFunctionPrototype, _regeneratorDefine2(u, "constructor", GeneratorFunctionPrototype), _regeneratorDefine2(GeneratorFunctionPrototype, "constructor", GeneratorFunction), GeneratorFunction.displayName = "GeneratorFunction", _regeneratorDefine2(GeneratorFunctionPrototype, o, "GeneratorFunction"), _regeneratorDefine2(u), _regeneratorDefine2(u, o, "Generator"), _regeneratorDefine2(u, n, function () { return this; }), _regeneratorDefine2(u, "toString", function () { return "[object Generator]"; }), (_regenerator = function _regenerator() { return { w: i, m: f }; })(); }
function _regeneratorDefine2(e, r, n, t) { var i = Object.defineProperty; try { i({}, "", {}); } catch (e) { i = 0; } _regeneratorDefine2 = function _regeneratorDefine(e, r, n, t) { function o(r, n) { _regeneratorDefine2(e, r, function (e) { return this._invoke(r, n, e); }); } r ? i ? i(e, r, { value: n, enumerable: !t, configurable: !t, writable: !t }) : e[r] = n : (o("next", 0), o("throw", 1), o("return", 2)); }, _regeneratorDefine2(e, r, n, t); }
function asyncGeneratorStep(n, t, e, r, o, a, c) { try { var i = n[a](c), u = i.value; } catch (n) { return void e(n); } i.done ? t(u) : Promise.resolve(u).then(r, o); }
function _asyncToGenerator(n) { return function () { var t = this, e = arguments; return new Promise(function (r, o) { var a = n.apply(t, e); function _next(n) { asyncGeneratorStep(a, r, o, _next, _throw, "next", n); } function _throw(n) { asyncGeneratorStep(a, r, o, _next, _throw, "throw", n); } _next(void 0); }); }; }
//#region ../common/refresh-utils.ts
var runtimePublicPath = "/@react-refresh";
var preambleCode = "import { injectIntoGlobalHook } from \"__BASE__".concat(runtimePublicPath.slice(1), "\";\ninjectIntoGlobalHook(window);\nwindow.$RefreshReg$ = () => {};\nwindow.$RefreshSig$ = () => (type) => type;");
var getPreambleCode = function getPreambleCode(base) {
  return preambleCode.replace("__BASE__", base);
};
function virtualPreamblePlugin(_ref) {
  var name = _ref.name,
    isEnabled = _ref.isEnabled;
  return {
    name: "vite:react-virtual-preamble",
    resolveId: {
      order: "pre",
      filter: {
        id: (0, _pluginutils.exactRegex)(name)
      },
      handler: function handler(source) {
        if (source === name) return "\0" + source;
      }
    },
    load: {
      filter: {
        id: (0, _pluginutils.exactRegex)("\0" + name)
      },
      handler: function handler(id) {
        if (id === "\0" + name) {
          if (isEnabled()) return preambleCode.replace("__BASE__", "/");
          return "";
        }
      }
    }
  };
}
//#endregion
//#region ../common/warning.ts
var silenceUseClientWarning = function silenceUseClientWarning(userConfig) {
  return {
    rollupOptions: {
      onwarn: function onwarn(warning, defaultHandler) {
        var _userConfig$build;
        if (warning.code === "MODULE_LEVEL_DIRECTIVE" && (warning.message.includes("use client") || warning.message.includes("use server"))) return;
        if (warning.code === "SOURCEMAP_ERROR" && warning.message.includes("resolve original location") && warning.pos === 0) return;
        if ((_userConfig$build = userConfig.build) !== null && _userConfig$build !== void 0 && (_userConfig$build = _userConfig$build.rollupOptions) !== null && _userConfig$build !== void 0 && _userConfig$build.onwarn) userConfig.build.rollupOptions.onwarn(warning, defaultHandler);else defaultHandler(warning);
      }
    }
  };
};
//#endregion
//#region src/reactCompilerPreset.ts
var reactCompilerPreset = exports.reactCompilerPreset = function reactCompilerPreset() {
  var options = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  return {
    preset: function preset() {
      return {
        plugins: [["babel-plugin-react-compiler", options]]
      };
    },
    rolldown: {
      filter: {
        code: options.compilationMode === "annotation" ? /['"]use memo['"]/ : /\b[A-Z]|\buse/
      },
      applyToEnvironmentHook: function applyToEnvironmentHook(env) {
        return env.config.consumer === "client";
      },
      optimizeDeps: {
        include: options.target === "17" || options.target === "18" ? ["react-compiler-runtime"] : ["react/compiler-runtime"]
      }
    }
  };
};
//#endregion
//#region src/index.ts
var refreshRuntimePath = (0, _nodePath.join)((0, _nodePath.dirname)((0, _nodeUrl.fileURLToPath)(import.meta.url)), "refresh-runtime.js");
var defaultIncludeRE = /\.[tj]sx?$/;
var defaultExcludeRE = /\/node_modules\//;
function viteReact() {
  var _opts$include, _opts$exclude, _opts$jsxImportSource;
  var opts = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  var include = (_opts$include = opts.include) !== null && _opts$include !== void 0 ? _opts$include : defaultIncludeRE;
  var exclude = (_opts$exclude = opts.exclude) !== null && _opts$exclude !== void 0 ? _opts$exclude : defaultExcludeRE;
  var jsxImportSource = (_opts$jsxImportSource = opts.jsxImportSource) !== null && _opts$jsxImportSource !== void 0 ? _opts$jsxImportSource : "react";
  var jsxImportRuntime = "".concat(jsxImportSource, "/jsx-runtime");
  var jsxImportDevRuntime = "".concat(jsxImportSource, "/jsx-dev-runtime");
  var runningInVite = false;
  var isProduction = true;
  var skipFastRefresh = true;
  var base;
  var isBundledDev = false;
  var viteBabel = {
    name: "vite:react-babel",
    enforce: "pre",
    config: function config(_userConfig, _ref2) {
      var command = _ref2.command;
      if (opts.jsxRuntime === "classic") return {
        oxc: {
          jsx: {
            runtime: "classic",
            refresh: command === "serve"
          },
          jsxRefreshInclude: (0, _pluginutils.makeIdFiltersToMatchWithQuery)(include),
          jsxRefreshExclude: (0, _pluginutils.makeIdFiltersToMatchWithQuery)(exclude)
        }
      };else return {
        oxc: {
          jsx: {
            runtime: "automatic",
            importSource: opts.jsxImportSource,
            refresh: command === "serve"
          },
          jsxRefreshInclude: (0, _pluginutils.makeIdFiltersToMatchWithQuery)(include),
          jsxRefreshExclude: (0, _pluginutils.makeIdFiltersToMatchWithQuery)(exclude)
        },
        optimizeDeps: {
          rolldownOptions: {
            transform: {
              jsx: {
                runtime: "automatic"
              }
            }
          }
        }
      };
    },
    configResolved: function configResolved(config) {
      runningInVite = true;
      base = config.base;
      if (config.experimental.bundledDev) isBundledDev = true;
      isProduction = config.isProduction;
      skipFastRefresh = isProduction || config.command === "build" || config.server.hmr === false;
    },
    options: function options(_options) {
      if (!runningInVite) {
        var _options$transform;
        (_options$transform = _options.transform) !== null && _options$transform !== void 0 ? _options$transform : _options.transform = {};
        _options.transform.jsx = {
          runtime: opts.jsxRuntime,
          importSource: opts.jsxImportSource
        };
        return _options;
      }
    }
  };
  var viteRefreshWrapper = {
    name: "vite:react:refresh-wrapper",
    apply: "serve",
    applyToEnvironment: function applyToEnvironment(env) {
      return _asyncToGenerator(/*#__PURE__*/_regenerator().m(function _callee() {
        var _opts$reactRefreshHos;
        return _regenerator().w(function (_context) {
          while (1) switch (_context.n) {
            case 0:
              if (!(env.config.consumer !== "client" || skipFastRefresh)) {
                _context.n = 1;
                break;
              }
              return _context.a(2, false);
            case 1:
              return _context.a(2, (0, _internal.reactRefreshWrapperPlugin)({
                cwd: process.cwd(),
                include: (0, _pluginutils.makeIdFiltersToMatchWithQuery)(include),
                exclude: (0, _pluginutils.makeIdFiltersToMatchWithQuery)(exclude),
                jsxImportSource: jsxImportSource,
                reactRefreshHost: (_opts$reactRefreshHos = opts.reactRefreshHost) !== null && _opts$reactRefreshHos !== void 0 ? _opts$reactRefreshHos : ""
              }));
          }
        }, _callee);
      }))();
    }
  };
  var viteConfigPost = {
    name: "vite:react:config-post",
    enforce: "post",
    config: function config(userConfig) {
      var _userConfig$server;
      if (((_userConfig$server = userConfig.server) === null || _userConfig$server === void 0 ? void 0 : _userConfig$server.hmr) === false) return {
        oxc: {
          jsx: {
            refresh: false
          }
        }
      };
    }
  };
  var viteReactRefreshBundledDevMode = {
    name: "vite:react-refresh-fbm",
    enforce: "pre",
    transformIndexHtml: {
      handler: function handler() {
        if (!skipFastRefresh && isBundledDev) return [{
          tag: "script",
          attrs: {
            type: "module"
          },
          children: getPreambleCode(base)
        }];
      },
      order: "pre"
    }
  };
  var dependencies = ["react", "react-dom", jsxImportDevRuntime, jsxImportRuntime];
  return [viteBabel, viteRefreshWrapper, viteConfigPost, viteReactRefreshBundledDevMode, {
    name: "vite:react-refresh",
    enforce: "pre",
    config: function config(userConfig) {
      return {
        build: silenceUseClientWarning(userConfig),
        optimizeDeps: {
          include: dependencies
        }
      };
    },
    resolveId: {
      filter: {
        id: (0, _pluginutils.exactRegex)(runtimePublicPath)
      },
      handler: function handler(id) {
        if (id === "/@react-refresh") return id;
      }
    },
    load: {
      filter: {
        id: (0, _pluginutils.exactRegex)(runtimePublicPath)
      },
      handler: function handler(id) {
        if (id === "/@react-refresh") return (0, _nodeFs.readFileSync)(refreshRuntimePath, "utf-8").replace(/__README_URL__/g, "https://github.com/vitejs/vite-plugin-react/tree/main/packages/plugin-react");
      }
    },
    transformIndexHtml: function transformIndexHtml() {
      if (!skipFastRefresh && !isBundledDev) return [{
        tag: "script",
        attrs: {
          type: "module"
        },
        children: getPreambleCode(base)
      }];
    }
  }, virtualPreamblePlugin({
    name: "@vitejs/plugin-react/preamble",
    isEnabled: function isEnabled() {
      return !skipFastRefresh && !isBundledDev;
    }
  })];
}
viteReact.preambleCode = preambleCode;
function viteReactForCjs(options) {
  return viteReact.call(this, options);
}
Object.assign(viteReactForCjs, {
  "default": viteReactForCjs,
  reactCompilerPreset: reactCompilerPreset
});
//#endregion
