// @jridgewell/source-map/dist/source-map.umd.js
"use strict";

function _classCallCheck(a, n) { if (!(a instanceof n)) throw new TypeError("Cannot call a class as a function"); }
function _defineProperties(e, r) { for (var t = 0; t < r.length; t++) { var o = r[t]; o.enumerable = o.enumerable || !1, o.configurable = !0, "value" in o && (o.writable = !0), Object.defineProperty(e, _toPropertyKey(o.key), o); } }
function _createClass(e, r, t) { return r && _defineProperties(e.prototype, r), t && _defineProperties(e, t), Object.defineProperty(e, "prototype", { writable: !1 }), e; }
function _toPropertyKey(t) { var i = _toPrimitive(t, "string"); return "symbol" == _typeof(i) ? i : i + ""; }
function _toPrimitive(t, r) { if ("object" != _typeof(t) || !t) return t; var e = t[Symbol.toPrimitive]; if (void 0 !== e) { var i = e.call(t, r || "default"); if ("object" != _typeof(i)) return i; throw new TypeError("@@toPrimitive must return a primitive value."); } return ("string" === r ? String : Number)(t); }
function _createForOfIteratorHelper(r, e) { var t = "undefined" != typeof Symbol && r[Symbol.iterator] || r["@@iterator"]; if (!t) { if (Array.isArray(r) || (t = _unsupportedIterableToArray(r)) || e && r && "number" == typeof r.length) { t && (r = t); var _n = 0, F = function F() {}; return { s: F, n: function n() { return _n >= r.length ? { done: !0 } : { done: !1, value: r[_n++] }; }, e: function e(r) { throw r; }, f: F }; } throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); } var o, a = !0, u = !1; return { s: function s() { t = t.call(r); }, n: function n() { var r = t.next(); return a = r.done, r; }, e: function e(r) { u = !0, o = r; }, f: function f() { try { a || null == t["return"] || t["return"](); } finally { if (u) throw o; } } }; }
function _unsupportedIterableToArray(r, a) { if (r) { if ("string" == typeof r) return _arrayLikeToArray(r, a); var t = {}.toString.call(r).slice(8, -1); return "Object" === t && r.constructor && (t = r.constructor.name), "Map" === t || "Set" === t ? Array.from(r) : "Arguments" === t || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(t) ? _arrayLikeToArray(r, a) : void 0; } }
function _arrayLikeToArray(r, a) { (null == a || a > r.length) && (a = r.length); for (var e = 0, n = Array(a); e < a; e++) n[e] = r[e]; return n; }
function _typeof(o) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (o) { return typeof o; } : function (o) { return o && "function" == typeof Symbol && o.constructor === Symbol && o !== Symbol.prototype ? "symbol" : typeof o; }, _typeof(o); }
(function (global, factory) {
  if ((typeof exports === "undefined" ? "undefined" : _typeof(exports)) === 'object' && typeof module !== 'undefined') {
    factory(module, require('@jridgewell/gen-mapping'), require('@jridgewell/trace-mapping'));
    module.exports = def(module);
  } else if (typeof define === 'function' && define.amd) {
    define(['module', '@jridgewell/gen-mapping', '@jridgewell/trace-mapping'], function (mod) {
      factory.apply(this, arguments);
      mod.exports = def(mod);
    });
  } else {
    var mod = {
      exports: {}
    };
    factory(mod, global.genMapping, global.traceMapping);
    global = typeof globalThis !== 'undefined' ? globalThis : global || self;
    global.sourceMap = def(mod);
  }
  function def(m) {
    return 'default' in m.exports ? m.exports["default"] : m.exports;
  }
})(void 0, function (module, require_genMapping, require_traceMapping) {
  "use strict";

  var __create = Object.create;
  var __defProp = Object.defineProperty;
  var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
  var __getOwnPropNames = Object.getOwnPropertyNames;
  var __getProtoOf = Object.getPrototypeOf;
  var __hasOwnProp = Object.prototype.hasOwnProperty;
  var __commonJS = function __commonJS(cb, mod) {
    return function __require() {
      return mod || (0, cb[__getOwnPropNames(cb)[0]])((mod = {
        exports: {}
      }).exports, mod), mod.exports;
    };
  };
  var __export = function __export(target, all) {
    for (var name in all) __defProp(target, name, {
      get: all[name],
      enumerable: true
    });
  };
  var __copyProps = function __copyProps(to, from, except, desc) {
    if (from && _typeof(from) === "object" || typeof from === "function") {
      var _iterator = _createForOfIteratorHelper(__getOwnPropNames(from)),
        _step;
      try {
        var _loop = function _loop() {
          var key = _step.value;
          if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
            get: function get() {
              return from[key];
            },
            enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
          });
        };
        for (_iterator.s(); !(_step = _iterator.n()).done;) {
          _loop();
        }
      } catch (err) {
        _iterator.e(err);
      } finally {
        _iterator.f();
      }
    }
    return to;
  };
  var __toESM = function __toESM(mod, isNodeMode, target) {
    return target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
    // If the importer is in node compatibility mode or this is not an ESM
    // file that has been converted to a CommonJS file using a Babel-
    // compatible transform (i.e. "__esModule" has not been set), then set
    // "default" to the CommonJS "module.exports" for node compatibility.
    isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", {
      value: mod,
      enumerable: true
    }) : target, mod);
  };
  var __toCommonJS = function __toCommonJS(mod) {
    return __copyProps(__defProp({}, "__esModule", {
      value: true
    }), mod);
  };

  // umd:@jridgewell/trace-mapping
  var require_trace_mapping = __commonJS({
    "umd:@jridgewell/trace-mapping": function umdJridgewell_traceMapping(exports, module2) {
      module2.exports = require_traceMapping;
    }
  });

  // umd:@jridgewell/gen-mapping
  var require_gen_mapping = __commonJS({
    "umd:@jridgewell/gen-mapping": function umdJridgewell_genMapping(exports, module2) {
      module2.exports = require_genMapping;
    }
  });

  // src/source-map.ts
  var source_map_exports = {};
  __export(source_map_exports, {
    SourceMapConsumer: function SourceMapConsumer() {
      return _SourceMapConsumer2;
    },
    SourceMapGenerator: function SourceMapGenerator() {
      return _SourceMapGenerator2;
    }
  });
  module.exports = __toCommonJS(source_map_exports);
  var import_trace_mapping = __toESM(require_trace_mapping());
  var import_gen_mapping = __toESM(require_gen_mapping());
  var _SourceMapConsumer2 = /*#__PURE__*/function () {
    function _SourceMapConsumer(map, mapUrl) {
      _classCallCheck(this, _SourceMapConsumer);
      var trace = this._map = new import_trace_mapping.AnyMap(map, mapUrl);
      this.file = trace.file;
      this.names = trace.names;
      this.sourceRoot = trace.sourceRoot;
      this.sources = trace.resolvedSources;
      this.sourcesContent = trace.sourcesContent;
      this.version = trace.version;
    }
    return _createClass(_SourceMapConsumer, [{
      key: "mappings",
      get: function get() {
        return (0, import_trace_mapping.encodedMappings)(this._map);
      }
    }, {
      key: "originalPositionFor",
      value: function originalPositionFor(needle) {
        return (0, import_trace_mapping.originalPositionFor)(this._map, needle);
      }
    }, {
      key: "generatedPositionFor",
      value: function generatedPositionFor(originalPosition) {
        return (0, import_trace_mapping.generatedPositionFor)(this._map, originalPosition);
      }
    }, {
      key: "allGeneratedPositionsFor",
      value: function allGeneratedPositionsFor(originalPosition) {
        return (0, import_trace_mapping.allGeneratedPositionsFor)(this._map, originalPosition);
      }
    }, {
      key: "hasContentsOfAllSources",
      value: function hasContentsOfAllSources() {
        if (!this.sourcesContent || this.sourcesContent.length !== this.sources.length) {
          return false;
        }
        var _iterator2 = _createForOfIteratorHelper(this.sourcesContent),
          _step2;
        try {
          for (_iterator2.s(); !(_step2 = _iterator2.n()).done;) {
            var content = _step2.value;
            if (content == null) {
              return false;
            }
          }
        } catch (err) {
          _iterator2.e(err);
        } finally {
          _iterator2.f();
        }
        return true;
      }
    }, {
      key: "sourceContentFor",
      value: function sourceContentFor(source, nullOnMissing) {
        var sourceContent = (0, import_trace_mapping.sourceContentFor)(this._map, source);
        if (sourceContent != null) {
          return sourceContent;
        }
        if (nullOnMissing) {
          return null;
        }
        throw new Error("\"".concat(source, "\" is not in the SourceMap."));
      }
    }, {
      key: "eachMapping",
      value: function eachMapping(callback, context) {
        (0, import_trace_mapping.eachMapping)(this._map, context ? callback.bind(context) : callback);
      }
    }, {
      key: "destroy",
      value: function destroy() {}
    }], [{
      key: "fromSourceMap",
      value: function fromSourceMap(map, mapUrl) {
        if (map.toDecodedMap) {
          return new _SourceMapConsumer(map.toDecodedMap(), mapUrl);
        }
        return new _SourceMapConsumer(map.toJSON(), mapUrl);
      }
    }]);
  }();
  var _SourceMapGenerator2 = /*#__PURE__*/function () {
    function _SourceMapGenerator(opts) {
      _classCallCheck(this, _SourceMapGenerator);
      this._map = opts instanceof import_gen_mapping.GenMapping ? opts : new import_gen_mapping.GenMapping(opts);
    }
    return _createClass(_SourceMapGenerator, [{
      key: "addMapping",
      value: function addMapping(mapping) {
        (0, import_gen_mapping.maybeAddMapping)(this._map, mapping);
      }
    }, {
      key: "setSourceContent",
      value: function setSourceContent(source, content) {
        (0, import_gen_mapping.setSourceContent)(this._map, source, content);
      }
    }, {
      key: "toJSON",
      value: function toJSON() {
        return (0, import_gen_mapping.toEncodedMap)(this._map);
      }
    }, {
      key: "toString",
      value: function toString() {
        return JSON.stringify(this.toJSON());
      }
    }, {
      key: "toDecodedMap",
      value: function toDecodedMap() {
        return (0, import_gen_mapping.toDecodedMap)(this._map);
      }
    }], [{
      key: "fromSourceMap",
      value: function fromSourceMap(consumer) {
        return new _SourceMapGenerator((0, import_gen_mapping.fromMap)(consumer));
      }
    }]);
  }();
});
