// @rolldown/pluginutils/dist/filter/composable-filters.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.and = and;
exports.code = code;
exports.exclude = exclude;
exports.exprInterpreter = exprInterpreter;
exports.id = id;
exports.importerId = importerId;
exports.include = include;
exports.interpreter = interpreter;
exports.interpreterImpl = interpreterImpl;
exports.moduleType = moduleType;
exports.not = not;
exports.or = or;
exports.queries = queries;
exports.query = query;
var _utils = require("../utils.js");
function _typeof(o) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (o) { return typeof o; } : function (o) { return o && "function" == typeof Symbol && o.constructor === Symbol && o !== Symbol.prototype ? "symbol" : typeof o; }, _typeof(o); }
function _createForOfIteratorHelper(r, e) { var t = "undefined" != typeof Symbol && r[Symbol.iterator] || r["@@iterator"]; if (!t) { if (Array.isArray(r) || (t = _unsupportedIterableToArray(r)) || e && r && "number" == typeof r.length) { t && (r = t); var _n = 0, F = function F() {}; return { s: F, n: function n() { return _n >= r.length ? { done: !0 } : { done: !1, value: r[_n++] }; }, e: function e(r) { throw r; }, f: F }; } throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); } var o, a = !0, u = !1; return { s: function s() { t = t.call(r); }, n: function n() { var r = t.next(); return a = r.done, r; }, e: function e(r) { u = !0, o = r; }, f: function f() { try { a || null == t["return"] || t["return"](); } finally { if (u) throw o; } } }; }
function _toConsumableArray(r) { return _arrayWithoutHoles(r) || _iterableToArray(r) || _unsupportedIterableToArray(r) || _nonIterableSpread(); }
function _nonIterableSpread() { throw new TypeError("Invalid attempt to spread non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }
function _iterableToArray(r) { if ("undefined" != typeof Symbol && null != r[Symbol.iterator] || null != r["@@iterator"]) return Array.from(r); }
function _arrayWithoutHoles(r) { if (Array.isArray(r)) return _arrayLikeToArray(r); }
function _slicedToArray(r, e) { return _arrayWithHoles(r) || _iterableToArrayLimit(r, e) || _unsupportedIterableToArray(r, e) || _nonIterableRest(); }
function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }
function _unsupportedIterableToArray(r, a) { if (r) { if ("string" == typeof r) return _arrayLikeToArray(r, a); var t = {}.toString.call(r).slice(8, -1); return "Object" === t && r.constructor && (t = r.constructor.name), "Map" === t || "Set" === t ? Array.from(r) : "Arguments" === t || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(t) ? _arrayLikeToArray(r, a) : void 0; } }
function _arrayLikeToArray(r, a) { (null == a || a > r.length) && (a = r.length); for (var e = 0, n = Array(a); e < a; e++) n[e] = r[e]; return n; }
function _iterableToArrayLimit(r, l) { var t = null == r ? null : "undefined" != typeof Symbol && r[Symbol.iterator] || r["@@iterator"]; if (null != t) { var e, n, i, u, a = [], f = !0, o = !1; try { if (i = (t = t.call(r)).next, 0 === l) { if (Object(t) !== t) return; f = !1; } else for (; !(f = (e = i.call(t)).done) && (a.push(e.value), a.length !== l); f = !0); } catch (r) { o = !0, n = r; } finally { try { if (!f && null != t["return"] && (u = t["return"](), Object(u) !== u)) return; } finally { if (o) throw n; } } return a; } }
function _arrayWithHoles(r) { if (Array.isArray(r)) return r; }
function _construct(t, e, r) { if (_isNativeReflectConstruct()) return Reflect.construct.apply(null, arguments); var o = [null]; o.push.apply(o, e); var p = new (t.bind.apply(t, o))(); return r && _setPrototypeOf(p, r.prototype), p; }
function _setPrototypeOf(t, e) { return _setPrototypeOf = Object.setPrototypeOf ? Object.setPrototypeOf.bind() : function (t, e) { return t.__proto__ = e, t; }, _setPrototypeOf(t, e); }
function _isNativeReflectConstruct() { try { var t = !Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); } catch (t) {} return (_isNativeReflectConstruct = function _isNativeReflectConstruct() { return !!t; })(); }
function _defineProperties(e, r) { for (var t = 0; t < r.length; t++) { var o = r[t]; o.enumerable = o.enumerable || !1, o.configurable = !0, "value" in o && (o.writable = !0), Object.defineProperty(e, _toPropertyKey(o.key), o); } }
function _createClass(e, r, t) { return r && _defineProperties(e.prototype, r), t && _defineProperties(e, t), Object.defineProperty(e, "prototype", { writable: !1 }), e; }
function _classCallCheck(a, n) { if (!(a instanceof n)) throw new TypeError("Cannot call a class as a function"); }
function _defineProperty(e, r, t) { return (r = _toPropertyKey(r)) in e ? Object.defineProperty(e, r, { value: t, enumerable: !0, configurable: !0, writable: !0 }) : e[r] = t, e; }
function _toPropertyKey(t) { var i = _toPrimitive(t, "string"); return "symbol" == _typeof(i) ? i : i + ""; }
function _toPrimitive(t, r) { if ("object" != _typeof(t) || !t) return t; var e = t[Symbol.toPrimitive]; if (void 0 !== e) { var i = e.call(t, r || "default"); if ("object" != _typeof(i)) return i; throw new TypeError("@@toPrimitive must return a primitive value."); } return ("string" === r ? String : Number)(t); }
var And = /*#__PURE__*/_createClass(function And() {
  _classCallCheck(this, And);
  _defineProperty(this, "kind", void 0);
  _defineProperty(this, "args", void 0);
  for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
    args[_key] = arguments[_key];
  }
  if (args.length === 0) {
    throw new Error('`And` expects at least one operand');
  }
  this.args = args;
  this.kind = 'and';
});
var Or = /*#__PURE__*/_createClass(function Or() {
  _classCallCheck(this, Or);
  _defineProperty(this, "kind", void 0);
  _defineProperty(this, "args", void 0);
  for (var _len2 = arguments.length, args = new Array(_len2), _key2 = 0; _key2 < _len2; _key2++) {
    args[_key2] = arguments[_key2];
  }
  if (args.length === 0) {
    throw new Error('`Or` expects at least one operand');
  }
  this.args = args;
  this.kind = 'or';
});
var Not = /*#__PURE__*/_createClass(function Not(expr) {
  _classCallCheck(this, Not);
  _defineProperty(this, "kind", void 0);
  _defineProperty(this, "expr", void 0);
  this.expr = expr;
  this.kind = 'not';
});
var Id = /*#__PURE__*/_createClass(function Id(pattern, params) {
  _classCallCheck(this, Id);
  _defineProperty(this, "kind", void 0);
  _defineProperty(this, "pattern", void 0);
  _defineProperty(this, "params", void 0);
  this.pattern = pattern;
  this.kind = 'id';
  this.params = params !== null && params !== void 0 ? params : {
    cleanUrl: false
  };
});
var ImporterId = /*#__PURE__*/_createClass(function ImporterId(pattern, params) {
  _classCallCheck(this, ImporterId);
  _defineProperty(this, "kind", void 0);
  _defineProperty(this, "pattern", void 0);
  _defineProperty(this, "params", void 0);
  this.pattern = pattern;
  this.kind = 'importerId';
  this.params = params !== null && params !== void 0 ? params : {
    cleanUrl: false
  };
});
var ModuleType = /*#__PURE__*/_createClass(function ModuleType(pattern) {
  _classCallCheck(this, ModuleType);
  _defineProperty(this, "kind", void 0);
  _defineProperty(this, "pattern", void 0);
  this.pattern = pattern;
  this.kind = 'moduleType';
});
var Code = /*#__PURE__*/_createClass(function Code(expr) {
  _classCallCheck(this, Code);
  _defineProperty(this, "kind", void 0);
  _defineProperty(this, "pattern", void 0);
  this.pattern = expr;
  this.kind = 'code';
});
var Query = /*#__PURE__*/_createClass(function Query(key, pattern) {
  _classCallCheck(this, Query);
  _defineProperty(this, "kind", void 0);
  _defineProperty(this, "key", void 0);
  _defineProperty(this, "pattern", void 0);
  this.pattern = pattern;
  this.key = key;
  this.kind = 'query';
});
var Include = /*#__PURE__*/_createClass(function Include(expr) {
  _classCallCheck(this, Include);
  _defineProperty(this, "kind", void 0);
  _defineProperty(this, "expr", void 0);
  this.expr = expr;
  this.kind = 'include';
});
var Exclude = /*#__PURE__*/_createClass(function Exclude(expr) {
  _classCallCheck(this, Exclude);
  _defineProperty(this, "kind", void 0);
  _defineProperty(this, "expr", void 0);
  this.expr = expr;
  this.kind = 'exclude';
});
function and() {
  for (var _len3 = arguments.length, args = new Array(_len3), _key3 = 0; _key3 < _len3; _key3++) {
    args[_key3] = arguments[_key3];
  }
  return _construct(And, args);
}
function or() {
  for (var _len4 = arguments.length, args = new Array(_len4), _key4 = 0; _key4 < _len4; _key4++) {
    args[_key4] = arguments[_key4];
  }
  return _construct(Or, args);
}
function not(expr) {
  return new Not(expr);
}
function id(pattern, params) {
  return new Id(pattern, params);
}
function importerId(pattern, params) {
  return new ImporterId(pattern, params);
}
function moduleType(pattern) {
  return new ModuleType(pattern);
}
function code(pattern) {
  return new Code(pattern);
}
/*
 * There are three kinds of conditions are supported:
 * 1. `boolean`: if the value is `true`, the key must exist and be truthy. if the value is `false`, the key must not exist or be falsy.
 * 2. `string`: the key must exist and be equal to the value.
 * 3. `RegExp`: the key must exist and match the value.
 */
function query(key, pattern) {
  return new Query(key, pattern);
}
function include(expr) {
  return new Include(expr);
}
function exclude(expr) {
  return new Exclude(expr);
}
/**
 * convert a queryObject to FilterExpression like
 * ```js
 *   and(query(k1, v1), query(k2, v2))
 * ```
 * @param queryFilterObject The query filter object needs to be matched.
 * @returns a `And` FilterExpression
 */
function queries(queryFilter) {
  var arr = Object.entries(queryFilter).map(function (_ref) {
    var _ref2 = _slicedToArray(_ref, 2),
      key = _ref2[0],
      value = _ref2[1];
    return new Query(key, value);
  });
  return and.apply(void 0, _toConsumableArray(arr));
}
function interpreter(exprs, code, id, moduleType, importerId) {
  var arr = [];
  if (Array.isArray(exprs)) {
    arr = exprs;
  } else {
    arr = [exprs];
  }
  return interpreterImpl(arr, code, id, moduleType, importerId);
}
function interpreterImpl(expr, code, id, moduleType, importerId) {
  var ctx = arguments.length > 5 && arguments[5] !== undefined ? arguments[5] : {};
  var hasInclude = false;
  var _iterator = _createForOfIteratorHelper(expr),
    _step;
  try {
    for (_iterator.s(); !(_step = _iterator.n()).done;) {
      var e = _step.value;
      switch (e.kind) {
        case 'include':
          {
            hasInclude = true;
            if (exprInterpreter(e.expr, code, id, moduleType, importerId, ctx)) {
              return true;
            }
            break;
          }
        case 'exclude':
          {
            if (exprInterpreter(e.expr, code, id, moduleType, importerId, ctx)) {
              return false;
            }
            break;
          }
      }
    }
  } catch (err) {
    _iterator.e(err);
  } finally {
    _iterator.f();
  }
  return !hasInclude;
}
function exprInterpreter(expr, code, id, moduleType, importerId) {
  var ctx = arguments.length > 5 && arguments[5] !== undefined ? arguments[5] : {};
  switch (expr.kind) {
    case 'and':
      {
        return expr.args.every(function (e) {
          return exprInterpreter(e, code, id, moduleType, importerId, ctx);
        });
      }
    case 'or':
      {
        return expr.args.some(function (e) {
          return exprInterpreter(e, code, id, moduleType, importerId, ctx);
        });
      }
    case 'not':
      {
        return !exprInterpreter(expr.expr, code, id, moduleType, importerId, ctx);
      }
    case 'id':
      {
        if (id === undefined) {
          throw new Error('`id` is required for `id` expression');
        }
        var idToMatch = id;
        if (expr.params.cleanUrl) {
          idToMatch = (0, _utils.cleanUrl)(idToMatch);
        }
        return typeof expr.pattern === 'string' ? idToMatch === expr.pattern : expr.pattern.test(idToMatch);
      }
    case 'importerId':
      {
        if (importerId === undefined) {
          return false; // Entry files have no importer, so no match
        }
        var importerIdToMatch = importerId;
        if (expr.params.cleanUrl) {
          importerIdToMatch = (0, _utils.cleanUrl)(importerIdToMatch);
        }
        return typeof expr.pattern === 'string' ? importerIdToMatch === expr.pattern : expr.pattern.test(importerIdToMatch);
      }
    case 'moduleType':
      {
        if (moduleType === undefined) {
          throw new Error('`moduleType` is required for `moduleType` expression');
        }
        return moduleType === expr.pattern;
      }
    case 'code':
      {
        if (code === undefined) {
          throw new Error('`code` is required for `code` expression');
        }
        return typeof expr.pattern === 'string' ? code.includes(expr.pattern) : expr.pattern.test(code);
      }
    case 'query':
      {
        if (id === undefined) {
          throw new Error('`id` is required for `Query` expression');
        }
        if (!ctx.urlSearchParamsCache) {
          var queryString = (0, _utils.extractQueryWithoutFragment)(id);
          ctx.urlSearchParamsCache = new URLSearchParams(queryString);
        }
        var urlParams = ctx.urlSearchParamsCache;
        if (typeof expr.pattern === 'boolean') {
          if (expr.pattern) {
            return urlParams.has(expr.key);
          } else {
            return !urlParams.has(expr.key);
          }
        } else if (typeof expr.pattern === 'string') {
          return urlParams.get(expr.key) === expr.pattern;
        } else {
          var _urlParams$get;
          return expr.pattern.test((_urlParams$get = urlParams.get(expr.key)) !== null && _urlParams$get !== void 0 ? _urlParams$get : '');
        }
      }
    default:
      {
        throw new Error("Expression ".concat(JSON.stringify(expr), " is not expected."));
      }
  }
}
