// @jridgewell/trace-mapping/dist/trace-mapping.umd.js
"use strict";

function _defineProperties(e, r) { for (var t = 0; t < r.length; t++) { var o = r[t]; o.enumerable = o.enumerable || !1, o.configurable = !0, "value" in o && (o.writable = !0), Object.defineProperty(e, _toPropertyKey(o.key), o); } }
function _createClass(e, r, t) { return r && _defineProperties(e.prototype, r), t && _defineProperties(e, t), Object.defineProperty(e, "prototype", { writable: !1 }), e; }
function _toPropertyKey(t) { var i = _toPrimitive(t, "string"); return "symbol" == _typeof(i) ? i : i + ""; }
function _toPrimitive(t, r) { if ("object" != _typeof(t) || !t) return t; var e = t[Symbol.toPrimitive]; if (void 0 !== e) { var i = e.call(t, r || "default"); if ("object" != _typeof(i)) return i; throw new TypeError("@@toPrimitive must return a primitive value."); } return ("string" === r ? String : Number)(t); }
function _classCallCheck(a, n) { if (!(a instanceof n)) throw new TypeError("Cannot call a class as a function"); }
function _createForOfIteratorHelper(r, e) { var t = "undefined" != typeof Symbol && r[Symbol.iterator] || r["@@iterator"]; if (!t) { if (Array.isArray(r) || (t = _unsupportedIterableToArray(r)) || e && r && "number" == typeof r.length) { t && (r = t); var _n = 0, F = function F() {}; return { s: F, n: function n() { return _n >= r.length ? { done: !0 } : { done: !1, value: r[_n++] }; }, e: function e(r) { throw r; }, f: F }; } throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); } var o, a = !0, u = !1; return { s: function s() { t = t.call(r); }, n: function n() { var r = t.next(); return a = r.done, r; }, e: function e(r) { u = !0, o = r; }, f: function f() { try { a || null == t["return"] || t["return"](); } finally { if (u) throw o; } } }; }
function _unsupportedIterableToArray(r, a) { if (r) { if ("string" == typeof r) return _arrayLikeToArray(r, a); var t = {}.toString.call(r).slice(8, -1); return "Object" === t && r.constructor && (t = r.constructor.name), "Map" === t || "Set" === t ? Array.from(r) : "Arguments" === t || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(t) ? _arrayLikeToArray(r, a) : void 0; } }
function _arrayLikeToArray(r, a) { (null == a || a > r.length) && (a = r.length); for (var e = 0, n = Array(a); e < a; e++) n[e] = r[e]; return n; }
function _typeof(o) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (o) { return typeof o; } : function (o) { return o && "function" == typeof Symbol && o.constructor === Symbol && o !== Symbol.prototype ? "symbol" : typeof o; }, _typeof(o); }
(function (global, factory) {
  if ((typeof exports === "undefined" ? "undefined" : _typeof(exports)) === 'object' && typeof module !== 'undefined') {
    factory(module, require('@jridgewell/resolve-uri'), require('@jridgewell/sourcemap-codec'));
    module.exports = def(module);
  } else if (typeof define === 'function' && define.amd) {
    define(['module', '@jridgewell/resolve-uri', '@jridgewell/sourcemap-codec'], function (mod) {
      factory.apply(this, arguments);
      mod.exports = def(mod);
    });
  } else {
    var mod = {
      exports: {}
    };
    factory(mod, global.resolveURI, global.sourcemapCodec);
    global = typeof globalThis !== 'undefined' ? globalThis : global || self;
    global.traceMapping = def(mod);
  }
  function def(m) {
    return 'default' in m.exports ? m.exports["default"] : m.exports;
  }
})(void 0, function (module, require_resolveURI, require_sourcemapCodec) {
  "use strict";

  var __create = Object.create;
  var __defProp = Object.defineProperty;
  var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
  var __getOwnPropNames = Object.getOwnPropertyNames;
  var __getProtoOf = Object.getPrototypeOf;
  var __hasOwnProp = Object.prototype.hasOwnProperty;
  var __commonJS = function __commonJS(cb, mod) {
    return function __require() {
      return mod || (0, cb[__getOwnPropNames(cb)[0]])((mod = {
        exports: {}
      }).exports, mod), mod.exports;
    };
  };
  var __export = function __export(target, all) {
    for (var name in all) __defProp(target, name, {
      get: all[name],
      enumerable: true
    });
  };
  var __copyProps = function __copyProps(to, from, except, desc) {
    if (from && _typeof(from) === "object" || typeof from === "function") {
      var _iterator = _createForOfIteratorHelper(__getOwnPropNames(from)),
        _step;
      try {
        var _loop = function _loop() {
          var key = _step.value;
          if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
            get: function get() {
              return from[key];
            },
            enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
          });
        };
        for (_iterator.s(); !(_step = _iterator.n()).done;) {
          _loop();
        }
      } catch (err) {
        _iterator.e(err);
      } finally {
        _iterator.f();
      }
    }
    return to;
  };
  var __toESM = function __toESM(mod, isNodeMode, target) {
    return target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
    // If the importer is in node compatibility mode or this is not an ESM
    // file that has been converted to a CommonJS file using a Babel-
    // compatible transform (i.e. "__esModule" has not been set), then set
    // "default" to the CommonJS "module.exports" for node compatibility.
    isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", {
      value: mod,
      enumerable: true
    }) : target, mod);
  };
  var __toCommonJS = function __toCommonJS(mod) {
    return __copyProps(__defProp({}, "__esModule", {
      value: true
    }), mod);
  };

  // umd:@jridgewell/sourcemap-codec
  var require_sourcemap_codec = __commonJS({
    "umd:@jridgewell/sourcemap-codec": function umdJridgewell_sourcemapCodec(exports, module2) {
      module2.exports = require_sourcemapCodec;
    }
  });

  // umd:@jridgewell/resolve-uri
  var require_resolve_uri = __commonJS({
    "umd:@jridgewell/resolve-uri": function umdJridgewell_resolveUri(exports, module2) {
      module2.exports = require_resolveURI;
    }
  });

  // src/trace-mapping.ts
  var trace_mapping_exports = {};
  __export(trace_mapping_exports, {
    AnyMap: function AnyMap() {
      return _FlattenMap;
    },
    FlattenMap: function FlattenMap() {
      return _FlattenMap;
    },
    GREATEST_LOWER_BOUND: function GREATEST_LOWER_BOUND() {
      return _GREATEST_LOWER_BOUND;
    },
    LEAST_UPPER_BOUND: function LEAST_UPPER_BOUND() {
      return _LEAST_UPPER_BOUND;
    },
    TraceMap: function TraceMap() {
      return _TraceMap;
    },
    allGeneratedPositionsFor: function allGeneratedPositionsFor() {
      return _allGeneratedPositionsFor;
    },
    decodedMap: function decodedMap() {
      return _decodedMap;
    },
    decodedMappings: function decodedMappings() {
      return _decodedMappings;
    },
    eachMapping: function eachMapping() {
      return _eachMapping;
    },
    encodedMap: function encodedMap() {
      return _encodedMap;
    },
    encodedMappings: function encodedMappings() {
      return _encodedMappings;
    },
    generatedPositionFor: function generatedPositionFor() {
      return _generatedPositionFor;
    },
    isIgnored: function isIgnored() {
      return _isIgnored;
    },
    originalPositionFor: function originalPositionFor() {
      return _originalPositionFor;
    },
    presortedDecodedMap: function presortedDecodedMap() {
      return _presortedDecodedMap;
    },
    sourceContentFor: function sourceContentFor() {
      return _sourceContentFor;
    },
    traceSegment: function traceSegment() {
      return _traceSegment;
    }
  });
  module.exports = __toCommonJS(trace_mapping_exports);
  var import_sourcemap_codec = __toESM(require_sourcemap_codec());

  // src/resolve.ts
  var import_resolve_uri = __toESM(require_resolve_uri());

  // src/strip-filename.ts
  function stripFilename(path) {
    if (!path) return "";
    var index = path.lastIndexOf("/");
    return path.slice(0, index + 1);
  }

  // src/resolve.ts
  function resolver(mapUrl, sourceRoot) {
    var from = stripFilename(mapUrl);
    var prefix = sourceRoot ? sourceRoot + "/" : "";
    return function (source) {
      return (0, import_resolve_uri["default"])(prefix + (source || ""), from);
    };
  }

  // src/sourcemap-segment.ts
  var COLUMN = 0;
  var SOURCES_INDEX = 1;
  var SOURCE_LINE = 2;
  var SOURCE_COLUMN = 3;
  var NAMES_INDEX = 4;
  var REV_GENERATED_LINE = 1;
  var REV_GENERATED_COLUMN = 2;

  // src/sort.ts
  function maybeSort(mappings, owned) {
    var unsortedIndex = nextUnsortedSegmentLine(mappings, 0);
    if (unsortedIndex === mappings.length) return mappings;
    if (!owned) mappings = mappings.slice();
    for (var i = unsortedIndex; i < mappings.length; i = nextUnsortedSegmentLine(mappings, i + 1)) {
      mappings[i] = sortSegments(mappings[i], owned);
    }
    return mappings;
  }
  function nextUnsortedSegmentLine(mappings, start) {
    for (var i = start; i < mappings.length; i++) {
      if (!isSorted(mappings[i])) return i;
    }
    return mappings.length;
  }
  function isSorted(line) {
    for (var j = 1; j < line.length; j++) {
      if (line[j][COLUMN] < line[j - 1][COLUMN]) {
        return false;
      }
    }
    return true;
  }
  function sortSegments(line, owned) {
    if (!owned) line = line.slice();
    return line.sort(sortComparator);
  }
  function sortComparator(a, b) {
    return a[COLUMN] - b[COLUMN];
  }

  // src/by-source.ts
  function buildBySources(decoded, memos) {
    var sources = memos.map(function () {
      return [];
    });
    for (var i = 0; i < decoded.length; i++) {
      var line = decoded[i];
      for (var j = 0; j < line.length; j++) {
        var seg = line[j];
        if (seg.length === 1) continue;
        var sourceIndex2 = seg[SOURCES_INDEX];
        var sourceLine = seg[SOURCE_LINE];
        var sourceColumn = seg[SOURCE_COLUMN];
        var source = sources[sourceIndex2];
        var segs = source[sourceLine] || (source[sourceLine] = []);
        segs.push([sourceColumn, i, seg[COLUMN]]);
      }
    }
    for (var _i = 0; _i < sources.length; _i++) {
      var _source = sources[_i];
      for (var _j = 0; _j < _source.length; _j++) {
        var _line = _source[_j];
        if (_line) _line.sort(sortComparator);
      }
    }
    return sources;
  }

  // src/binary-search.ts
  var found = false;
  function binarySearch(haystack, needle, low, high) {
    while (low <= high) {
      var mid = low + (high - low >> 1);
      var cmp = haystack[mid][COLUMN] - needle;
      if (cmp === 0) {
        found = true;
        return mid;
      }
      if (cmp < 0) {
        low = mid + 1;
      } else {
        high = mid - 1;
      }
    }
    found = false;
    return low - 1;
  }
  function upperBound(haystack, needle, index) {
    for (var i = index + 1; i < haystack.length; index = i++) {
      if (haystack[i][COLUMN] !== needle) break;
    }
    return index;
  }
  function lowerBound(haystack, needle, index) {
    for (var i = index - 1; i >= 0; index = i--) {
      if (haystack[i][COLUMN] !== needle) break;
    }
    return index;
  }
  function memoizedState() {
    return {
      lastKey: -1,
      lastNeedle: -1,
      lastIndex: -1
    };
  }
  function memoizedBinarySearch(haystack, needle, state, key) {
    var lastKey = state.lastKey,
      lastNeedle = state.lastNeedle,
      lastIndex = state.lastIndex;
    var low = 0;
    var high = haystack.length - 1;
    if (key === lastKey) {
      if (needle === lastNeedle) {
        found = lastIndex !== -1 && haystack[lastIndex][COLUMN] === needle;
        return lastIndex;
      }
      if (needle >= lastNeedle) {
        low = lastIndex === -1 ? 0 : lastIndex;
      } else {
        high = lastIndex;
      }
    }
    state.lastKey = key;
    state.lastNeedle = needle;
    return state.lastIndex = binarySearch(haystack, needle, low, high);
  }

  // src/types.ts
  function parse(map) {
    return typeof map === "string" ? JSON.parse(map) : map;
  }

  // src/flatten-map.ts
  var _FlattenMap = function _FlattenMap(map, mapUrl) {
    var parsed = parse(map);
    if (!("sections" in parsed)) {
      return new _TraceMap(parsed, mapUrl);
    }
    var mappings = [];
    var sources = [];
    var sourcesContent = [];
    var names = [];
    var ignoreList = [];
    recurse(parsed, mapUrl, mappings, sources, sourcesContent, names, ignoreList, 0, 0, Infinity, Infinity);
    var joined = {
      version: 3,
      file: parsed.file,
      names: names,
      sources: sources,
      sourcesContent: sourcesContent,
      mappings: mappings,
      ignoreList: ignoreList
    };
    return _presortedDecodedMap(joined);
  };
  function recurse(input, mapUrl, mappings, sources, sourcesContent, names, ignoreList, lineOffset, columnOffset, stopLine, stopColumn) {
    var sections = input.sections;
    for (var i = 0; i < sections.length; i++) {
      var _sections$i = sections[i],
        map = _sections$i.map,
        offset = _sections$i.offset;
      var sl = stopLine;
      var sc = stopColumn;
      if (i + 1 < sections.length) {
        var nextOffset = sections[i + 1].offset;
        sl = Math.min(stopLine, lineOffset + nextOffset.line);
        if (sl === stopLine) {
          sc = Math.min(stopColumn, columnOffset + nextOffset.column);
        } else if (sl < stopLine) {
          sc = columnOffset + nextOffset.column;
        }
      }
      addSection(map, mapUrl, mappings, sources, sourcesContent, names, ignoreList, lineOffset + offset.line, columnOffset + offset.column, sl, sc);
    }
  }
  function addSection(input, mapUrl, mappings, sources, sourcesContent, names, ignoreList, lineOffset, columnOffset, stopLine, stopColumn) {
    var parsed = parse(input);
    if ("sections" in parsed) return recurse.apply(void 0, arguments);
    var map = new _TraceMap(parsed, mapUrl);
    var sourcesOffset = sources.length;
    var namesOffset = names.length;
    var decoded = _decodedMappings(map);
    var resolvedSources = map.resolvedSources,
      contents = map.sourcesContent,
      ignores = map.ignoreList;
    append(sources, resolvedSources);
    append(names, map.names);
    if (contents) append(sourcesContent, contents);else for (var i = 0; i < resolvedSources.length; i++) sourcesContent.push(null);
    if (ignores) for (var _i2 = 0; _i2 < ignores.length; _i2++) ignoreList.push(ignores[_i2] + sourcesOffset);
    for (var _i3 = 0; _i3 < decoded.length; _i3++) {
      var lineI = lineOffset + _i3;
      if (lineI > stopLine) return;
      var out = getLine(mappings, lineI);
      var cOffset = _i3 === 0 ? columnOffset : 0;
      var line = decoded[_i3];
      for (var j = 0; j < line.length; j++) {
        var seg = line[j];
        var column = cOffset + seg[COLUMN];
        if (lineI === stopLine && column >= stopColumn) return;
        if (seg.length === 1) {
          out.push([column]);
          continue;
        }
        var sourcesIndex = sourcesOffset + seg[SOURCES_INDEX];
        var sourceLine = seg[SOURCE_LINE];
        var sourceColumn = seg[SOURCE_COLUMN];
        out.push(seg.length === 4 ? [column, sourcesIndex, sourceLine, sourceColumn] : [column, sourcesIndex, sourceLine, sourceColumn, namesOffset + seg[NAMES_INDEX]]);
      }
    }
  }
  function append(arr, other) {
    for (var i = 0; i < other.length; i++) arr.push(other[i]);
  }
  function getLine(arr, index) {
    for (var i = arr.length; i <= index; i++) arr[i] = [];
    return arr[index];
  }

  // src/trace-mapping.ts
  var LINE_GTR_ZERO = "`line` must be greater than 0 (lines start at line 1)";
  var COL_GTR_EQ_ZERO = "`column` must be greater than or equal to 0 (columns start at column 0)";
  var _LEAST_UPPER_BOUND = -1;
  var _GREATEST_LOWER_BOUND = 1;
  var _TraceMap = /*#__PURE__*/_createClass(function _TraceMap(map, mapUrl) {
    _classCallCheck(this, _TraceMap);
    var isString = typeof map === "string";
    if (!isString && map._decodedMemo) return map;
    var parsed = parse(map);
    var version = parsed.version,
      file = parsed.file,
      names = parsed.names,
      sourceRoot = parsed.sourceRoot,
      sources = parsed.sources,
      sourcesContent = parsed.sourcesContent;
    this.version = version;
    this.file = file;
    this.names = names || [];
    this.sourceRoot = sourceRoot;
    this.sources = sources;
    this.sourcesContent = sourcesContent;
    this.ignoreList = parsed.ignoreList || parsed.x_google_ignoreList || void 0;
    var resolve = resolver(mapUrl, sourceRoot);
    this.resolvedSources = sources.map(resolve);
    var mappings = parsed.mappings;
    if (typeof mappings === "string") {
      this._encoded = mappings;
      this._decoded = void 0;
    } else if (Array.isArray(mappings)) {
      this._encoded = void 0;
      this._decoded = maybeSort(mappings, isString);
    } else if (parsed.sections) {
      throw new Error("TraceMap passed sectioned source map, please use FlattenMap export instead");
    } else {
      throw new Error("invalid source map: ".concat(JSON.stringify(parsed)));
    }
    this._decodedMemo = memoizedState();
    this._bySources = void 0;
    this._bySourceMemos = void 0;
  });
  function cast(map) {
    return map;
  }
  function _encodedMappings(map) {
    var _a, _b;
    return (_b = (_a = cast(map))._encoded) != null ? _b : _a._encoded = (0, import_sourcemap_codec.encode)(cast(map)._decoded);
  }
  function _decodedMappings(map) {
    var _a;
    return (_a = cast(map))._decoded || (_a._decoded = (0, import_sourcemap_codec.decode)(cast(map)._encoded));
  }
  function _traceSegment(map, line, column) {
    var decoded = _decodedMappings(map);
    if (line >= decoded.length) return null;
    var segments = decoded[line];
    var index = traceSegmentInternal(segments, cast(map)._decodedMemo, line, column, _GREATEST_LOWER_BOUND);
    return index === -1 ? null : segments[index];
  }
  function _originalPositionFor(map, needle) {
    var line = needle.line,
      column = needle.column,
      bias = needle.bias;
    line--;
    if (line < 0) throw new Error(LINE_GTR_ZERO);
    if (column < 0) throw new Error(COL_GTR_EQ_ZERO);
    var decoded = _decodedMappings(map);
    if (line >= decoded.length) return OMapping(null, null, null, null);
    var segments = decoded[line];
    var index = traceSegmentInternal(segments, cast(map)._decodedMemo, line, column, bias || _GREATEST_LOWER_BOUND);
    if (index === -1) return OMapping(null, null, null, null);
    var segment = segments[index];
    if (segment.length === 1) return OMapping(null, null, null, null);
    var names = map.names,
      resolvedSources = map.resolvedSources;
    return OMapping(resolvedSources[segment[SOURCES_INDEX]], segment[SOURCE_LINE] + 1, segment[SOURCE_COLUMN], segment.length === 5 ? names[segment[NAMES_INDEX]] : null);
  }
  function _generatedPositionFor(map, needle) {
    var source = needle.source,
      line = needle.line,
      column = needle.column,
      bias = needle.bias;
    return generatedPosition(map, source, line, column, bias || _GREATEST_LOWER_BOUND, false);
  }
  function _allGeneratedPositionsFor(map, needle) {
    var source = needle.source,
      line = needle.line,
      column = needle.column,
      bias = needle.bias;
    return generatedPosition(map, source, line, column, bias || _LEAST_UPPER_BOUND, true);
  }
  function _eachMapping(map, cb) {
    var decoded = _decodedMappings(map);
    var names = map.names,
      resolvedSources = map.resolvedSources;
    for (var i = 0; i < decoded.length; i++) {
      var line = decoded[i];
      for (var j = 0; j < line.length; j++) {
        var seg = line[j];
        var generatedLine = i + 1;
        var generatedColumn = seg[0];
        var source = null;
        var originalLine = null;
        var originalColumn = null;
        var name = null;
        if (seg.length !== 1) {
          source = resolvedSources[seg[1]];
          originalLine = seg[2] + 1;
          originalColumn = seg[3];
        }
        if (seg.length === 5) name = names[seg[4]];
        cb({
          generatedLine: generatedLine,
          generatedColumn: generatedColumn,
          source: source,
          originalLine: originalLine,
          originalColumn: originalColumn,
          name: name
        });
      }
    }
  }
  function sourceIndex(map, source) {
    var sources = map.sources,
      resolvedSources = map.resolvedSources;
    var index = sources.indexOf(source);
    if (index === -1) index = resolvedSources.indexOf(source);
    return index;
  }
  function _sourceContentFor(map, source) {
    var sourcesContent = map.sourcesContent;
    if (sourcesContent == null) return null;
    var index = sourceIndex(map, source);
    return index === -1 ? null : sourcesContent[index];
  }
  function _isIgnored(map, source) {
    var ignoreList = map.ignoreList;
    if (ignoreList == null) return false;
    var index = sourceIndex(map, source);
    return index === -1 ? false : ignoreList.includes(index);
  }
  function _presortedDecodedMap(map, mapUrl) {
    var tracer = new _TraceMap(clone(map, []), mapUrl);
    cast(tracer)._decoded = map.mappings;
    return tracer;
  }
  function _decodedMap(map) {
    return clone(map, _decodedMappings(map));
  }
  function _encodedMap(map) {
    return clone(map, _encodedMappings(map));
  }
  function clone(map, mappings) {
    return {
      version: map.version,
      file: map.file,
      names: map.names,
      sourceRoot: map.sourceRoot,
      sources: map.sources,
      sourcesContent: map.sourcesContent,
      mappings: mappings,
      ignoreList: map.ignoreList || map.x_google_ignoreList
    };
  }
  function OMapping(source, line, column, name) {
    return {
      source: source,
      line: line,
      column: column,
      name: name
    };
  }
  function GMapping(line, column) {
    return {
      line: line,
      column: column
    };
  }
  function traceSegmentInternal(segments, memo, line, column, bias) {
    var index = memoizedBinarySearch(segments, column, memo, line);
    if (found) {
      index = (bias === _LEAST_UPPER_BOUND ? upperBound : lowerBound)(segments, column, index);
    } else if (bias === _LEAST_UPPER_BOUND) index++;
    if (index === -1 || index === segments.length) return -1;
    return index;
  }
  function sliceGeneratedPositions(segments, memo, line, column, bias) {
    var min = traceSegmentInternal(segments, memo, line, column, _GREATEST_LOWER_BOUND);
    if (!found && bias === _LEAST_UPPER_BOUND) min++;
    if (min === -1 || min === segments.length) return [];
    var matchedColumn = found ? column : segments[min][COLUMN];
    if (!found) min = lowerBound(segments, matchedColumn, min);
    var max = upperBound(segments, matchedColumn, min);
    var result = [];
    for (; min <= max; min++) {
      var segment = segments[min];
      result.push(GMapping(segment[REV_GENERATED_LINE] + 1, segment[REV_GENERATED_COLUMN]));
    }
    return result;
  }
  function generatedPosition(map, source, line, column, bias, all) {
    var _a, _b;
    line--;
    if (line < 0) throw new Error(LINE_GTR_ZERO);
    if (column < 0) throw new Error(COL_GTR_EQ_ZERO);
    var sources = map.sources,
      resolvedSources = map.resolvedSources;
    var sourceIndex2 = sources.indexOf(source);
    if (sourceIndex2 === -1) sourceIndex2 = resolvedSources.indexOf(source);
    if (sourceIndex2 === -1) return all ? [] : GMapping(null, null);
    var bySourceMemos = (_a = cast(map))._bySourceMemos || (_a._bySourceMemos = sources.map(memoizedState));
    var generated = (_b = cast(map))._bySources || (_b._bySources = buildBySources(_decodedMappings(map), bySourceMemos));
    var segments = generated[sourceIndex2][line];
    if (segments == null) return all ? [] : GMapping(null, null);
    var memo = bySourceMemos[sourceIndex2];
    if (all) return sliceGeneratedPositions(segments, memo, line, column, bias);
    var index = traceSegmentInternal(segments, memo, line, column, bias);
    if (index === -1) return GMapping(null, null);
    var segment = segments[index];
    return GMapping(segment[REV_GENERATED_LINE] + 1, segment[REV_GENERATED_COLUMN]);
  }
});
