// @webassemblyjs/ast/esm/nodes.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.assertValtypeLiteral = exports.assertTypeInstruction = exports.assertTable = exports.assertStringLiteral = exports.assertStart = exports.assertSignature = exports.assertSectionMetadata = exports.assertQuoteModule = exports.assertProgram = exports.assertProducersSectionMetadata = exports.assertProducerMetadataVersionedName = exports.assertProducerMetadata = exports.assertNumberLiteral = exports.assertModuleNameMetadata = exports.assertModuleMetadata = exports.assertModuleImport = exports.assertModuleExportDescr = exports.assertModuleExport = exports.assertModule = exports.assertMemory = exports.assertLoopInstruction = exports.assertLongNumberLiteral = exports.assertLocalNameMetadata = exports.assertLimit = exports.assertLeadingComment = exports.assertInternalGoto = exports.assertInternalEndAndReturn = exports.assertInternalCallExtern = exports.assertInternalBrUnless = exports.assertInstr = exports.assertIndexInFuncSection = exports.assertIfInstruction = exports.assertIdentifier = exports.assertGlobalType = exports.assertGlobal = exports.assertFunctionNameMetadata = exports.assertFuncImportDescr = exports.assertFunc = exports.assertFloatLiteral = exports.assertElem = exports.assertData = exports.assertCallInstruction = exports.assertCallIndirectInstruction = exports.assertByteArray = exports.assertBlockInstruction = exports.assertBlockComment = exports.assertBinaryModule = void 0;
exports.binaryModule = binaryModule;
exports.blockComment = blockComment;
exports.blockInstruction = blockInstruction;
exports.byteArray = byteArray;
exports.callIndirectInstruction = callIndirectInstruction;
exports.callInstruction = callInstruction;
exports.data = data;
exports.elem = elem;
exports.floatLiteral = floatLiteral;
exports.func = func;
exports.funcImportDescr = funcImportDescr;
exports.functionNameMetadata = functionNameMetadata;
exports.global = global;
exports.globalType = globalType;
exports.identifier = identifier;
exports.ifInstruction = ifInstruction;
exports.indexInFuncSection = indexInFuncSection;
exports.instr = instr;
exports.internalBrUnless = internalBrUnless;
exports.internalCallExtern = internalCallExtern;
exports.internalEndAndReturn = internalEndAndReturn;
exports.internalGoto = internalGoto;
exports.isLongNumberLiteral = exports.isLocalNameMetadata = exports.isLimit = exports.isLeadingComment = exports.isIntrinsic = exports.isInternalGoto = exports.isInternalEndAndReturn = exports.isInternalCallExtern = exports.isInternalBrUnless = exports.isInstruction = exports.isInstr = exports.isIndexInFuncSection = exports.isImportDescr = exports.isIfInstruction = exports.isIdentifier = exports.isGlobalType = exports.isGlobal = exports.isFunctionNameMetadata = exports.isFuncImportDescr = exports.isFunc = exports.isFloatLiteral = exports.isExpression = exports.isElem = exports.isData = exports.isCallInstruction = exports.isCallIndirectInstruction = exports.isByteArray = exports.isBlockInstruction = exports.isBlockComment = exports.isBlock = exports.isBinaryModule = void 0;
exports.isValtypeLiteral = exports.isTypeInstruction = exports.isTable = exports.isStringLiteral = exports.isStart = exports.isSignature = exports.isSectionMetadata = exports.isQuoteModule = exports.isProgram = exports.isProducersSectionMetadata = exports.isProducerMetadataVersionedName = exports.isProducerMetadata = exports.isNumericLiteral = exports.isNumberLiteral = exports.isNode = exports.isModuleNameMetadata = exports.isModuleMetadata = exports.isModuleImport = exports.isModuleExportDescr = exports.isModuleExport = exports.isModule = exports.isMemory = exports.isLoopInstruction = void 0;
exports.leadingComment = leadingComment;
exports.limit = limit;
exports.localNameMetadata = localNameMetadata;
exports.longNumberLiteral = longNumberLiteral;
exports.loopInstruction = loopInstruction;
exports.memory = memory;
exports.module = _module;
exports.moduleExport = moduleExport;
exports.moduleExportDescr = moduleExportDescr;
exports.moduleImport = moduleImport;
exports.moduleMetadata = moduleMetadata;
exports.moduleNameMetadata = moduleNameMetadata;
exports.nodeAndUnionTypes = void 0;
exports.numberLiteral = numberLiteral;
exports.producerMetadata = producerMetadata;
exports.producerMetadataVersionedName = producerMetadataVersionedName;
exports.producersSectionMetadata = producersSectionMetadata;
exports.program = program;
exports.quoteModule = quoteModule;
exports.sectionMetadata = sectionMetadata;
exports.signature = signature;
exports.start = start;
exports.stringLiteral = stringLiteral;
exports.table = table;
exports.typeInstruction = typeInstruction;
exports.unionTypesMap = void 0;
exports.valtypeLiteral = valtypeLiteral;
function _typeof(obj) {
  "@babel/helpers - typeof";

  if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") {
    _typeof = function _typeof(obj) {
      return typeof obj;
    };
  } else {
    _typeof = function _typeof(obj) {
      return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj;
    };
  }
  return _typeof(obj);
}

// THIS FILE IS AUTOGENERATED
// see scripts/generateNodeUtils.js
function isTypeOf(t) {
  return function (n) {
    return n.type === t;
  };
}
function assertTypeOf(t) {
  return function (n) {
    return function () {
      if (!(n.type === t)) {
        throw new Error('n.type === t' + " error: " + (undefined || "unknown"));
      }
    }();
  };
}
function _module(id, fields, metadata) {
  if (id !== null && id !== undefined) {
    if (!(typeof id === "string")) {
      throw new Error('typeof id === "string"' + " error: " + ("Argument id must be of type string, given: " + _typeof(id) || "unknown"));
    }
  }
  if (!(_typeof(fields) === "object" && typeof fields.length !== "undefined")) {
    throw new Error('typeof fields === "object" && typeof fields.length !== "undefined"' + " error: " + (undefined || "unknown"));
  }
  var node = {
    type: "Module",
    id: id,
    fields: fields
  };
  if (typeof metadata !== "undefined") {
    node.metadata = metadata;
  }
  return node;
}
function moduleMetadata(sections, functionNames, localNames, producers) {
  if (!(_typeof(sections) === "object" && typeof sections.length !== "undefined")) {
    throw new Error('typeof sections === "object" && typeof sections.length !== "undefined"' + " error: " + (undefined || "unknown"));
  }
  if (functionNames !== null && functionNames !== undefined) {
    if (!(_typeof(functionNames) === "object" && typeof functionNames.length !== "undefined")) {
      throw new Error('typeof functionNames === "object" && typeof functionNames.length !== "undefined"' + " error: " + (undefined || "unknown"));
    }
  }
  if (localNames !== null && localNames !== undefined) {
    if (!(_typeof(localNames) === "object" && typeof localNames.length !== "undefined")) {
      throw new Error('typeof localNames === "object" && typeof localNames.length !== "undefined"' + " error: " + (undefined || "unknown"));
    }
  }
  if (producers !== null && producers !== undefined) {
    if (!(_typeof(producers) === "object" && typeof producers.length !== "undefined")) {
      throw new Error('typeof producers === "object" && typeof producers.length !== "undefined"' + " error: " + (undefined || "unknown"));
    }
  }
  var node = {
    type: "ModuleMetadata",
    sections: sections
  };
  if (typeof functionNames !== "undefined" && functionNames.length > 0) {
    node.functionNames = functionNames;
  }
  if (typeof localNames !== "undefined" && localNames.length > 0) {
    node.localNames = localNames;
  }
  if (typeof producers !== "undefined" && producers.length > 0) {
    node.producers = producers;
  }
  return node;
}
function moduleNameMetadata(value) {
  if (!(typeof value === "string")) {
    throw new Error('typeof value === "string"' + " error: " + ("Argument value must be of type string, given: " + _typeof(value) || "unknown"));
  }
  var node = {
    type: "ModuleNameMetadata",
    value: value
  };
  return node;
}
function functionNameMetadata(value, index) {
  if (!(typeof value === "string")) {
    throw new Error('typeof value === "string"' + " error: " + ("Argument value must be of type string, given: " + _typeof(value) || "unknown"));
  }
  if (!(typeof index === "number")) {
    throw new Error('typeof index === "number"' + " error: " + ("Argument index must be of type number, given: " + _typeof(index) || "unknown"));
  }
  var node = {
    type: "FunctionNameMetadata",
    value: value,
    index: index
  };
  return node;
}
function localNameMetadata(value, localIndex, functionIndex) {
  if (!(typeof value === "string")) {
    throw new Error('typeof value === "string"' + " error: " + ("Argument value must be of type string, given: " + _typeof(value) || "unknown"));
  }
  if (!(typeof localIndex === "number")) {
    throw new Error('typeof localIndex === "number"' + " error: " + ("Argument localIndex must be of type number, given: " + _typeof(localIndex) || "unknown"));
  }
  if (!(typeof functionIndex === "number")) {
    throw new Error('typeof functionIndex === "number"' + " error: " + ("Argument functionIndex must be of type number, given: " + _typeof(functionIndex) || "unknown"));
  }
  var node = {
    type: "LocalNameMetadata",
    value: value,
    localIndex: localIndex,
    functionIndex: functionIndex
  };
  return node;
}
function binaryModule(id, blob) {
  if (id !== null && id !== undefined) {
    if (!(typeof id === "string")) {
      throw new Error('typeof id === "string"' + " error: " + ("Argument id must be of type string, given: " + _typeof(id) || "unknown"));
    }
  }
  if (!(_typeof(blob) === "object" && typeof blob.length !== "undefined")) {
    throw new Error('typeof blob === "object" && typeof blob.length !== "undefined"' + " error: " + (undefined || "unknown"));
  }
  var node = {
    type: "BinaryModule",
    id: id,
    blob: blob
  };
  return node;
}
function quoteModule(id, string) {
  if (id !== null && id !== undefined) {
    if (!(typeof id === "string")) {
      throw new Error('typeof id === "string"' + " error: " + ("Argument id must be of type string, given: " + _typeof(id) || "unknown"));
    }
  }
  if (!(_typeof(string) === "object" && typeof string.length !== "undefined")) {
    throw new Error('typeof string === "object" && typeof string.length !== "undefined"' + " error: " + (undefined || "unknown"));
  }
  var node = {
    type: "QuoteModule",
    id: id,
    string: string
  };
  return node;
}
function sectionMetadata(section, startOffset, size, vectorOfSize) {
  if (!(typeof startOffset === "number")) {
    throw new Error('typeof startOffset === "number"' + " error: " + ("Argument startOffset must be of type number, given: " + _typeof(startOffset) || "unknown"));
  }
  var node = {
    type: "SectionMetadata",
    section: section,
    startOffset: startOffset,
    size: size,
    vectorOfSize: vectorOfSize
  };
  return node;
}
function producersSectionMetadata(producers) {
  if (!(_typeof(producers) === "object" && typeof producers.length !== "undefined")) {
    throw new Error('typeof producers === "object" && typeof producers.length !== "undefined"' + " error: " + (undefined || "unknown"));
  }
  var node = {
    type: "ProducersSectionMetadata",
    producers: producers
  };
  return node;
}
function producerMetadata(language, processedBy, sdk) {
  if (!(_typeof(language) === "object" && typeof language.length !== "undefined")) {
    throw new Error('typeof language === "object" && typeof language.length !== "undefined"' + " error: " + (undefined || "unknown"));
  }
  if (!(_typeof(processedBy) === "object" && typeof processedBy.length !== "undefined")) {
    throw new Error('typeof processedBy === "object" && typeof processedBy.length !== "undefined"' + " error: " + (undefined || "unknown"));
  }
  if (!(_typeof(sdk) === "object" && typeof sdk.length !== "undefined")) {
    throw new Error('typeof sdk === "object" && typeof sdk.length !== "undefined"' + " error: " + (undefined || "unknown"));
  }
  var node = {
    type: "ProducerMetadata",
    language: language,
    processedBy: processedBy,
    sdk: sdk
  };
  return node;
}
function producerMetadataVersionedName(name, version) {
  if (!(typeof name === "string")) {
    throw new Error('typeof name === "string"' + " error: " + ("Argument name must be of type string, given: " + _typeof(name) || "unknown"));
  }
  if (!(typeof version === "string")) {
    throw new Error('typeof version === "string"' + " error: " + ("Argument version must be of type string, given: " + _typeof(version) || "unknown"));
  }
  var node = {
    type: "ProducerMetadataVersionedName",
    name: name,
    version: version
  };
  return node;
}
function loopInstruction(label, resulttype, instr) {
  if (!(_typeof(instr) === "object" && typeof instr.length !== "undefined")) {
    throw new Error('typeof instr === "object" && typeof instr.length !== "undefined"' + " error: " + (undefined || "unknown"));
  }
  var node = {
    type: "LoopInstruction",
    id: "loop",
    label: label,
    resulttype: resulttype,
    instr: instr
  };
  return node;
}
function instr(id, object, args, namedArgs) {
  if (!(typeof id === "string")) {
    throw new Error('typeof id === "string"' + " error: " + ("Argument id must be of type string, given: " + _typeof(id) || "unknown"));
  }
  if (!(_typeof(args) === "object" && typeof args.length !== "undefined")) {
    throw new Error('typeof args === "object" && typeof args.length !== "undefined"' + " error: " + (undefined || "unknown"));
  }
  var node = {
    type: "Instr",
    id: id,
    args: args
  };
  if (typeof object !== "undefined") {
    node.object = object;
  }
  if (typeof namedArgs !== "undefined" && Object.keys(namedArgs).length !== 0) {
    node.namedArgs = namedArgs;
  }
  return node;
}
function ifInstruction(testLabel, test, result, consequent, alternate) {
  if (!(_typeof(test) === "object" && typeof test.length !== "undefined")) {
    throw new Error('typeof test === "object" && typeof test.length !== "undefined"' + " error: " + (undefined || "unknown"));
  }
  if (!(_typeof(consequent) === "object" && typeof consequent.length !== "undefined")) {
    throw new Error('typeof consequent === "object" && typeof consequent.length !== "undefined"' + " error: " + (undefined || "unknown"));
  }
  if (!(_typeof(alternate) === "object" && typeof alternate.length !== "undefined")) {
    throw new Error('typeof alternate === "object" && typeof alternate.length !== "undefined"' + " error: " + (undefined || "unknown"));
  }
  var node = {
    type: "IfInstruction",
    id: "if",
    testLabel: testLabel,
    test: test,
    result: result,
    consequent: consequent,
    alternate: alternate
  };
  return node;
}
function stringLiteral(value) {
  if (!(typeof value === "string")) {
    throw new Error('typeof value === "string"' + " error: " + ("Argument value must be of type string, given: " + _typeof(value) || "unknown"));
  }
  var node = {
    type: "StringLiteral",
    value: value
  };
  return node;
}
function numberLiteral(value, raw) {
  if (!(typeof value === "number")) {
    throw new Error('typeof value === "number"' + " error: " + ("Argument value must be of type number, given: " + _typeof(value) || "unknown"));
  }
  if (!(typeof raw === "string")) {
    throw new Error('typeof raw === "string"' + " error: " + ("Argument raw must be of type string, given: " + _typeof(raw) || "unknown"));
  }
  var node = {
    type: "NumberLiteral",
    value: value,
    raw: raw
  };
  return node;
}
function longNumberLiteral(value, raw) {
  if (!(typeof raw === "string")) {
    throw new Error('typeof raw === "string"' + " error: " + ("Argument raw must be of type string, given: " + _typeof(raw) || "unknown"));
  }
  var node = {
    type: "LongNumberLiteral",
    value: value,
    raw: raw
  };
  return node;
}
function floatLiteral(value, nan, inf, raw) {
  if (!(typeof value === "number")) {
    throw new Error('typeof value === "number"' + " error: " + ("Argument value must be of type number, given: " + _typeof(value) || "unknown"));
  }
  if (nan !== null && nan !== undefined) {
    if (!(typeof nan === "boolean")) {
      throw new Error('typeof nan === "boolean"' + " error: " + ("Argument nan must be of type boolean, given: " + _typeof(nan) || "unknown"));
    }
  }
  if (inf !== null && inf !== undefined) {
    if (!(typeof inf === "boolean")) {
      throw new Error('typeof inf === "boolean"' + " error: " + ("Argument inf must be of type boolean, given: " + _typeof(inf) || "unknown"));
    }
  }
  if (!(typeof raw === "string")) {
    throw new Error('typeof raw === "string"' + " error: " + ("Argument raw must be of type string, given: " + _typeof(raw) || "unknown"));
  }
  var node = {
    type: "FloatLiteral",
    value: value,
    raw: raw
  };
  if (nan === true) {
    node.nan = true;
  }
  if (inf === true) {
    node.inf = true;
  }
  return node;
}
function elem(table, offset, funcs) {
  if (!(_typeof(offset) === "object" && typeof offset.length !== "undefined")) {
    throw new Error('typeof offset === "object" && typeof offset.length !== "undefined"' + " error: " + (undefined || "unknown"));
  }
  if (!(_typeof(funcs) === "object" && typeof funcs.length !== "undefined")) {
    throw new Error('typeof funcs === "object" && typeof funcs.length !== "undefined"' + " error: " + (undefined || "unknown"));
  }
  var node = {
    type: "Elem",
    table: table,
    offset: offset,
    funcs: funcs
  };
  return node;
}
function indexInFuncSection(index) {
  var node = {
    type: "IndexInFuncSection",
    index: index
  };
  return node;
}
function valtypeLiteral(name) {
  var node = {
    type: "ValtypeLiteral",
    name: name
  };
  return node;
}
function typeInstruction(id, functype) {
  var node = {
    type: "TypeInstruction",
    id: id,
    functype: functype
  };
  return node;
}
function start(index) {
  var node = {
    type: "Start",
    index: index
  };
  return node;
}
function globalType(valtype, mutability) {
  var node = {
    type: "GlobalType",
    valtype: valtype,
    mutability: mutability
  };
  return node;
}
function leadingComment(value) {
  if (!(typeof value === "string")) {
    throw new Error('typeof value === "string"' + " error: " + ("Argument value must be of type string, given: " + _typeof(value) || "unknown"));
  }
  var node = {
    type: "LeadingComment",
    value: value
  };
  return node;
}
function blockComment(value) {
  if (!(typeof value === "string")) {
    throw new Error('typeof value === "string"' + " error: " + ("Argument value must be of type string, given: " + _typeof(value) || "unknown"));
  }
  var node = {
    type: "BlockComment",
    value: value
  };
  return node;
}
function data(memoryIndex, offset, init) {
  var node = {
    type: "Data",
    memoryIndex: memoryIndex,
    offset: offset,
    init: init
  };
  return node;
}
function global(globalType, init, name) {
  if (!(_typeof(init) === "object" && typeof init.length !== "undefined")) {
    throw new Error('typeof init === "object" && typeof init.length !== "undefined"' + " error: " + (undefined || "unknown"));
  }
  var node = {
    type: "Global",
    globalType: globalType,
    init: init,
    name: name
  };
  return node;
}
function table(elementType, limits, name, elements) {
  if (!(limits.type === "Limit")) {
    throw new Error('limits.type === "Limit"' + " error: " + ("Argument limits must be of type Limit, given: " + limits.type || "unknown"));
  }
  if (elements !== null && elements !== undefined) {
    if (!(_typeof(elements) === "object" && typeof elements.length !== "undefined")) {
      throw new Error('typeof elements === "object" && typeof elements.length !== "undefined"' + " error: " + (undefined || "unknown"));
    }
  }
  var node = {
    type: "Table",
    elementType: elementType,
    limits: limits,
    name: name
  };
  if (typeof elements !== "undefined" && elements.length > 0) {
    node.elements = elements;
  }
  return node;
}
function memory(limits, id) {
  var node = {
    type: "Memory",
    limits: limits,
    id: id
  };
  return node;
}
function funcImportDescr(id, signature) {
  var node = {
    type: "FuncImportDescr",
    id: id,
    signature: signature
  };
  return node;
}
function moduleImport(module, name, descr) {
  if (!(typeof module === "string")) {
    throw new Error('typeof module === "string"' + " error: " + ("Argument module must be of type string, given: " + _typeof(module) || "unknown"));
  }
  if (!(typeof name === "string")) {
    throw new Error('typeof name === "string"' + " error: " + ("Argument name must be of type string, given: " + _typeof(name) || "unknown"));
  }
  var node = {
    type: "ModuleImport",
    module: module,
    name: name,
    descr: descr
  };
  return node;
}
function moduleExportDescr(exportType, id) {
  var node = {
    type: "ModuleExportDescr",
    exportType: exportType,
    id: id
  };
  return node;
}
function moduleExport(name, descr) {
  if (!(typeof name === "string")) {
    throw new Error('typeof name === "string"' + " error: " + ("Argument name must be of type string, given: " + _typeof(name) || "unknown"));
  }
  var node = {
    type: "ModuleExport",
    name: name,
    descr: descr
  };
  return node;
}
function limit(min, max, shared) {
  if (!(typeof min === "number")) {
    throw new Error('typeof min === "number"' + " error: " + ("Argument min must be of type number, given: " + _typeof(min) || "unknown"));
  }
  if (max !== null && max !== undefined) {
    if (!(typeof max === "number")) {
      throw new Error('typeof max === "number"' + " error: " + ("Argument max must be of type number, given: " + _typeof(max) || "unknown"));
    }
  }
  if (shared !== null && shared !== undefined) {
    if (!(typeof shared === "boolean")) {
      throw new Error('typeof shared === "boolean"' + " error: " + ("Argument shared must be of type boolean, given: " + _typeof(shared) || "unknown"));
    }
  }
  var node = {
    type: "Limit",
    min: min
  };
  if (typeof max !== "undefined") {
    node.max = max;
  }
  if (shared === true) {
    node.shared = true;
  }
  return node;
}
function signature(params, results) {
  if (!(_typeof(params) === "object" && typeof params.length !== "undefined")) {
    throw new Error('typeof params === "object" && typeof params.length !== "undefined"' + " error: " + (undefined || "unknown"));
  }
  if (!(_typeof(results) === "object" && typeof results.length !== "undefined")) {
    throw new Error('typeof results === "object" && typeof results.length !== "undefined"' + " error: " + (undefined || "unknown"));
  }
  var node = {
    type: "Signature",
    params: params,
    results: results
  };
  return node;
}
function program(body) {
  if (!(_typeof(body) === "object" && typeof body.length !== "undefined")) {
    throw new Error('typeof body === "object" && typeof body.length !== "undefined"' + " error: " + (undefined || "unknown"));
  }
  var node = {
    type: "Program",
    body: body
  };
  return node;
}
function identifier(value, raw) {
  if (!(typeof value === "string")) {
    throw new Error('typeof value === "string"' + " error: " + ("Argument value must be of type string, given: " + _typeof(value) || "unknown"));
  }
  if (raw !== null && raw !== undefined) {
    if (!(typeof raw === "string")) {
      throw new Error('typeof raw === "string"' + " error: " + ("Argument raw must be of type string, given: " + _typeof(raw) || "unknown"));
    }
  }
  var node = {
    type: "Identifier",
    value: value
  };
  if (typeof raw !== "undefined") {
    node.raw = raw;
  }
  return node;
}
function blockInstruction(label, instr, result) {
  if (!(_typeof(instr) === "object" && typeof instr.length !== "undefined")) {
    throw new Error('typeof instr === "object" && typeof instr.length !== "undefined"' + " error: " + (undefined || "unknown"));
  }
  var node = {
    type: "BlockInstruction",
    id: "block",
    label: label,
    instr: instr,
    result: result
  };
  return node;
}
function callInstruction(index, instrArgs, numeric) {
  if (instrArgs !== null && instrArgs !== undefined) {
    if (!(_typeof(instrArgs) === "object" && typeof instrArgs.length !== "undefined")) {
      throw new Error('typeof instrArgs === "object" && typeof instrArgs.length !== "undefined"' + " error: " + (undefined || "unknown"));
    }
  }
  var node = {
    type: "CallInstruction",
    id: "call",
    index: index
  };
  if (typeof instrArgs !== "undefined" && instrArgs.length > 0) {
    node.instrArgs = instrArgs;
  }
  if (typeof numeric !== "undefined") {
    node.numeric = numeric;
  }
  return node;
}
function callIndirectInstruction(signature, intrs) {
  if (intrs !== null && intrs !== undefined) {
    if (!(_typeof(intrs) === "object" && typeof intrs.length !== "undefined")) {
      throw new Error('typeof intrs === "object" && typeof intrs.length !== "undefined"' + " error: " + (undefined || "unknown"));
    }
  }
  var node = {
    type: "CallIndirectInstruction",
    id: "call_indirect",
    signature: signature
  };
  if (typeof intrs !== "undefined" && intrs.length > 0) {
    node.intrs = intrs;
  }
  return node;
}
function byteArray(values) {
  if (!(_typeof(values) === "object" && typeof values.length !== "undefined")) {
    throw new Error('typeof values === "object" && typeof values.length !== "undefined"' + " error: " + (undefined || "unknown"));
  }
  var node = {
    type: "ByteArray",
    values: values
  };
  return node;
}
function func(name, signature, body, isExternal, metadata) {
  if (!(_typeof(body) === "object" && typeof body.length !== "undefined")) {
    throw new Error('typeof body === "object" && typeof body.length !== "undefined"' + " error: " + (undefined || "unknown"));
  }
  if (isExternal !== null && isExternal !== undefined) {
    if (!(typeof isExternal === "boolean")) {
      throw new Error('typeof isExternal === "boolean"' + " error: " + ("Argument isExternal must be of type boolean, given: " + _typeof(isExternal) || "unknown"));
    }
  }
  var node = {
    type: "Func",
    name: name,
    signature: signature,
    body: body
  };
  if (isExternal === true) {
    node.isExternal = true;
  }
  if (typeof metadata !== "undefined") {
    node.metadata = metadata;
  }
  return node;
}
function internalBrUnless(target) {
  if (!(typeof target === "number")) {
    throw new Error('typeof target === "number"' + " error: " + ("Argument target must be of type number, given: " + _typeof(target) || "unknown"));
  }
  var node = {
    type: "InternalBrUnless",
    target: target
  };
  return node;
}
function internalGoto(target) {
  if (!(typeof target === "number")) {
    throw new Error('typeof target === "number"' + " error: " + ("Argument target must be of type number, given: " + _typeof(target) || "unknown"));
  }
  var node = {
    type: "InternalGoto",
    target: target
  };
  return node;
}
function internalCallExtern(target) {
  if (!(typeof target === "number")) {
    throw new Error('typeof target === "number"' + " error: " + ("Argument target must be of type number, given: " + _typeof(target) || "unknown"));
  }
  var node = {
    type: "InternalCallExtern",
    target: target
  };
  return node;
}
function internalEndAndReturn() {
  var node = {
    type: "InternalEndAndReturn"
  };
  return node;
}
var isModule = exports.isModule = isTypeOf("Module");
var isModuleMetadata = exports.isModuleMetadata = isTypeOf("ModuleMetadata");
var isModuleNameMetadata = exports.isModuleNameMetadata = isTypeOf("ModuleNameMetadata");
var isFunctionNameMetadata = exports.isFunctionNameMetadata = isTypeOf("FunctionNameMetadata");
var isLocalNameMetadata = exports.isLocalNameMetadata = isTypeOf("LocalNameMetadata");
var isBinaryModule = exports.isBinaryModule = isTypeOf("BinaryModule");
var isQuoteModule = exports.isQuoteModule = isTypeOf("QuoteModule");
var isSectionMetadata = exports.isSectionMetadata = isTypeOf("SectionMetadata");
var isProducersSectionMetadata = exports.isProducersSectionMetadata = isTypeOf("ProducersSectionMetadata");
var isProducerMetadata = exports.isProducerMetadata = isTypeOf("ProducerMetadata");
var isProducerMetadataVersionedName = exports.isProducerMetadataVersionedName = isTypeOf("ProducerMetadataVersionedName");
var isLoopInstruction = exports.isLoopInstruction = isTypeOf("LoopInstruction");
var isInstr = exports.isInstr = isTypeOf("Instr");
var isIfInstruction = exports.isIfInstruction = isTypeOf("IfInstruction");
var isStringLiteral = exports.isStringLiteral = isTypeOf("StringLiteral");
var isNumberLiteral = exports.isNumberLiteral = isTypeOf("NumberLiteral");
var isLongNumberLiteral = exports.isLongNumberLiteral = isTypeOf("LongNumberLiteral");
var isFloatLiteral = exports.isFloatLiteral = isTypeOf("FloatLiteral");
var isElem = exports.isElem = isTypeOf("Elem");
var isIndexInFuncSection = exports.isIndexInFuncSection = isTypeOf("IndexInFuncSection");
var isValtypeLiteral = exports.isValtypeLiteral = isTypeOf("ValtypeLiteral");
var isTypeInstruction = exports.isTypeInstruction = isTypeOf("TypeInstruction");
var isStart = exports.isStart = isTypeOf("Start");
var isGlobalType = exports.isGlobalType = isTypeOf("GlobalType");
var isLeadingComment = exports.isLeadingComment = isTypeOf("LeadingComment");
var isBlockComment = exports.isBlockComment = isTypeOf("BlockComment");
var isData = exports.isData = isTypeOf("Data");
var isGlobal = exports.isGlobal = isTypeOf("Global");
var isTable = exports.isTable = isTypeOf("Table");
var isMemory = exports.isMemory = isTypeOf("Memory");
var isFuncImportDescr = exports.isFuncImportDescr = isTypeOf("FuncImportDescr");
var isModuleImport = exports.isModuleImport = isTypeOf("ModuleImport");
var isModuleExportDescr = exports.isModuleExportDescr = isTypeOf("ModuleExportDescr");
var isModuleExport = exports.isModuleExport = isTypeOf("ModuleExport");
var isLimit = exports.isLimit = isTypeOf("Limit");
var isSignature = exports.isSignature = isTypeOf("Signature");
var isProgram = exports.isProgram = isTypeOf("Program");
var isIdentifier = exports.isIdentifier = isTypeOf("Identifier");
var isBlockInstruction = exports.isBlockInstruction = isTypeOf("BlockInstruction");
var isCallInstruction = exports.isCallInstruction = isTypeOf("CallInstruction");
var isCallIndirectInstruction = exports.isCallIndirectInstruction = isTypeOf("CallIndirectInstruction");
var isByteArray = exports.isByteArray = isTypeOf("ByteArray");
var isFunc = exports.isFunc = isTypeOf("Func");
var isInternalBrUnless = exports.isInternalBrUnless = isTypeOf("InternalBrUnless");
var isInternalGoto = exports.isInternalGoto = isTypeOf("InternalGoto");
var isInternalCallExtern = exports.isInternalCallExtern = isTypeOf("InternalCallExtern");
var isInternalEndAndReturn = exports.isInternalEndAndReturn = isTypeOf("InternalEndAndReturn");
var isNode = exports.isNode = function isNode(node) {
  return isModule(node) || isModuleMetadata(node) || isModuleNameMetadata(node) || isFunctionNameMetadata(node) || isLocalNameMetadata(node) || isBinaryModule(node) || isQuoteModule(node) || isSectionMetadata(node) || isProducersSectionMetadata(node) || isProducerMetadata(node) || isProducerMetadataVersionedName(node) || isLoopInstruction(node) || isInstr(node) || isIfInstruction(node) || isStringLiteral(node) || isNumberLiteral(node) || isLongNumberLiteral(node) || isFloatLiteral(node) || isElem(node) || isIndexInFuncSection(node) || isValtypeLiteral(node) || isTypeInstruction(node) || isStart(node) || isGlobalType(node) || isLeadingComment(node) || isBlockComment(node) || isData(node) || isGlobal(node) || isTable(node) || isMemory(node) || isFuncImportDescr(node) || isModuleImport(node) || isModuleExportDescr(node) || isModuleExport(node) || isLimit(node) || isSignature(node) || isProgram(node) || isIdentifier(node) || isBlockInstruction(node) || isCallInstruction(node) || isCallIndirectInstruction(node) || isByteArray(node) || isFunc(node) || isInternalBrUnless(node) || isInternalGoto(node) || isInternalCallExtern(node) || isInternalEndAndReturn(node);
};
var isBlock = exports.isBlock = function isBlock(node) {
  return isLoopInstruction(node) || isBlockInstruction(node) || isFunc(node);
};
var isInstruction = exports.isInstruction = function isInstruction(node) {
  return isLoopInstruction(node) || isInstr(node) || isIfInstruction(node) || isTypeInstruction(node) || isBlockInstruction(node) || isCallInstruction(node) || isCallIndirectInstruction(node);
};
var isExpression = exports.isExpression = function isExpression(node) {
  return isInstr(node) || isStringLiteral(node) || isNumberLiteral(node) || isLongNumberLiteral(node) || isFloatLiteral(node) || isValtypeLiteral(node) || isIdentifier(node);
};
var isNumericLiteral = exports.isNumericLiteral = function isNumericLiteral(node) {
  return isNumberLiteral(node) || isLongNumberLiteral(node) || isFloatLiteral(node);
};
var isImportDescr = exports.isImportDescr = function isImportDescr(node) {
  return isGlobalType(node) || isTable(node) || isMemory(node) || isFuncImportDescr(node);
};
var isIntrinsic = exports.isIntrinsic = function isIntrinsic(node) {
  return isInternalBrUnless(node) || isInternalGoto(node) || isInternalCallExtern(node) || isInternalEndAndReturn(node);
};
var assertModule = exports.assertModule = assertTypeOf("Module");
var assertModuleMetadata = exports.assertModuleMetadata = assertTypeOf("ModuleMetadata");
var assertModuleNameMetadata = exports.assertModuleNameMetadata = assertTypeOf("ModuleNameMetadata");
var assertFunctionNameMetadata = exports.assertFunctionNameMetadata = assertTypeOf("FunctionNameMetadata");
var assertLocalNameMetadata = exports.assertLocalNameMetadata = assertTypeOf("LocalNameMetadata");
var assertBinaryModule = exports.assertBinaryModule = assertTypeOf("BinaryModule");
var assertQuoteModule = exports.assertQuoteModule = assertTypeOf("QuoteModule");
var assertSectionMetadata = exports.assertSectionMetadata = assertTypeOf("SectionMetadata");
var assertProducersSectionMetadata = exports.assertProducersSectionMetadata = assertTypeOf("ProducersSectionMetadata");
var assertProducerMetadata = exports.assertProducerMetadata = assertTypeOf("ProducerMetadata");
var assertProducerMetadataVersionedName = exports.assertProducerMetadataVersionedName = assertTypeOf("ProducerMetadataVersionedName");
var assertLoopInstruction = exports.assertLoopInstruction = assertTypeOf("LoopInstruction");
var assertInstr = exports.assertInstr = assertTypeOf("Instr");
var assertIfInstruction = exports.assertIfInstruction = assertTypeOf("IfInstruction");
var assertStringLiteral = exports.assertStringLiteral = assertTypeOf("StringLiteral");
var assertNumberLiteral = exports.assertNumberLiteral = assertTypeOf("NumberLiteral");
var assertLongNumberLiteral = exports.assertLongNumberLiteral = assertTypeOf("LongNumberLiteral");
var assertFloatLiteral = exports.assertFloatLiteral = assertTypeOf("FloatLiteral");
var assertElem = exports.assertElem = assertTypeOf("Elem");
var assertIndexInFuncSection = exports.assertIndexInFuncSection = assertTypeOf("IndexInFuncSection");
var assertValtypeLiteral = exports.assertValtypeLiteral = assertTypeOf("ValtypeLiteral");
var assertTypeInstruction = exports.assertTypeInstruction = assertTypeOf("TypeInstruction");
var assertStart = exports.assertStart = assertTypeOf("Start");
var assertGlobalType = exports.assertGlobalType = assertTypeOf("GlobalType");
var assertLeadingComment = exports.assertLeadingComment = assertTypeOf("LeadingComment");
var assertBlockComment = exports.assertBlockComment = assertTypeOf("BlockComment");
var assertData = exports.assertData = assertTypeOf("Data");
var assertGlobal = exports.assertGlobal = assertTypeOf("Global");
var assertTable = exports.assertTable = assertTypeOf("Table");
var assertMemory = exports.assertMemory = assertTypeOf("Memory");
var assertFuncImportDescr = exports.assertFuncImportDescr = assertTypeOf("FuncImportDescr");
var assertModuleImport = exports.assertModuleImport = assertTypeOf("ModuleImport");
var assertModuleExportDescr = exports.assertModuleExportDescr = assertTypeOf("ModuleExportDescr");
var assertModuleExport = exports.assertModuleExport = assertTypeOf("ModuleExport");
var assertLimit = exports.assertLimit = assertTypeOf("Limit");
var assertSignature = exports.assertSignature = assertTypeOf("Signature");
var assertProgram = exports.assertProgram = assertTypeOf("Program");
var assertIdentifier = exports.assertIdentifier = assertTypeOf("Identifier");
var assertBlockInstruction = exports.assertBlockInstruction = assertTypeOf("BlockInstruction");
var assertCallInstruction = exports.assertCallInstruction = assertTypeOf("CallInstruction");
var assertCallIndirectInstruction = exports.assertCallIndirectInstruction = assertTypeOf("CallIndirectInstruction");
var assertByteArray = exports.assertByteArray = assertTypeOf("ByteArray");
var assertFunc = exports.assertFunc = assertTypeOf("Func");
var assertInternalBrUnless = exports.assertInternalBrUnless = assertTypeOf("InternalBrUnless");
var assertInternalGoto = exports.assertInternalGoto = assertTypeOf("InternalGoto");
var assertInternalCallExtern = exports.assertInternalCallExtern = assertTypeOf("InternalCallExtern");
var assertInternalEndAndReturn = exports.assertInternalEndAndReturn = assertTypeOf("InternalEndAndReturn");
var unionTypesMap = exports.unionTypesMap = {
  Module: ["Node"],
  ModuleMetadata: ["Node"],
  ModuleNameMetadata: ["Node"],
  FunctionNameMetadata: ["Node"],
  LocalNameMetadata: ["Node"],
  BinaryModule: ["Node"],
  QuoteModule: ["Node"],
  SectionMetadata: ["Node"],
  ProducersSectionMetadata: ["Node"],
  ProducerMetadata: ["Node"],
  ProducerMetadataVersionedName: ["Node"],
  LoopInstruction: ["Node", "Block", "Instruction"],
  Instr: ["Node", "Expression", "Instruction"],
  IfInstruction: ["Node", "Instruction"],
  StringLiteral: ["Node", "Expression"],
  NumberLiteral: ["Node", "NumericLiteral", "Expression"],
  LongNumberLiteral: ["Node", "NumericLiteral", "Expression"],
  FloatLiteral: ["Node", "NumericLiteral", "Expression"],
  Elem: ["Node"],
  IndexInFuncSection: ["Node"],
  ValtypeLiteral: ["Node", "Expression"],
  TypeInstruction: ["Node", "Instruction"],
  Start: ["Node"],
  GlobalType: ["Node", "ImportDescr"],
  LeadingComment: ["Node"],
  BlockComment: ["Node"],
  Data: ["Node"],
  Global: ["Node"],
  Table: ["Node", "ImportDescr"],
  Memory: ["Node", "ImportDescr"],
  FuncImportDescr: ["Node", "ImportDescr"],
  ModuleImport: ["Node"],
  ModuleExportDescr: ["Node"],
  ModuleExport: ["Node"],
  Limit: ["Node"],
  Signature: ["Node"],
  Program: ["Node"],
  Identifier: ["Node", "Expression"],
  BlockInstruction: ["Node", "Block", "Instruction"],
  CallInstruction: ["Node", "Instruction"],
  CallIndirectInstruction: ["Node", "Instruction"],
  ByteArray: ["Node"],
  Func: ["Node", "Block"],
  InternalBrUnless: ["Node", "Intrinsic"],
  InternalGoto: ["Node", "Intrinsic"],
  InternalCallExtern: ["Node", "Intrinsic"],
  InternalEndAndReturn: ["Node", "Intrinsic"]
};
var nodeAndUnionTypes = exports.nodeAndUnionTypes = ["Module", "ModuleMetadata", "ModuleNameMetadata", "FunctionNameMetadata", "LocalNameMetadata", "BinaryModule", "QuoteModule", "SectionMetadata", "ProducersSectionMetadata", "ProducerMetadata", "ProducerMetadataVersionedName", "LoopInstruction", "Instr", "IfInstruction", "StringLiteral", "NumberLiteral", "LongNumberLiteral", "FloatLiteral", "Elem", "IndexInFuncSection", "ValtypeLiteral", "TypeInstruction", "Start", "GlobalType", "LeadingComment", "BlockComment", "Data", "Global", "Table", "Memory", "FuncImportDescr", "ModuleImport", "ModuleExportDescr", "ModuleExport", "Limit", "Signature", "Program", "Identifier", "BlockInstruction", "CallInstruction", "CallIndirectInstruction", "ByteArray", "Func", "InternalBrUnless", "InternalGoto", "InternalCallExtern", "InternalEndAndReturn", "Node", "Block", "Instruction", "Expression", "NumericLiteral", "ImportDescr", "Intrinsic"];
