// @jridgewell/gen-mapping/dist/gen-mapping.umd.js
"use strict";

function _defineProperties(e, r) { for (var t = 0; t < r.length; t++) { var o = r[t]; o.enumerable = o.enumerable || !1, o.configurable = !0, "value" in o && (o.writable = !0), Object.defineProperty(e, _toPropertyKey(o.key), o); } }
function _createClass(e, r, t) { return r && _defineProperties(e.prototype, r), t && _defineProperties(e, t), Object.defineProperty(e, "prototype", { writable: !1 }), e; }
function _toPropertyKey(t) { var i = _toPrimitive(t, "string"); return "symbol" == _typeof(i) ? i : i + ""; }
function _toPrimitive(t, r) { if ("object" != _typeof(t) || !t) return t; var e = t[Symbol.toPrimitive]; if (void 0 !== e) { var i = e.call(t, r || "default"); if ("object" != _typeof(i)) return i; throw new TypeError("@@toPrimitive must return a primitive value."); } return ("string" === r ? String : Number)(t); }
function _classCallCheck(a, n) { if (!(a instanceof n)) throw new TypeError("Cannot call a class as a function"); }
function _createForOfIteratorHelper(r, e) { var t = "undefined" != typeof Symbol && r[Symbol.iterator] || r["@@iterator"]; if (!t) { if (Array.isArray(r) || (t = _unsupportedIterableToArray(r)) || e && r && "number" == typeof r.length) { t && (r = t); var _n = 0, F = function F() {}; return { s: F, n: function n() { return _n >= r.length ? { done: !0 } : { done: !1, value: r[_n++] }; }, e: function e(r) { throw r; }, f: F }; } throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); } var o, a = !0, u = !1; return { s: function s() { t = t.call(r); }, n: function n() { var r = t.next(); return a = r.done, r; }, e: function e(r) { u = !0, o = r; }, f: function f() { try { a || null == t["return"] || t["return"](); } finally { if (u) throw o; } } }; }
function _unsupportedIterableToArray(r, a) { if (r) { if ("string" == typeof r) return _arrayLikeToArray(r, a); var t = {}.toString.call(r).slice(8, -1); return "Object" === t && r.constructor && (t = r.constructor.name), "Map" === t || "Set" === t ? Array.from(r) : "Arguments" === t || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(t) ? _arrayLikeToArray(r, a) : void 0; } }
function _arrayLikeToArray(r, a) { (null == a || a > r.length) && (a = r.length); for (var e = 0, n = Array(a); e < a; e++) n[e] = r[e]; return n; }
function _typeof(o) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (o) { return typeof o; } : function (o) { return o && "function" == typeof Symbol && o.constructor === Symbol && o !== Symbol.prototype ? "symbol" : typeof o; }, _typeof(o); }
(function (global, factory) {
  if ((typeof exports === "undefined" ? "undefined" : _typeof(exports)) === 'object' && typeof module !== 'undefined') {
    factory(module, require('@jridgewell/sourcemap-codec'), require('@jridgewell/trace-mapping'));
    module.exports = def(module);
  } else if (typeof define === 'function' && define.amd) {
    define(['module', '@jridgewell/sourcemap-codec', '@jridgewell/trace-mapping'], function (mod) {
      factory.apply(this, arguments);
      mod.exports = def(mod);
    });
  } else {
    var mod = {
      exports: {}
    };
    factory(mod, global.sourcemapCodec, global.traceMapping);
    global = typeof globalThis !== 'undefined' ? globalThis : global || self;
    global.genMapping = def(mod);
  }
  function def(m) {
    return 'default' in m.exports ? m.exports["default"] : m.exports;
  }
})(void 0, function (module, require_sourcemapCodec, require_traceMapping) {
  "use strict";

  var __create = Object.create;
  var __defProp = Object.defineProperty;
  var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
  var __getOwnPropNames = Object.getOwnPropertyNames;
  var __getProtoOf = Object.getPrototypeOf;
  var __hasOwnProp = Object.prototype.hasOwnProperty;
  var __commonJS = function __commonJS(cb, mod) {
    return function __require() {
      return mod || (0, cb[__getOwnPropNames(cb)[0]])((mod = {
        exports: {}
      }).exports, mod), mod.exports;
    };
  };
  var __export = function __export(target, all) {
    for (var name in all) __defProp(target, name, {
      get: all[name],
      enumerable: true
    });
  };
  var __copyProps = function __copyProps(to, from, except, desc) {
    if (from && _typeof(from) === "object" || typeof from === "function") {
      var _iterator = _createForOfIteratorHelper(__getOwnPropNames(from)),
        _step;
      try {
        var _loop = function _loop() {
          var key = _step.value;
          if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
            get: function get() {
              return from[key];
            },
            enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
          });
        };
        for (_iterator.s(); !(_step = _iterator.n()).done;) {
          _loop();
        }
      } catch (err) {
        _iterator.e(err);
      } finally {
        _iterator.f();
      }
    }
    return to;
  };
  var __toESM = function __toESM(mod, isNodeMode, target) {
    return target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
    // If the importer is in node compatibility mode or this is not an ESM
    // file that has been converted to a CommonJS file using a Babel-
    // compatible transform (i.e. "__esModule" has not been set), then set
    // "default" to the CommonJS "module.exports" for node compatibility.
    isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", {
      value: mod,
      enumerable: true
    }) : target, mod);
  };
  var __toCommonJS = function __toCommonJS(mod) {
    return __copyProps(__defProp({}, "__esModule", {
      value: true
    }), mod);
  };

  // umd:@jridgewell/sourcemap-codec
  var require_sourcemap_codec = __commonJS({
    "umd:@jridgewell/sourcemap-codec": function umdJridgewell_sourcemapCodec(exports, module2) {
      module2.exports = require_sourcemapCodec;
    }
  });

  // umd:@jridgewell/trace-mapping
  var require_trace_mapping = __commonJS({
    "umd:@jridgewell/trace-mapping": function umdJridgewell_traceMapping(exports, module2) {
      module2.exports = require_traceMapping;
    }
  });

  // src/gen-mapping.ts
  var gen_mapping_exports = {};
  __export(gen_mapping_exports, {
    GenMapping: function GenMapping() {
      return _GenMapping;
    },
    addMapping: function addMapping() {
      return _addMapping;
    },
    addSegment: function addSegment() {
      return _addSegment;
    },
    allMappings: function allMappings() {
      return _allMappings;
    },
    fromMap: function fromMap() {
      return _fromMap;
    },
    maybeAddMapping: function maybeAddMapping() {
      return _maybeAddMapping;
    },
    maybeAddSegment: function maybeAddSegment() {
      return _maybeAddSegment;
    },
    setIgnore: function setIgnore() {
      return _setIgnore;
    },
    setSourceContent: function setSourceContent() {
      return _setSourceContent;
    },
    toDecodedMap: function toDecodedMap() {
      return _toDecodedMap;
    },
    toEncodedMap: function toEncodedMap() {
      return _toEncodedMap;
    }
  });
  module.exports = __toCommonJS(gen_mapping_exports);

  // src/set-array.ts
  var SetArray = /*#__PURE__*/_createClass(function SetArray() {
    _classCallCheck(this, SetArray);
    this._indexes = {
      __proto__: null
    };
    this.array = [];
  });
  function cast(set) {
    return set;
  }
  function get(setarr, key) {
    return cast(setarr)._indexes[key];
  }
  function put(setarr, key) {
    var index = get(setarr, key);
    if (index !== void 0) return index;
    var _cast = cast(setarr),
      array = _cast.array,
      indexes = _cast._indexes;
    var length = array.push(key);
    return indexes[key] = length - 1;
  }
  function remove(setarr, key) {
    var index = get(setarr, key);
    if (index === void 0) return;
    var _cast2 = cast(setarr),
      array = _cast2.array,
      indexes = _cast2._indexes;
    for (var i = index + 1; i < array.length; i++) {
      var k = array[i];
      array[i - 1] = k;
      indexes[k]--;
    }
    indexes[key] = void 0;
    array.pop();
  }

  // src/gen-mapping.ts
  var import_sourcemap_codec = __toESM(require_sourcemap_codec());
  var import_trace_mapping = __toESM(require_trace_mapping());

  // src/sourcemap-segment.ts
  var COLUMN = 0;
  var SOURCES_INDEX = 1;
  var SOURCE_LINE = 2;
  var SOURCE_COLUMN = 3;
  var NAMES_INDEX = 4;

  // src/gen-mapping.ts
  var NO_NAME = -1;
  var _GenMapping = /*#__PURE__*/_createClass(function _GenMapping() {
    var _ref = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {},
      file = _ref.file,
      sourceRoot = _ref.sourceRoot;
    _classCallCheck(this, _GenMapping);
    this._names = new SetArray();
    this._sources = new SetArray();
    this._sourcesContent = [];
    this._mappings = [];
    this.file = file;
    this.sourceRoot = sourceRoot;
    this._ignoreList = new SetArray();
  });
  function cast2(map) {
    return map;
  }
  function _addSegment(map, genLine, genColumn, source, sourceLine, sourceColumn, name, content) {
    return addSegmentInternal(false, map, genLine, genColumn, source, sourceLine, sourceColumn, name, content);
  }
  function _addMapping(map, mapping) {
    return addMappingInternal(false, map, mapping);
  }
  var _maybeAddSegment = function _maybeAddSegment(map, genLine, genColumn, source, sourceLine, sourceColumn, name, content) {
    return addSegmentInternal(true, map, genLine, genColumn, source, sourceLine, sourceColumn, name, content);
  };
  var _maybeAddMapping = function _maybeAddMapping(map, mapping) {
    return addMappingInternal(true, map, mapping);
  };
  function _setSourceContent(map, source, content) {
    var _cast3 = cast2(map),
      sources = _cast3._sources,
      sourcesContent = _cast3._sourcesContent;
    var index = put(sources, source);
    sourcesContent[index] = content;
  }
  function _setIgnore(map, source) {
    var ignore = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : true;
    var _cast4 = cast2(map),
      sources = _cast4._sources,
      sourcesContent = _cast4._sourcesContent,
      ignoreList = _cast4._ignoreList;
    var index = put(sources, source);
    if (index === sourcesContent.length) sourcesContent[index] = null;
    if (ignore) put(ignoreList, index);else remove(ignoreList, index);
  }
  function _toDecodedMap(map) {
    var _cast5 = cast2(map),
      mappings = _cast5._mappings,
      sources = _cast5._sources,
      sourcesContent = _cast5._sourcesContent,
      names = _cast5._names,
      ignoreList = _cast5._ignoreList;
    removeEmptyFinalLines(mappings);
    return {
      version: 3,
      file: map.file || void 0,
      names: names.array,
      sourceRoot: map.sourceRoot || void 0,
      sources: sources.array,
      sourcesContent: sourcesContent,
      mappings: mappings,
      // originalScopes,
      // generatedRanges,
      ignoreList: ignoreList.array
    };
  }
  function _toEncodedMap(map) {
    var decoded = _toDecodedMap(map);
    return Object.assign({}, decoded, {
      // originalScopes: decoded.originalScopes.map((os) => encodeOriginalScopes(os)),
      // generatedRanges: encodeGeneratedRanges(decoded.generatedRanges as GeneratedRange[]),
      mappings: (0, import_sourcemap_codec.encode)(decoded.mappings)
    });
  }
  function _fromMap(input) {
    var map = new import_trace_mapping.TraceMap(input);
    var gen = new _GenMapping({
      file: map.file,
      sourceRoot: map.sourceRoot
    });
    putAll(cast2(gen)._names, map.names);
    putAll(cast2(gen)._sources, map.sources);
    cast2(gen)._sourcesContent = map.sourcesContent || map.sources.map(function () {
      return null;
    });
    cast2(gen)._mappings = (0, import_trace_mapping.decodedMappings)(map);
    if (map.ignoreList) putAll(cast2(gen)._ignoreList, map.ignoreList);
    return gen;
  }
  function _allMappings(map) {
    var out = [];
    var _cast6 = cast2(map),
      mappings = _cast6._mappings,
      sources = _cast6._sources,
      names = _cast6._names;
    for (var i = 0; i < mappings.length; i++) {
      var line = mappings[i];
      for (var j = 0; j < line.length; j++) {
        var seg = line[j];
        var generated = {
          line: i + 1,
          column: seg[COLUMN]
        };
        var source = void 0;
        var original = void 0;
        var name = void 0;
        if (seg.length !== 1) {
          source = sources.array[seg[SOURCES_INDEX]];
          original = {
            line: seg[SOURCE_LINE] + 1,
            column: seg[SOURCE_COLUMN]
          };
          if (seg.length === 5) name = names.array[seg[NAMES_INDEX]];
        }
        out.push({
          generated: generated,
          source: source,
          original: original,
          name: name
        });
      }
    }
    return out;
  }
  function addSegmentInternal(skipable, map, genLine, genColumn, source, sourceLine, sourceColumn, name, content) {
    var _cast7 = cast2(map),
      mappings = _cast7._mappings,
      sources = _cast7._sources,
      sourcesContent = _cast7._sourcesContent,
      names = _cast7._names;
    var line = getIndex(mappings, genLine);
    var index = getColumnIndex(line, genColumn);
    if (!source) {
      if (skipable && skipSourceless(line, index)) return;
      return insert(line, index, [genColumn]);
    }
    assert(sourceLine);
    assert(sourceColumn);
    var sourcesIndex = put(sources, source);
    var namesIndex = name ? put(names, name) : NO_NAME;
    if (sourcesIndex === sourcesContent.length) sourcesContent[sourcesIndex] = content != null ? content : null;
    if (skipable && skipSource(line, index, sourcesIndex, sourceLine, sourceColumn, namesIndex)) {
      return;
    }
    return insert(line, index, name ? [genColumn, sourcesIndex, sourceLine, sourceColumn, namesIndex] : [genColumn, sourcesIndex, sourceLine, sourceColumn]);
  }
  function assert(_val) {}
  function getIndex(arr, index) {
    for (var i = arr.length; i <= index; i++) {
      arr[i] = [];
    }
    return arr[index];
  }
  function getColumnIndex(line, genColumn) {
    var index = line.length;
    for (var i = index - 1; i >= 0; index = i--) {
      var current = line[i];
      if (genColumn >= current[COLUMN]) break;
    }
    return index;
  }
  function insert(array, index, value) {
    for (var i = array.length; i > index; i--) {
      array[i] = array[i - 1];
    }
    array[index] = value;
  }
  function removeEmptyFinalLines(mappings) {
    var length = mappings.length;
    var len = length;
    for (var i = len - 1; i >= 0; len = i, i--) {
      if (mappings[i].length > 0) break;
    }
    if (len < length) mappings.length = len;
  }
  function putAll(setarr, array) {
    for (var i = 0; i < array.length; i++) put(setarr, array[i]);
  }
  function skipSourceless(line, index) {
    if (index === 0) return true;
    var prev = line[index - 1];
    return prev.length === 1;
  }
  function skipSource(line, index, sourcesIndex, sourceLine, sourceColumn, namesIndex) {
    if (index === 0) return false;
    var prev = line[index - 1];
    if (prev.length === 1) return false;
    return sourcesIndex === prev[SOURCES_INDEX] && sourceLine === prev[SOURCE_LINE] && sourceColumn === prev[SOURCE_COLUMN] && namesIndex === (prev.length === 5 ? prev[NAMES_INDEX] : NO_NAME);
  }
  function addMappingInternal(skipable, map, mapping) {
    var generated = mapping.generated,
      source = mapping.source,
      original = mapping.original,
      name = mapping.name,
      content = mapping.content;
    if (!source) {
      return addSegmentInternal(skipable, map, generated.line - 1, generated.column, null, null, null, null, null);
    }
    assert(original);
    return addSegmentInternal(skipable, map, generated.line - 1, generated.column, source, original.line - 1, original.column, name, content);
  }
});
