// @discoveryjs/json-ext/src/parse-chunked.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.parseChunked = parseChunked;
var _utils = require("./utils.js");
function _regenerator() { /*! regenerator-runtime -- Copyright (c) 2014-present, Facebook, Inc. -- license (MIT): https://github.com/babel/babel/blob/main/packages/babel-helpers/LICENSE */ var e, t, r = "function" == typeof Symbol ? Symbol : {}, n = r.iterator || "@@iterator", o = r.toStringTag || "@@toStringTag"; function i(r, n, o, i) { var c = n && n.prototype instanceof Generator ? n : Generator, u = Object.create(c.prototype); return _regeneratorDefine2(u, "_invoke", function (r, n, o) { var i, c, u, f = 0, p = o || [], y = !1, G = { p: 0, n: 0, v: e, a: d, f: d.bind(e, 4), d: function d(t, r) { return i = t, c = 0, u = e, G.n = r, a; } }; function d(r, n) { for (c = r, u = n, t = 0; !y && f && !o && t < p.length; t++) { var o, i = p[t], d = G.p, l = i[2]; r > 3 ? (o = l === n) && (u = i[(c = i[4]) ? 5 : (c = 3, 3)], i[4] = i[5] = e) : i[0] <= d && ((o = r < 2 && d < i[1]) ? (c = 0, G.v = n, G.n = i[1]) : d < l && (o = r < 3 || i[0] > n || n > l) && (i[4] = r, i[5] = n, G.n = l, c = 0)); } if (o || r > 1) return a; throw y = !0, n; } return function (o, p, l) { if (f > 1) throw TypeError("Generator is already running"); for (y && 1 === p && d(p, l), c = p, u = l; (t = c < 2 ? e : u) || !y;) { i || (c ? c < 3 ? (c > 1 && (G.n = -1), d(c, u)) : G.n = u : G.v = u); try { if (f = 2, i) { if (c || (o = "next"), t = i[o]) { if (!(t = t.call(i, u))) throw TypeError("iterator result is not an object"); if (!t.done) return t; u = t.value, c < 2 && (c = 0); } else 1 === c && (t = i["return"]) && t.call(i), c < 2 && (u = TypeError("The iterator does not provide a '" + o + "' method"), c = 1); i = e; } else if ((t = (y = G.n < 0) ? u : r.call(n, G)) !== a) break; } catch (t) { i = e, c = 1, u = t; } finally { f = 1; } } return { value: t, done: y }; }; }(r, o, i), !0), u; } var a = {}; function Generator() {} function GeneratorFunction() {} function GeneratorFunctionPrototype() {} t = Object.getPrototypeOf; var c = [][n] ? t(t([][n]())) : (_regeneratorDefine2(t = {}, n, function () { return this; }), t), u = GeneratorFunctionPrototype.prototype = Generator.prototype = Object.create(c); function f(e) { return Object.setPrototypeOf ? Object.setPrototypeOf(e, GeneratorFunctionPrototype) : (e.__proto__ = GeneratorFunctionPrototype, _regeneratorDefine2(e, o, "GeneratorFunction")), e.prototype = Object.create(u), e; } return GeneratorFunction.prototype = GeneratorFunctionPrototype, _regeneratorDefine2(u, "constructor", GeneratorFunctionPrototype), _regeneratorDefine2(GeneratorFunctionPrototype, "constructor", GeneratorFunction), GeneratorFunction.displayName = "GeneratorFunction", _regeneratorDefine2(GeneratorFunctionPrototype, o, "GeneratorFunction"), _regeneratorDefine2(u), _regeneratorDefine2(u, o, "Generator"), _regeneratorDefine2(u, n, function () { return this; }), _regeneratorDefine2(u, "toString", function () { return "[object Generator]"; }), (_regenerator = function _regenerator() { return { w: i, m: f }; })(); }
function _regeneratorDefine2(e, r, n, t) { var i = Object.defineProperty; try { i({}, "", {}); } catch (e) { i = 0; } _regeneratorDefine2 = function _regeneratorDefine(e, r, n, t) { function o(r, n) { _regeneratorDefine2(e, r, function (e) { return this._invoke(r, n, e); }); } r ? i ? i(e, r, { value: n, enumerable: !t, configurable: !t, writable: !t }) : e[r] = n : (o("next", 0), o("throw", 1), o("return", 2)); }, _regeneratorDefine2(e, r, n, t); }
function _toConsumableArray(r) { return _arrayWithoutHoles(r) || _iterableToArray(r) || _unsupportedIterableToArray(r) || _nonIterableSpread(); }
function _nonIterableSpread() { throw new TypeError("Invalid attempt to spread non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }
function _unsupportedIterableToArray(r, a) { if (r) { if ("string" == typeof r) return _arrayLikeToArray(r, a); var t = {}.toString.call(r).slice(8, -1); return "Object" === t && r.constructor && (t = r.constructor.name), "Map" === t || "Set" === t ? Array.from(r) : "Arguments" === t || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(t) ? _arrayLikeToArray(r, a) : void 0; } }
function _iterableToArray(r) { if ("undefined" != typeof Symbol && null != r[Symbol.iterator] || null != r["@@iterator"]) return Array.from(r); }
function _arrayWithoutHoles(r) { if (Array.isArray(r)) return _arrayLikeToArray(r); }
function _arrayLikeToArray(r, a) { (null == a || a > r.length) && (a = r.length); for (var e = 0, n = Array(a); e < a; e++) n[e] = r[e]; return n; }
function asyncGeneratorStep(n, t, e, r, o, a, c) { try { var i = n[a](c), u = i.value; } catch (n) { return void e(n); } i.done ? t(u) : Promise.resolve(u).then(r, o); }
function _asyncToGenerator(n) { return function () { var t = this, e = arguments; return new Promise(function (r, o) { var a = n.apply(t, e); function _next(n) { asyncGeneratorStep(a, r, o, _next, _throw, "next", n); } function _throw(n) { asyncGeneratorStep(a, r, o, _next, _throw, "throw", n); } _next(void 0); }); }; }
function _typeof(o) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (o) { return typeof o; } : function (o) { return o && "function" == typeof Symbol && o.constructor === Symbol && o !== Symbol.prototype ? "symbol" : typeof o; }, _typeof(o); }
function _asyncIterator(r) { var n, t, o, e = 2; for ("undefined" != typeof Symbol && (t = Symbol.asyncIterator, o = Symbol.iterator); e--;) { if (t && null != (n = r[t])) return n.call(r); if (o && null != (n = r[o])) return new AsyncFromSyncIterator(n.call(r)); t = "@@asyncIterator", o = "@@iterator"; } throw new TypeError("Object is not async iterable"); }
function AsyncFromSyncIterator(r) { function AsyncFromSyncIteratorContinuation(r) { if (Object(r) !== r) return Promise.reject(new TypeError(r + " is not an object.")); var n = r.done; return Promise.resolve(r.value).then(function (r) { return { value: r, done: n }; }); } return AsyncFromSyncIterator = function AsyncFromSyncIterator(r) { this.s = r, this.n = r.next; }, AsyncFromSyncIterator.prototype = { s: null, n: null, next: function next() { return AsyncFromSyncIteratorContinuation(this.n.apply(this.s, arguments)); }, "return": function _return(r) { var n = this.s["return"]; return void 0 === n ? Promise.resolve({ value: r, done: !0 }) : AsyncFromSyncIteratorContinuation(n.apply(this.s, arguments)); }, "throw": function _throw(r) { var n = this.s["return"]; return void 0 === n ? Promise.reject(r) : AsyncFromSyncIteratorContinuation(n.apply(this.s, arguments)); } }, new AsyncFromSyncIterator(r); }
var NO_VALUE = Symbol('empty');
var STACK_OBJECT = 1;
var STACK_ARRAY = 2;
var MODE_JSON = 0;
var MODE_JSONL = 1;
var MODE_JSONL_AUTO = 2;
var decoder = new TextDecoder();
function adjustPosition(error, jsonParseOffset) {
  if (error.name === 'SyntaxError' && jsonParseOffset) {
    error.message = error.message.replace(/at position (\d+)/, function (_, pos) {
      return 'at position ' + (Number(pos) + jsonParseOffset);
    });
  }
  return error;
}
function append(array, elements) {
  // Note: Avoid using array.push(...elements) since it may lead to
  // "RangeError: Maximum call stack size exceeded" for long arrays
  var initialLength = array.length;
  array.length += elements.length;
  for (var i = 0; i < elements.length; i++) {
    array[initialLength + i] = elements[i];
  }
}
function resolveParseMode(mode) {
  switch (mode) {
    case 'json':
      return MODE_JSON;
    case 'jsonl':
      return MODE_JSONL;
    case 'auto':
      return MODE_JSONL_AUTO;
    default:
      throw new TypeError('Invalid options: `mode` should be "json", "jsonl", or "auto"');
  }
}
function parseChunkedOptions(value) {
  var _options$mode, _options$reviver, _options$onRootValue, _options$onChunk;
  var options = typeof value === 'function' ? {
    reviver: value
  } : value || {};
  return {
    mode: resolveParseMode((_options$mode = options.mode) !== null && _options$mode !== void 0 ? _options$mode : 'json'),
    reviver: (_options$reviver = options.reviver) !== null && _options$reviver !== void 0 ? _options$reviver : null,
    onRootValue: (_options$onRootValue = options.onRootValue) !== null && _options$onRootValue !== void 0 ? _options$onRootValue : null,
    onChunk: (_options$onChunk = options.onChunk) !== null && _options$onChunk !== void 0 ? _options$onChunk : null
  };
}
function applyReviver(value, reviver) {
  return walk({
    '': value
  }, '', value);
  function walk(holder, key, value) {
    if (value && _typeof(value) === 'object') {
      for (var _i = 0, _Object$keys = Object.keys(value); _i < _Object$keys.length; _i++) {
        var childKey = _Object$keys[_i];
        var childValue = value[childKey];
        var newValue = walk(value, childKey, childValue);
        if (newValue === undefined) {
          delete value[childKey];
        } else if (newValue !== childValue) {
          value[childKey] = newValue;
        }
      }
    }
    return reviver.call(holder, key, value);
  }
}
function parseChunked(_x, _x2) {
  return _parseChunked.apply(this, arguments);
}
function _parseChunked() {
  _parseChunked = _asyncToGenerator(/*#__PURE__*/_regenerator().m(function _callee(chunkEmitter, optionsOrReviver) {
    var _parseChunkedOptions, mode, reviver, onRootValue, onChunk, iterable, parser, _iteratorAbruptCompletion, _didIteratorError, _iteratorError, _iterator, _step, chunk, _t, _t2;
    return _regenerator().w(function (_context) {
      while (1) switch (_context.p = _context.n) {
        case 0:
          _parseChunkedOptions = parseChunkedOptions(optionsOrReviver), mode = _parseChunkedOptions.mode, reviver = _parseChunkedOptions.reviver, onRootValue = _parseChunkedOptions.onRootValue, onChunk = _parseChunkedOptions.onChunk;
          iterable = typeof chunkEmitter === 'function' ? chunkEmitter() : chunkEmitter;
          if (!(0, _utils.isIterable)(iterable)) {
            _context.n = 16;
            break;
          }
          parser = createChunkParser(mode, reviver, onRootValue, onChunk);
          _context.p = 1;
          _iteratorAbruptCompletion = false;
          _didIteratorError = false;
          _context.p = 2;
          _iterator = _asyncIterator(iterable);
        case 3:
          _context.n = 4;
          return _iterator.next();
        case 4:
          if (!(_iteratorAbruptCompletion = !(_step = _context.v).done)) {
            _context.n = 7;
            break;
          }
          chunk = _step.value;
          if (!(typeof chunk !== 'string' && !ArrayBuffer.isView(chunk))) {
            _context.n = 5;
            break;
          }
          throw new TypeError('Invalid chunk: Expected string, TypedArray or Buffer');
        case 5:
          parser.push(chunk);
        case 6:
          _iteratorAbruptCompletion = false;
          _context.n = 3;
          break;
        case 7:
          _context.n = 9;
          break;
        case 8:
          _context.p = 8;
          _t = _context.v;
          _didIteratorError = true;
          _iteratorError = _t;
        case 9:
          _context.p = 9;
          _context.p = 10;
          if (!(_iteratorAbruptCompletion && _iterator["return"] != null)) {
            _context.n = 11;
            break;
          }
          _context.n = 11;
          return _iterator["return"]();
        case 11:
          _context.p = 11;
          if (!_didIteratorError) {
            _context.n = 12;
            break;
          }
          throw _iteratorError;
        case 12:
          return _context.f(11);
        case 13:
          return _context.f(9);
        case 14:
          return _context.a(2, parser.finish());
        case 15:
          _context.p = 15;
          _t2 = _context.v;
          throw adjustPosition(_t2, parser.jsonParseOffset);
        case 16:
          throw new TypeError('Invalid chunk emitter: Expected an Iterable, AsyncIterable, generator, ' + 'async generator, or a function returning an Iterable or AsyncIterable');
        case 17:
          return _context.a(2);
      }
    }, _callee, null, [[10,, 11, 13], [2, 8, 9, 14], [1, 15]]);
  }));
  return _parseChunked.apply(this, arguments);
}
;
function createChunkParser(parseMode, reviver, onRootValue, onChunk) {
  var rootValues = parseMode === MODE_JSONL ? [] : null;
  var rootValuesCount = 0;
  var currentRootValue = NO_VALUE;
  var currentRootValueCursor = null;
  var consumedChunkLength = 0;
  var parsedChunkLength = 0;
  var prevArray = null;
  var prevArraySlices = [];
  var stack = new Array(100);
  var lastFlushDepth = 0;
  var flushDepth = 0;
  var stateString = false;
  var stateStringEscape = false;
  var seenNonWhiteSpace = false;
  var allowNewRootValue = true;
  var pendingByteSeq = null;
  var pendingChunk = null;
  var jsonParseOffset = 0;
  var state = Object.freeze({
    get mode() {
      return parseMode === MODE_JSONL ? 'jsonl' : 'json';
    },
    get returnValue() {
      return typeof onRootValue === 'function' ? rootValuesCount : rootValues !== null ? rootValues : currentRootValue !== NO_VALUE ? currentRootValue : undefined;
    },
    get currentRootValue() {
      return currentRootValue !== NO_VALUE ? currentRootValue : undefined;
    },
    get rootValuesCount() {
      return rootValuesCount;
    },
    get consumed() {
      return consumedChunkLength;
    },
    get parsed() {
      return parsedChunkLength;
    }
  });
  return {
    push: push,
    finish: finish,
    state: state,
    get jsonParseOffset() {
      return jsonParseOffset;
    }
  };
  function startRootValue(fragment) {
    // Extra non-whitespace after complete root value should fail to parse
    if (!allowNewRootValue) {
      jsonParseOffset -= 2;
      JSON.parse('[]' + fragment);
    }

    // In "auto" mode, switch to JSONL when a second root value is starting after a newline
    if (currentRootValue !== NO_VALUE && parseMode === MODE_JSONL_AUTO) {
      parseMode = MODE_JSONL;
      rootValues = [currentRootValue];
    }

    // Block parsing of an additional root value until a newline is encountered
    allowNewRootValue = false;

    // Parse fragment as a new root value
    currentRootValue = JSON.parse(fragment);
  }
  function finishRootValue() {
    rootValuesCount++;
    if (typeof reviver === 'function') {
      currentRootValue = applyReviver(currentRootValue, reviver);
    }
    if (typeof onRootValue === 'function') {
      onRootValue(currentRootValue, state);
    } else if (parseMode === MODE_JSONL) {
      rootValues.push(currentRootValue);
    }
  }
  function mergeArraySlices() {
    if (prevArray === null) {
      return;
    }
    if (prevArraySlices.length !== 0) {
      var _prevArray;
      var newArray = prevArraySlices.length === 1 ? prevArray.concat(prevArraySlices[0]) : (_prevArray = prevArray).concat.apply(_prevArray, _toConsumableArray(prevArraySlices));
      if (currentRootValueCursor.prev !== null) {
        currentRootValueCursor.prev.value[currentRootValueCursor.key] = newArray;
      } else {
        currentRootValue = newArray;
      }
      currentRootValueCursor.value = newArray;
      prevArraySlices = [];
    }
    prevArray = null;
  }
  function parseAndAppend(fragment, wrap) {
    // Append new entries or elements
    if (stack[lastFlushDepth - 1] === STACK_OBJECT) {
      if (wrap) {
        jsonParseOffset--;
        fragment = '{' + fragment + '}';
      }
      Object.assign(currentRootValueCursor.value, JSON.parse(fragment));
    } else {
      if (wrap) {
        jsonParseOffset--;
        fragment = '[' + fragment + ']';
      }
      if (prevArray === currentRootValueCursor.value) {
        prevArraySlices.push(JSON.parse(fragment));
      } else {
        append(currentRootValueCursor.value, JSON.parse(fragment));
        prevArray = currentRootValueCursor.value;
      }
    }
  }
  function prepareAddition(fragment) {
    var _currentRootValueCurs = currentRootValueCursor,
      value = _currentRootValueCurs.value;
    var expectComma = Array.isArray(value) ? value.length !== 0 : Object.keys(value).length !== 0;
    if (expectComma) {
      // Skip a comma at the beginning of fragment, otherwise it would
      // fail to parse
      if (fragment[0] === ',') {
        jsonParseOffset++;
        return fragment.slice(1);
      }

      // When value (an object or array) is not empty and a fragment
      // doesn't start with a comma, a single valid fragment starting
      // is a closing bracket. If it's not, a prefix is adding to fail
      // parsing. Otherwise, the sequence of chunks can be successfully
      // parsed, although it should not, e.g. ["[{}", "{}]"]
      if (fragment[0] !== '}' && fragment[0] !== ']') {
        jsonParseOffset -= 3;
        return '[[]' + fragment;
      }
    }
    return fragment;
  }
  function flush(chunk, start, end) {
    var fragment = chunk.slice(start, end);

    // Save position correction for an error in JSON.parse() if any
    jsonParseOffset = consumedChunkLength + start;
    parsedChunkLength += end - start;

    // Prepend pending chunk if any
    if (pendingChunk !== null) {
      fragment = pendingChunk + fragment;
      jsonParseOffset -= pendingChunk.length;
      parsedChunkLength += pendingChunk.length;
      pendingChunk = null;
    }
    if (flushDepth === lastFlushDepth) {
      // Depth didn't change, so it's a continuation of the current value or entire value if it's a root one
      if (lastFlushDepth === 0) {
        startRootValue(fragment);
      } else {
        parseAndAppend(prepareAddition(fragment), true);
      }
    } else if (flushDepth > lastFlushDepth) {
      // Add missed closing brackets/parentheses
      for (var i = flushDepth - 1; i >= lastFlushDepth; i--) {
        fragment += stack[i] === STACK_OBJECT ? '}' : ']';
      }
      if (lastFlushDepth === 0) {
        startRootValue(fragment);
        currentRootValueCursor = {
          value: currentRootValue,
          key: null,
          prev: null
        };
      } else {
        parseAndAppend(prepareAddition(fragment), true);
        mergeArraySlices();
      }

      // Move down to the depths to the last object/array, which is current now
      for (var _i2 = lastFlushDepth || 1; _i2 < flushDepth; _i2++) {
        var _currentRootValueCurs2 = currentRootValueCursor,
          value = _currentRootValueCurs2.value;
        var key = null;
        if (stack[_i2 - 1] === STACK_OBJECT) {
          // Find last entry
          // eslint-disable-next-line curly
          for (key in value);
          value = value[key];
        } else {
          // Last element
          key = value.length - 1;
          value = value[key];
        }
        currentRootValueCursor = {
          value: value,
          key: key,
          prev: currentRootValueCursor
        };
      }
    } else /* flushDepth < lastFlushDepth */{
        fragment = prepareAddition(fragment);

        // Add missed opening brackets/parentheses
        for (var _i3 = lastFlushDepth - 1; _i3 >= flushDepth; _i3--) {
          jsonParseOffset--;
          fragment = (stack[_i3] === STACK_OBJECT ? '{' : '[') + fragment;
        }
        parseAndAppend(fragment, false);
        mergeArraySlices();
        for (var _i4 = lastFlushDepth - 1; _i4 >= flushDepth; _i4--) {
          currentRootValueCursor = currentRootValueCursor.prev;
        }
      }
    if (flushDepth === 0) {
      finishRootValue();
    }
    lastFlushDepth = flushDepth;
    seenNonWhiteSpace = false;
  }
  function ensureChunkString(chunk) {
    if (typeof chunk !== 'string') {
      // Suppose chunk is Buffer or Uint8Array

      // Prepend uncompleted byte sequence if any
      if (pendingByteSeq !== null) {
        var origRawChunk = chunk;
        chunk = new Uint8Array(pendingByteSeq.length + origRawChunk.length);
        chunk.set(pendingByteSeq);
        chunk.set(origRawChunk, pendingByteSeq.length);
        pendingByteSeq = null;
      }

      // In case Buffer/Uint8Array, an input is encoded in UTF8
      // Seek for parts of uncompleted UTF8 symbol on the ending
      // This makes sense only if we expect more chunks and last char is not multi-bytes
      if (chunk[chunk.length - 1] > 127) {
        for (var seqLength = 0; seqLength < chunk.length; seqLength++) {
          var _byte = chunk[chunk.length - 1 - seqLength];

          // 10xxxxxx - 2nd, 3rd or 4th byte
          // 110xxxxx – first byte of 2-byte sequence
          // 1110xxxx - first byte of 3-byte sequence
          // 11110xxx - first byte of 4-byte sequence
          if (_byte >> 6 === 3) {
            seqLength++;

            // If the sequence is really incomplete, then preserve it
            // for the future chunk and cut off it from the current chunk
            if (seqLength !== 4 && _byte >> 3 === 30 || seqLength !== 3 && _byte >> 4 === 14 || seqLength !== 2 && _byte >> 5 === 6) {
              pendingByteSeq = chunk.slice(chunk.length - seqLength); // use slice to avoid tying chunk
              chunk = chunk.subarray(0, -seqLength); // use subarray to avoid buffer copy
            }
            break;
          }
        }
      }

      // Convert chunk to a string, since single decode per chunk
      // is much effective than decode multiple small substrings
      chunk = decoder.decode(chunk);
    }
    return chunk;
  }
  function push(chunk) {
    chunk = ensureChunkString(chunk);
    var chunkLength = chunk.length;
    var prevParsedChunkLength = parsedChunkLength;
    var lastFlushPoint = 0;
    var flushPoint = 0;

    // Main scan loop
    scan: for (var i = 0; i < chunkLength; i++) {
      if (stateString) {
        for (; i < chunkLength; i++) {
          if (stateStringEscape) {
            stateStringEscape = false;
          } else {
            switch (chunk.charCodeAt(i)) {
              case 0x22:
                /* " */
                stateString = false;
                continue scan;
              case 0x5C:
                /* \ */
                stateStringEscape = true;
            }
          }
        }
        break;
      }
      switch (chunk.charCodeAt(i)) {
        case 0x22:
          /* " */
          stateString = true;
          stateStringEscape = false;
          seenNonWhiteSpace = true;
          break;
        case 0x2C:
          /* , */
          flushPoint = i;
          break;
        case 0x7B:
          /* { */
          // Open an object
          flushPoint = i + 1;
          stack[flushDepth++] = STACK_OBJECT;
          seenNonWhiteSpace = true;
          break;
        case 0x5B:
          /* [ */
          // Open an array
          flushPoint = i + 1;
          stack[flushDepth++] = STACK_ARRAY;
          seenNonWhiteSpace = true;
          break;
        case 0x5D: /* ] */
        case 0x7D:
          /* } */
          // Close an object or array
          flushPoint = i + 1;
          if (flushDepth === 0) {
            // Unmatched closing bracket/brace at top level, should fail to parse
            break scan;
          }
          flushDepth--;

          // Flush on depth decrease related to last flush, otherwise wait for more chunks to flush together
          if (flushDepth < lastFlushDepth) {
            flush(chunk, lastFlushPoint, flushPoint);
            lastFlushPoint = flushPoint;
          }
          break;
        case 0x09: /* \t */
        case 0x0A: /* \n */
        case 0x0D: /* \r */
        case 0x20:
          /* space */
          if (flushDepth === 0) {
            if (seenNonWhiteSpace) {
              flushPoint = i;
              flush(chunk, lastFlushPoint, flushPoint);
              lastFlushPoint = flushPoint;
            }
            if (parseMode !== MODE_JSON && allowNewRootValue === false && (chunk.charCodeAt(i) === 0x0A || chunk.charCodeAt(i) === 0x0D)) {
              allowNewRootValue = true;
            }
            if (flushPoint === i) {
              parsedChunkLength++;
            }
          }

          // Move points forward when they point to current position and it's a whitespace
          if (lastFlushPoint === i) {
            lastFlushPoint++;
          }
          if (flushPoint === i) {
            flushPoint++;
          }
          break;
        default:
          seenNonWhiteSpace = true;
      }
    }
    if (flushPoint > lastFlushPoint) {
      flush(chunk, lastFlushPoint, flushPoint);
    }

    // Produce pendingChunk if something left
    if (flushPoint < chunkLength) {
      if (pendingChunk !== null) {
        // When there is already a pending chunk then no flush happened,
        // appending entire chunk to pending one
        pendingChunk += chunk;
      } else {
        // Create a pending chunk, it will start with non-whitespace since
        // flushPoint was moved forward away from whitespaces on scan
        pendingChunk = chunk.slice(flushPoint, chunkLength);
      }
    }
    consumedChunkLength += chunkLength;
    if (typeof onChunk === 'function') {
      onChunk(parsedChunkLength - prevParsedChunkLength, chunk, pendingChunk, state);
    }
  }
  function finish() {
    if (pendingChunk !== null || currentRootValue === NO_VALUE && parseMode !== MODE_JSONL) {
      // Force the `flushDepth < lastFlushDepth` branch in flush() to prepend missed
      // opening brackets/parentheses and produce a natural JSON.parse() EOF error
      flushDepth = 0;
      flush('', 0, 0);
    }
    if (typeof onChunk === 'function') {
      parsedChunkLength = consumedChunkLength;
      onChunk(0, null, null, state);
    }
    var result = state.returnValue;
    rootValues = null;
    currentRootValue = NO_VALUE;
    return result;
  }
}
