// @discoveryjs/json-ext/src/stringify-info.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.stringifyInfo = stringifyInfo;
var _utils = require("./utils.js");
var _excluded = ["replacer", "getKeys"];
function _createForOfIteratorHelper(r, e) { var t = "undefined" != typeof Symbol && r[Symbol.iterator] || r["@@iterator"]; if (!t) { if (Array.isArray(r) || (t = _unsupportedIterableToArray(r)) || e && r && "number" == typeof r.length) { t && (r = t); var _n = 0, F = function F() {}; return { s: F, n: function n() { return _n >= r.length ? { done: !0 } : { done: !1, value: r[_n++] }; }, e: function e(r) { throw r; }, f: F }; } throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); } var o, a = !0, u = !1; return { s: function s() { t = t.call(r); }, n: function n() { var r = t.next(); return a = r.done, r; }, e: function e(r) { u = !0, o = r; }, f: function f() { try { a || null == t["return"] || t["return"](); } finally { if (u) throw o; } } }; }
function _toConsumableArray(r) { return _arrayWithoutHoles(r) || _iterableToArray(r) || _unsupportedIterableToArray(r) || _nonIterableSpread(); }
function _nonIterableSpread() { throw new TypeError("Invalid attempt to spread non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }
function _unsupportedIterableToArray(r, a) { if (r) { if ("string" == typeof r) return _arrayLikeToArray(r, a); var t = {}.toString.call(r).slice(8, -1); return "Object" === t && r.constructor && (t = r.constructor.name), "Map" === t || "Set" === t ? Array.from(r) : "Arguments" === t || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(t) ? _arrayLikeToArray(r, a) : void 0; } }
function _iterableToArray(r) { if ("undefined" != typeof Symbol && null != r[Symbol.iterator] || null != r["@@iterator"]) return Array.from(r); }
function _arrayWithoutHoles(r) { if (Array.isArray(r)) return _arrayLikeToArray(r); }
function _arrayLikeToArray(r, a) { (null == a || a > r.length) && (a = r.length); for (var e = 0, n = Array(a); e < a; e++) n[e] = r[e]; return n; }
function _objectWithoutProperties(e, t) { if (null == e) return {}; var o, r, i = _objectWithoutPropertiesLoose(e, t); if (Object.getOwnPropertySymbols) { var n = Object.getOwnPropertySymbols(e); for (r = 0; r < n.length; r++) o = n[r], -1 === t.indexOf(o) && {}.propertyIsEnumerable.call(e, o) && (i[o] = e[o]); } return i; }
function _objectWithoutPropertiesLoose(r, e) { if (null == r) return {}; var t = {}; for (var n in r) if ({}.hasOwnProperty.call(r, n)) { if (-1 !== e.indexOf(n)) continue; t[n] = r[n]; } return t; }
function _typeof(o) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (o) { return typeof o; } : function (o) { return o && "function" == typeof Symbol && o.constructor === Symbol && o !== Symbol.prototype ? "symbol" : typeof o; }, _typeof(o); }
var hasOwn = typeof Object.hasOwn === 'function' ? Object.hasOwn : function (object, key) {
  return Object.hasOwnProperty.call(object, key);
};

// https://tc39.es/ecma262/#table-json-single-character-escapes
var escapableCharCodeSubstitution = {
  // JSON Single Character Escape Sequences
  0x08: '\\b',
  0x09: '\\t',
  0x0a: '\\n',
  0x0c: '\\f',
  0x0d: '\\r',
  0x22: '\\\"',
  0x5c: '\\\\'
};
var charLength2048 = Uint8Array.from({
  length: 2048
}, function (_, code) {
  if (hasOwn(escapableCharCodeSubstitution, code)) {
    return 2; // \X
  }
  if (code < 0x20) {
    return 6; // \uXXXX
  }
  return code < 128 ? 1 : 2; // UTF8 bytes
});
function isLeadingSurrogate(code) {
  return code >= 0xD800 && code <= 0xDBFF;
}
function isTrailingSurrogate(code) {
  return code >= 0xDC00 && code <= 0xDFFF;
}
function stringLength(str) {
  // Fast path to compute length when a string contains only characters encoded as single bytes
  if (!/[^\x20\x21\x23-\x5B\x5D-\x7F]/.test(str)) {
    return str.length + 2;
  }
  var len = 0;
  var prevLeadingSurrogate = false;
  for (var i = 0; i < str.length; i++) {
    var code = str.charCodeAt(i);
    if (code < 2048) {
      len += charLength2048[code];
    } else if (isLeadingSurrogate(code)) {
      len += 6; // \uXXXX since no pair with trailing surrogate yet
      prevLeadingSurrogate = true;
      continue;
    } else if (isTrailingSurrogate(code)) {
      len = prevLeadingSurrogate ? len - 2 // surrogate pair (4 bytes), since we calculate prev leading surrogate as 6 bytes, substruct 2 bytes
      : len + 6; // \uXXXX
    } else {
      len += 3; // code >= 2048 is 3 bytes length for UTF8
    }
    prevLeadingSurrogate = false;
  }
  return len + 2; // +2 for quotes
}

// avoid producing a string from a number
function intLength(num) {
  var len = 0;
  if (num < 0) {
    len = 1;
    num = -num;
  }
  if (num >= 1e9) {
    len += 9;
    num = (num - num % 1e9) / 1e9;
  }
  if (num >= 1e4) {
    if (num >= 1e6) {
      return len + (num >= 1e8 ? 9 : num >= 1e7 ? 8 : 7);
    }
    return len + (num >= 1e5 ? 6 : 5);
  }
  return len + (num >= 1e2 ? num >= 1e3 ? 4 : 3 : num >= 10 ? 2 : 1);
}
;
function primitiveLength(value) {
  switch (_typeof(value)) {
    case 'string':
      return stringLength(value);
    case 'number':
      return Number.isFinite(value) ? Number.isInteger(value) ? intLength(value) : String(value).length : 4 /* null */;
    case 'boolean':
      return value ? 4 /* true */ : 5 /* false */;
    case 'undefined':
    case 'object':
      return 4;
    /* null */

    default:
      return 0;
  }
}
function stringifyInfo(value) {
  var _options$space;
  for (var _len = arguments.length, args = new Array(_len > 1 ? _len - 1 : 0), _key = 1; _key < _len; _key++) {
    args[_key - 1] = arguments[_key];
  }
  var _normalizeStringifyOp = _utils.normalizeStringifyOptions.apply(void 0, args),
    replacer = _normalizeStringifyOp.replacer,
    getKeys = _normalizeStringifyOp.getKeys,
    options = _objectWithoutProperties(_normalizeStringifyOp, _excluded);
  var continueOnCircular = Boolean(options.continueOnCircular);
  var space = ((_options$space = options.space) === null || _options$space === void 0 ? void 0 : _options$space.length) || 0;
  var roots = (0, _utils.resolveStringifyMode)(options.mode) === 'jsonl' && Array.isArray(value) ? value : [value];
  var keysLength = new Map();
  var visited = new Map();
  var circular = new Set();
  var stack = [];
  var stop = false;
  var bytes = 0;
  var spaceBytes = 0;
  var objects = 0;
  for (var i = 0; i < roots.length; i++) {
    if (i > 0) {
      bytes += 1; // newline separator
    }
    walk({
      '': roots[i]
    }, '', roots[i]);
  }

  // when value is undefined or replaced for undefined
  if (bytes === 0 && roots.length === 1) {
    bytes += 9; // FIXME: that's the length of undefined, should we normalize behaviour to convert it to null?
  }
  return {
    bytes: isNaN(bytes) ? Infinity : bytes + spaceBytes,
    spaceBytes: space > 0 && isNaN(bytes) ? Infinity : spaceBytes,
    circular: _toConsumableArray(circular)
  };
  function walk(holder, key, value) {
    if (stop) {
      return;
    }
    value = (0, _utils.replaceValue)(holder, key, value, replacer);
    if (value === null || _typeof(value) !== 'object') {
      // primitive
      if (value !== undefined || Array.isArray(holder)) {
        bytes += primitiveLength(value);
      }
    } else {
      // check for circular references
      if (stack.includes(value)) {
        circular.add(value);
        bytes += 4; // treat as null

        if (!continueOnCircular) {
          stop = true;
        }
        return;
      }

      // Using 'visited' allows avoiding hang-ups in cases of highly interconnected object graphs;
      // for example, a list of git commits with references to parents can lead to N^2 complexity for traversal,
      // and N when 'visited' is used
      if (visited.has(value)) {
        bytes += visited.get(value);
        return;
      }
      objects++;
      var prevObjects = objects;
      var valueBytes = bytes;
      var valueLength = 0;
      stack.push(value);
      if (Array.isArray(value)) {
        // array
        valueLength = value.length;
        for (var _i = 0; _i < valueLength; _i++) {
          walk(value, _i, value[_i]);
        }
      } else {
        // object
        var prevLength = bytes;
        var _iterator = _createForOfIteratorHelper(getKeys(value)),
          _step;
        try {
          for (_iterator.s(); !(_step = _iterator.n()).done;) {
            var _key2 = _step.value;
            walk(value, _key2, value[_key2]);
            if (prevLength !== bytes) {
              var keyLen = keysLength.get(_key2);
              if (keyLen === undefined) {
                keysLength.set(_key2, keyLen = stringLength(_key2) + 1); // "key":
              }

              // value is printed
              bytes += keyLen;
              valueLength++;
              prevLength = bytes;
            }
          }
        } catch (err) {
          _iterator.e(err);
        } finally {
          _iterator.f();
        }
      }
      bytes += valueLength === 0 ? 2 // {} or []
      : 1 + valueLength; // {} or [] + commas

      if (space > 0 && valueLength > 0) {
        spaceBytes +=
        // a space between ":" and a value for each object entry
        (Array.isArray(value) ? 0 : valueLength) +
        // the formula results from folding the following components:
        // - for each key-value or element: ident + newline
        //   (1 + stack.length * space) * valueLength
        // - ident (one space less) before "}" or "]" + newline
        //   (stack.length - 1) * space + 1
        (1 + stack.length * space) * (valueLength + 1) - space;
      }
      stack.pop();

      // add to 'visited' only objects that contain nested objects
      if (prevObjects !== objects) {
        visited.set(value, bytes - valueBytes);
      }
    }
  }
}
;
