// @discoveryjs/json-ext/src/stringify-chunked.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.stringifyChunked = stringifyChunked;
var _utils = require("./utils.js");
var _excluded = ["replacer", "getKeys", "space"];
function _regeneratorValues(e) { if (null != e) { var t = e["function" == typeof Symbol && Symbol.iterator || "@@iterator"], r = 0; if (t) return t.call(e); if ("function" == typeof e.next) return e; if (!isNaN(e.length)) return { next: function next() { return e && r >= e.length && (e = void 0), { value: e && e[r++], done: !e }; } }; } throw new TypeError(_typeof(e) + " is not iterable"); }
var _marked = /*#__PURE__*/_regenerator().m(stringifyChunked);
function _regenerator() { /*! regenerator-runtime -- Copyright (c) 2014-present, Facebook, Inc. -- license (MIT): https://github.com/babel/babel/blob/main/packages/babel-helpers/LICENSE */ var e, t, r = "function" == typeof Symbol ? Symbol : {}, n = r.iterator || "@@iterator", o = r.toStringTag || "@@toStringTag"; function i(r, n, o, i) { var c = n && n.prototype instanceof Generator ? n : Generator, u = Object.create(c.prototype); return _regeneratorDefine2(u, "_invoke", function (r, n, o) { var i, c, u, f = 0, p = o || [], y = !1, G = { p: 0, n: 0, v: e, a: d, f: d.bind(e, 4), d: function d(t, r) { return i = t, c = 0, u = e, G.n = r, a; } }; function d(r, n) { for (c = r, u = n, t = 0; !y && f && !o && t < p.length; t++) { var o, i = p[t], d = G.p, l = i[2]; r > 3 ? (o = l === n) && (u = i[(c = i[4]) ? 5 : (c = 3, 3)], i[4] = i[5] = e) : i[0] <= d && ((o = r < 2 && d < i[1]) ? (c = 0, G.v = n, G.n = i[1]) : d < l && (o = r < 3 || i[0] > n || n > l) && (i[4] = r, i[5] = n, G.n = l, c = 0)); } if (o || r > 1) return a; throw y = !0, n; } return function (o, p, l) { if (f > 1) throw TypeError("Generator is already running"); for (y && 1 === p && d(p, l), c = p, u = l; (t = c < 2 ? e : u) || !y;) { i || (c ? c < 3 ? (c > 1 && (G.n = -1), d(c, u)) : G.n = u : G.v = u); try { if (f = 2, i) { if (c || (o = "next"), t = i[o]) { if (!(t = t.call(i, u))) throw TypeError("iterator result is not an object"); if (!t.done) return t; u = t.value, c < 2 && (c = 0); } else 1 === c && (t = i["return"]) && t.call(i), c < 2 && (u = TypeError("The iterator does not provide a '" + o + "' method"), c = 1); i = e; } else if ((t = (y = G.n < 0) ? u : r.call(n, G)) !== a) break; } catch (t) { i = e, c = 1, u = t; } finally { f = 1; } } return { value: t, done: y }; }; }(r, o, i), !0), u; } var a = {}; function Generator() {} function GeneratorFunction() {} function GeneratorFunctionPrototype() {} t = Object.getPrototypeOf; var c = [][n] ? t(t([][n]())) : (_regeneratorDefine2(t = {}, n, function () { return this; }), t), u = GeneratorFunctionPrototype.prototype = Generator.prototype = Object.create(c); function f(e) { return Object.setPrototypeOf ? Object.setPrototypeOf(e, GeneratorFunctionPrototype) : (e.__proto__ = GeneratorFunctionPrototype, _regeneratorDefine2(e, o, "GeneratorFunction")), e.prototype = Object.create(u), e; } return GeneratorFunction.prototype = GeneratorFunctionPrototype, _regeneratorDefine2(u, "constructor", GeneratorFunctionPrototype), _regeneratorDefine2(GeneratorFunctionPrototype, "constructor", GeneratorFunction), GeneratorFunction.displayName = "GeneratorFunction", _regeneratorDefine2(GeneratorFunctionPrototype, o, "GeneratorFunction"), _regeneratorDefine2(u), _regeneratorDefine2(u, o, "Generator"), _regeneratorDefine2(u, n, function () { return this; }), _regeneratorDefine2(u, "toString", function () { return "[object Generator]"; }), (_regenerator = function _regenerator() { return { w: i, m: f }; })(); }
function _regeneratorDefine2(e, r, n, t) { var i = Object.defineProperty; try { i({}, "", {}); } catch (e) { i = 0; } _regeneratorDefine2 = function _regeneratorDefine(e, r, n, t) { function o(r, n) { _regeneratorDefine2(e, r, function (e) { return this._invoke(r, n, e); }); } r ? i ? i(e, r, { value: n, enumerable: !t, configurable: !t, writable: !t }) : e[r] = n : (o("next", 0), o("throw", 1), o("return", 2)); }, _regeneratorDefine2(e, r, n, t); }
function _typeof(o) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (o) { return typeof o; } : function (o) { return o && "function" == typeof Symbol && o.constructor === Symbol && o !== Symbol.prototype ? "symbol" : typeof o; }, _typeof(o); }
function _objectWithoutProperties(e, t) { if (null == e) return {}; var o, r, i = _objectWithoutPropertiesLoose(e, t); if (Object.getOwnPropertySymbols) { var n = Object.getOwnPropertySymbols(e); for (r = 0; r < n.length; r++) o = n[r], -1 === t.indexOf(o) && {}.propertyIsEnumerable.call(e, o) && (i[o] = e[o]); } return i; }
function _objectWithoutPropertiesLoose(r, e) { if (null == r) return {}; var t = {}; for (var n in r) if ({}.hasOwnProperty.call(r, n)) { if (-1 !== e.indexOf(n)) continue; t[n] = r[n]; } return t; }
function encodeString(value) {
  if (/[^\x20\x21\x23-\x5B\x5D-\uD799]/.test(value)) {
    // [^\x20-\uD799]|[\x22\x5c]
    return JSON.stringify(value);
  }
  return '"' + value + '"';
}
function stringifyChunked(value) {
  var _len,
    args,
    _key,
    _normalizeStringifyOp,
    replacer,
    getKeys,
    space,
    options,
    highWaterMark,
    roots,
    rootCount,
    keyStrings,
    stack,
    rootValue,
    prevState,
    state,
    stateValue,
    stateEmpty,
    stateKeys,
    stateIndex,
    buffer,
    _loop,
    i,
    printObject,
    printArray,
    printEntryPrelude,
    printEntry,
    pushPrimitive,
    pushState,
    popState,
    _args2 = arguments;
  return _regenerator().w(function (_context2) {
    while (1) switch (_context2.n) {
      case 0:
        popState = function _popState() {
          stack.pop();
          var value = stack.length > 0 ? stack[stack.length - 1] : rootValue;

          // restore state
          state = Array.isArray(value) ? printArray : printObject;
          stateValue = value;
          stateEmpty = false;
          stateKeys = prevState.keys;
          stateIndex = prevState.index;

          // pop state
          prevState = prevState.prev;
        };
        pushState = function _pushState() {
          prevState = {
            keys: stateKeys,
            index: stateIndex,
            prev: prevState
          };
        };
        pushPrimitive = function _pushPrimitive(value) {
          var _value$constructor;
          switch (_typeof(value)) {
            case 'string':
              buffer += encodeString(value);
              break;
            case 'number':
              buffer += Number.isFinite(value) ? String(value) : 'null';
              break;
            case 'boolean':
              buffer += value ? 'true' : 'false';
              break;
            case 'undefined':
            case 'object':
              // typeof null === 'object'
              buffer += 'null';
              break;
            default:
              throw new TypeError("Do not know how to serialize a ".concat(((_value$constructor = value.constructor) === null || _value$constructor === void 0 ? void 0 : _value$constructor.name) || _typeof(value)));
          }
        };
        printEntry = function _printEntry(key, value) {
          value = (0, _utils.replaceValue)(stateValue, key, value, replacer);
          if (value === null || _typeof(value) !== 'object') {
            // primitive
            if (state !== printObject || value !== undefined) {
              printEntryPrelude(key);
              pushPrimitive(value);
            }
          } else {
            // If the visited set does not change after adding a value, then it is already in the set
            if (stack.includes(value)) {
              throw new TypeError('Converting circular structure to JSON');
            }
            printEntryPrelude(key);
            stack.push(value);
            pushState();
            state = Array.isArray(value) ? printArray : printObject;
            stateValue = value;
            stateEmpty = true;
            stateIndex = 0;
          }
        };
        printEntryPrelude = function _printEntryPrelude(key) {
          if (stateEmpty) {
            stateEmpty = false;
          } else {
            buffer += ',';
          }
          if (space && prevState !== null) {
            buffer += "\n".concat(space.repeat(stack.length));
          }
          if (state === printObject) {
            var keyString = keyStrings.get(key);
            if (keyString === undefined) {
              keyStrings.set(key, keyString = encodeString(key) + (space ? ': ' : ':'));
            }
            buffer += keyString;
          }
        };
        printArray = function _printArray() {
          if (stateIndex === 0) {
            buffer += '[';
          }
          if (stateIndex === stateValue.length) {
            buffer += space && !stateEmpty ? "\n".concat(space.repeat(stack.length - 1), "]") : ']';
            popState();
            return;
          }
          printEntry(stateIndex, stateValue[stateIndex++]);
        };
        printObject = function _printObject() {
          if (stateIndex === 0) {
            stateKeys = getKeys(stateValue);
            buffer += '{';
          }

          // when no keys left
          if (stateIndex === stateKeys.length) {
            buffer += space && !stateEmpty ? "\n".concat(space.repeat(stack.length - 1), "}") : '}';
            popState();
            return;
          }
          var key = stateKeys[stateIndex++];
          printEntry(key, stateValue[key]);
        };
        for (_len = _args2.length, args = new Array(_len > 1 ? _len - 1 : 0), _key = 1; _key < _len; _key++) {
          args[_key - 1] = _args2[_key];
        }
        _normalizeStringifyOp = _utils.normalizeStringifyOptions.apply(void 0, args), replacer = _normalizeStringifyOp.replacer, getKeys = _normalizeStringifyOp.getKeys, space = _normalizeStringifyOp.space, options = _objectWithoutProperties(_normalizeStringifyOp, _excluded);
        highWaterMark = Number(options.highWaterMark) || 0x4000; // 16kb by default
        roots = (0, _utils.resolveStringifyMode)(options.mode) === 'jsonl' && Array.isArray(value) ? value : [value];
        rootCount = roots.length;
        keyStrings = new Map();
        stack = [];
        rootValue = null;
        prevState = null;
        state = null;
        stateValue = null;
        stateEmpty = true;
        stateKeys = [];
        stateIndex = 0;
        buffer = '';
        _loop = /*#__PURE__*/_regenerator().m(function _loop(i) {
          return _regenerator().w(function (_context) {
            while (1) switch (_context.n) {
              case 0:
                if (rootValue !== null) {
                  buffer += '\n';
                }
                rootValue = {
                  '': roots[i]
                };
                prevState = null;
                state = function state() {
                  return printEntry('', roots[i]);
                };
                stateValue = rootValue;
                stateEmpty = true;
                stateKeys = [''];
                stateIndex = 0;
              case 1:
                state();
                if (!(buffer.length >= highWaterMark || prevState === null && i === rootCount - 1)) {
                  _context.n = 3;
                  break;
                }
                _context.n = 2;
                return buffer;
              case 2:
                buffer = '';
              case 3:
                if (prevState !== null) {
                  _context.n = 1;
                  break;
                }
              case 4:
                return _context.a(2);
            }
          }, _loop);
        });
        i = 0;
      case 1:
        if (!(i < rootCount)) {
          _context2.n = 3;
          break;
        }
        return _context2.d(_regeneratorValues(_loop(i)), 2);
      case 2:
        i++;
        _context2.n = 1;
        break;
      case 3:
        return _context2.a(2);
    }
  }, _marked);
}
;
