// @discoveryjs/json-ext/src/utils.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.isIterable = isIterable;
exports.normalizeReplacer = normalizeReplacer;
exports.normalizeSpace = normalizeSpace;
exports.normalizeStringifyOptions = normalizeStringifyOptions;
exports.replaceValue = replaceValue;
exports.resolveStringifyMode = resolveStringifyMode;
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function _defineProperty(e, r, t) { return (r = _toPropertyKey(r)) in e ? Object.defineProperty(e, r, { value: t, enumerable: !0, configurable: !0, writable: !0 }) : e[r] = t, e; }
function _toPropertyKey(t) { var i = _toPrimitive(t, "string"); return "symbol" == _typeof(i) ? i : i + ""; }
function _toPrimitive(t, r) { if ("object" != _typeof(t) || !t) return t; var e = t[Symbol.toPrimitive]; if (void 0 !== e) { var i = e.call(t, r || "default"); if ("object" != _typeof(i)) return i; throw new TypeError("@@toPrimitive must return a primitive value."); } return ("string" === r ? String : Number)(t); }
function _toConsumableArray(r) { return _arrayWithoutHoles(r) || _iterableToArray(r) || _unsupportedIterableToArray(r) || _nonIterableSpread(); }
function _nonIterableSpread() { throw new TypeError("Invalid attempt to spread non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }
function _unsupportedIterableToArray(r, a) { if (r) { if ("string" == typeof r) return _arrayLikeToArray(r, a); var t = {}.toString.call(r).slice(8, -1); return "Object" === t && r.constructor && (t = r.constructor.name), "Map" === t || "Set" === t ? Array.from(r) : "Arguments" === t || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(t) ? _arrayLikeToArray(r, a) : void 0; } }
function _iterableToArray(r) { if ("undefined" != typeof Symbol && null != r[Symbol.iterator] || null != r["@@iterator"]) return Array.from(r); }
function _arrayWithoutHoles(r) { if (Array.isArray(r)) return _arrayLikeToArray(r); }
function _arrayLikeToArray(r, a) { (null == a || a > r.length) && (a = r.length); for (var e = 0, n = Array(a); e < a; e++) n[e] = r[e]; return n; }
function _typeof(o) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (o) { return typeof o; } : function (o) { return o && "function" == typeof Symbol && o.constructor === Symbol && o !== Symbol.prototype ? "symbol" : typeof o; }, _typeof(o); }
function isIterable(value) {
  return _typeof(value) === 'object' && value !== null && (typeof value[Symbol.iterator] === 'function' || typeof value[Symbol.asyncIterator] === 'function');
}
function replaceValue(holder, key, value, replacer) {
  if (value && typeof value.toJSON === 'function') {
    value = value.toJSON();
  }
  if (replacer !== null) {
    value = replacer.call(holder, String(key), value);
  }
  switch (_typeof(value)) {
    case 'function':
    case 'symbol':
      value = undefined;
      break;
    case 'object':
      if (value !== null) {
        var cls = value.constructor;
        if (cls === String || cls === Number || cls === Boolean) {
          value = value.valueOf();
        }
      }
      break;
  }
  return value;
}
function normalizeReplacer(replacer) {
  if (typeof replacer === 'function') {
    return replacer;
  }
  if (Array.isArray(replacer)) {
    var allowlist = new Set(replacer.map(function (item) {
      var cls = item && item.constructor;
      return cls === String || cls === Number ? String(item) : null;
    }).filter(function (item) {
      return typeof item === 'string';
    }));
    return _toConsumableArray(allowlist);
  }
  return null;
}
function normalizeSpace(space) {
  if (typeof space === 'number') {
    if (!Number.isFinite(space) || space < 1) {
      return false;
    }
    return ' '.repeat(Math.min(space, 10));
  }
  if (typeof space === 'string') {
    return space.slice(0, 10) || false;
  }
  return false;
}
function normalizeStringifyOptions(optionsOrReplacer, space) {
  if (optionsOrReplacer === null || Array.isArray(optionsOrReplacer) || _typeof(optionsOrReplacer) !== 'object') {
    optionsOrReplacer = {
      replacer: optionsOrReplacer,
      space: space
    };
  }
  var replacer = normalizeReplacer(optionsOrReplacer.replacer);
  var getKeys = Object.keys;
  if (Array.isArray(replacer)) {
    var allowlist = replacer;
    getKeys = function getKeys() {
      return allowlist;
    };
    replacer = null;
  }
  return _objectSpread(_objectSpread({}, optionsOrReplacer), {}, {
    replacer: replacer,
    getKeys: getKeys,
    space: normalizeSpace(optionsOrReplacer.space)
  });
}
function resolveStringifyMode() {
  var mode = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 'json';
  if (mode === 'json' || mode === 'jsonl') {
    return mode;
  }
  throw new TypeError('Invalid options: `mode` should be "json" or "jsonl"');
}
